<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>specimen.hqtb.core.bidirectional_blast &mdash; SPECIMEN 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css?v=559046d4" />
      <link rel="stylesheet" type="text/css" href="../../../../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../../../../_static/copybutton.css?v=76b2166b" />
      <link rel="stylesheet" type="text/css" href="../../../../_static/custom_theme.css?v=ef5873f9" />

  
    <link rel="shortcut icon" href="../../../../_static/LogoSPECIMEN.png"/>
  <!--[if lt IE 9]>
    <script src="../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../../../_static/documentation_options.js?v=8d563738"></script>
        <script src="../../../../_static/doctools.js?v=888ff710"></script>
        <script src="../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script src="../../../../_static/clipboard.min.js?v=a7894cd8"></script>
        <script src="../../../../_static/copybutton.js?v=7859de39"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html">
            
              <img src="../../../../_static/LogoSPECIMEN.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Content</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../overview-pipes.html">Overview of the Pipelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../specimen.html">Contents of SPECIMEN</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../help.html">Help &amp; FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../dev-notes.html">Notes for Developers</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">SPECIMEN</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">specimen.hqtb.core.bidirectional_blast</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for specimen.hqtb.core.bidirectional_blast</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Perform a bidirectional blastp using DIAMOND on an input and a template (annotated genomes).</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">__author__</span> <span class="o">=</span> <span class="s1">&#39;Carolin Brune&#39;</span>

<span class="c1">################################################################################</span>
<span class="c1"># requirements</span>
<span class="c1">################################################################################</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">pd</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">mode</span><span class="o">.</span><span class="n">chained_assignment</span> <span class="o">=</span> <span class="kc">None</span> <span class="c1"># editing on copy and saving is elsewhere</span>

<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">os.path</span>

<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Literal</span>

<span class="c1">################################################################################</span>
<span class="c1"># functions</span>
<span class="c1">################################################################################</span>

<div class="viewcode-block" id="extract_cds">
<a class="viewcode-back" href="../../../../modules/core.html#specimen.hqtb.core.bidirectional_blast.extract_cds">[docs]</a>
<span class="k">def</span> <span class="nf">extract_cds</span><span class="p">(</span><span class="n">file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">collect_info</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">identifier</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extract the CDS from a genbank file (annotated genome). </span>
<span class="sd">    Produces a FASTA-file.</span>

<span class="sd">    Args:</span>
<span class="sd">        - file (str): </span>
<span class="sd">            File to extract CDS from.</span>
<span class="sd">        - name (str): </span>
<span class="sd">            Name of the genome.</span>
<span class="sd">        - dir (str): </span>
<span class="sd">            Directory for the ouput.</span>
<span class="sd">        - collect_info (list): </span>
<span class="sd">            Feature identifiers to collect information from.</span>
<span class="sd">        - identifier (str): </span>
<span class="sd">            Feature identifier to use of the header of the FASTA.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: </span>
<span class="sd">            Name of the FASTA-file </span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="n">extension</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">))[</span><span class="mi">1</span><span class="p">]</span>


    <span class="k">match</span> <span class="n">extension</span><span class="p">:</span>

        <span class="k">case</span> <span class="s1">&#39;.gbff&#39;</span><span class="p">:</span>
            <span class="n">fasta_name</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s1">&#39;FASTA&#39;</span><span class="p">,</span><span class="n">name</span><span class="o">+</span><span class="s1">&#39;_prot.fa&#39;</span><span class="p">)</span>
            <span class="c1"># check, if CDS were already extracted</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s1">&#39;FASTA&#39;</span><span class="p">,</span><span class="n">name</span><span class="o">+</span><span class="s1">&#39;_nucs.fa&#39;</span><span class="p">))</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s1">&#39;FASTA&#39;</span><span class="p">,</span><span class="n">name</span><span class="o">+</span><span class="s1">&#39;_prot.fa&#39;</span><span class="p">)):</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;CDS for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1"> were already extracted&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># run extraction process</span>
                <span class="n">info_list</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">with</span> <span class="p">(</span>
                    <span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">,</span>
                    <span class="nb">open</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s1">&#39;FASTA&#39;</span><span class="p">,</span><span class="n">name</span><span class="o">+</span><span class="s1">&#39;_nucs.fa&#39;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">nucs_f</span><span class="p">,</span>
                    <span class="nb">open</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s1">&#39;FASTA&#39;</span><span class="p">,</span><span class="n">name</span><span class="o">+</span><span class="s1">&#39;_prot.fa&#39;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">prot_f</span><span class="p">,</span>
                <span class="p">):</span>

                    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">f</span><span class="p">,</span><span class="s2">&quot;genbank&quot;</span><span class="p">):</span>
                        <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                            <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>

                                <span class="c1"># -----------------</span>
                                <span class="c1"># extract sequences</span>
                                <span class="c1"># -----------------</span>

                                <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;CDS&#39;</span><span class="p">:</span>
                                    <span class="c1"># extract nucleotide sequence</span>
                                    <span class="n">nuc_seq</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">record</span><span class="p">)</span><span class="o">.</span><span class="n">seq</span>
                                    <span class="c1"># extract protein sequence or translate</span>
                                    <span class="k">if</span> <span class="s1">&#39;translation&#39;</span> <span class="ow">in</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                                        <span class="n">prot_seq</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;translation&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                                    <span class="k">else</span><span class="p">:</span>
                                        <span class="n">prot_seq</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">nuc_seq</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span>
                                    <span class="c1"># create header</span>
                                    <span class="k">if</span> <span class="n">identifier</span> <span class="ow">in</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                                        <span class="n">header</span> <span class="o">=</span> <span class="sa">F</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="n">identifier</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                                    <span class="k">else</span><span class="p">:</span>
                                        <span class="n">header</span> <span class="o">=</span> <span class="sa">F</span><span class="s2">&quot;&gt;unidentified_</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">&quot;</span>
                                        <span class="n">x</span> <span class="o">+=</span> <span class="mi">1</span>
                                    <span class="c1"># write to file</span>
                                    <span class="n">nucs_f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">header</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">nuc_seq</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                                    <span class="n">prot_f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">header</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">prot_seq</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

                                    <span class="c1"># ---------------------------------</span>
                                    <span class="c1"># extract extra feature information</span>
                                    <span class="c1"># ---------------------------------</span>
                                    <span class="n">temp_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">collect_info</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="n">collect_info</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="mi">0</span><span class="p">]}</span>
                                    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">collect_info</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
                                        <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                                            <span class="n">temp_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                                        <span class="k">else</span><span class="p">:</span>
                                            <span class="n">temp_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
                                    <span class="n">info_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">temp_dict</span><span class="p">)</span>

                <span class="n">info_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">info_list</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">info_df</span><span class="p">)</span><span class="si">}</span><span class="s1"> CDS were extracted from </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
                <span class="n">info_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="n">name</span><span class="o">+</span><span class="s1">&#39;_info.csv&#39;</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">fasta_name</span>

        <span class="k">case</span> <span class="s1">&#39;.faa&#39;</span><span class="p">:</span>
            <span class="c1"># no CDS extraction needed, only the info-file</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;faa extension detected. Assuming no CDS extraction is needed.&#39;</span><span class="p">)</span>
            <span class="n">info_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span><span class="n">seq</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)],</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">])</span>
            <span class="n">info_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="n">name</span><span class="o">+</span><span class="s1">&#39;_info.csv&#39;</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">fasta_name</span> <span class="o">=</span> <span class="n">file</span>
            <span class="k">return</span> <span class="n">fasta_name</span>

        <span class="k">case</span><span class="w"> </span><span class="k">_</span><span class="p">:</span>
            <span class="n">mes</span> <span class="o">=</span> <span class="sa">F</span><span class="s1">&#39;Extract_cds: Unkown file extension </span><span class="si">{</span><span class="n">extension</span><span class="si">}</span><span class="s1"> for file </span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s1">.&#39;</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">mes</span><span class="p">)</span></div>



<div class="viewcode-block" id="create_diamond_db">
<a class="viewcode-back" href="../../../../modules/core.html#specimen.hqtb.core.bidirectional_blast.create_diamond_db">[docs]</a>
<span class="k">def</span> <span class="nf">create_diamond_db</span><span class="p">(</span><span class="nb">dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">threads</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a DIAMOND database for a given protein FASTA file.</span>

<span class="sd">    Args:</span>
<span class="sd">        - dir (str): </span>
<span class="sd">            Path to the data directory.</span>
<span class="sd">        - name (str): </span>
<span class="sd">            Name of the genome/database.</span>
<span class="sd">        - path (str): </span>
<span class="sd">            Path to the FASTA-file.</span>
<span class="sd">        - threads (int): </span>
<span class="sd">            Number of threads to use.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># check if database already exists</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s1">&#39;DIAMONDdb&#39;</span><span class="p">,</span><span class="n">name</span><span class="o">+</span><span class="s1">&#39;.dmnd&#39;</span><span class="p">)):</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;database for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1"> already exists&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># generate new database using diamond makedb</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;create DIAMOND database for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1"> using:&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;diamond makedb --in </span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s1"> -d </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s2">&quot;db&quot;</span><span class="p">,</span><span class="n">name</span><span class="p">))</span><span class="si">}</span><span class="s1"> -p </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">threads</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
        <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="sa">F</span><span class="s1">&#39;diamond makedb --in </span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s1"> -d </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s2">&quot;db&quot;</span><span class="p">,</span><span class="n">name</span><span class="p">))</span><span class="si">}</span><span class="s1"> -p </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">threads</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">],</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1"> time: </span><span class="si">{</span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="si">}</span><span class="s1">s&#39;</span><span class="p">)</span></div>



<div class="viewcode-block" id="run_diamond_blastp">
<a class="viewcode-back" href="../../../../modules/core.html#specimen.hqtb.core.bidirectional_blast.run_diamond_blastp">[docs]</a>
<span class="k">def</span> <span class="nf">run_diamond_blastp</span><span class="p">(</span><span class="nb">dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">db</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">query</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">fasta_path</span><span class="p">:</span><span class="nb">str</span> <span class="p">,</span> <span class="n">sensitivity</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">threads</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Run DIAMOND blastp for a given database name and FASTA - relies on the structure</span>
<span class="sd">    created by this file (bidirectional_blast.py).</span>

<span class="sd">    Args:</span>
<span class="sd">        - dir (str): </span>
<span class="sd">            Parent directory of the place to save the files to.</span>
<span class="sd">        - db (str): </span>
<span class="sd">            Name of the genome/database used as the database.</span>
<span class="sd">        - query (str): </span>
<span class="sd">            Name of the genome used as the query.</span>
<span class="sd">        - fasta_path (str): </span>
<span class="sd">            Path to the FASTA-file containing the CDS.</span>
<span class="sd">        - sensitivity (str): </span>
<span class="sd">            Sensitivity mode to use for DIAMOND blastp.</span>
<span class="sd">        - threads (int): </span>
<span class="sd">            Number of threads that will be used for running DIAMOND</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">outname</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s1">&#39;DIAMONDblastp&#39;</span><span class="p">,</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">query</span><span class="si">}</span><span class="s1">_vs_</span><span class="si">{</span><span class="n">db</span><span class="si">}</span><span class="s1">.tsv&#39;</span><span class="p">)</span>
    <span class="c1"># check if file already exists</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">outname</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;file or filename for </span><span class="si">{</span><span class="n">query</span><span class="si">}</span><span class="s1"> vs. </span><span class="si">{</span><span class="n">db</span><span class="si">}</span><span class="s1"> blastp already exists&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># blast file</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;blast </span><span class="si">{</span><span class="n">query</span><span class="si">}</span><span class="s1"> against </span><span class="si">{</span><span class="n">db</span><span class="si">}</span><span class="s1"> using:&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;diamond blastp -d </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s2">&quot;db/&quot;</span><span class="p">,</span><span class="n">db</span><span class="o">+</span><span class="s2">&quot;.dmnd&quot;</span><span class="p">))</span><span class="si">}</span><span class="s1"> -q </span><span class="si">{</span><span class="sa">F</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="si">}</span><span class="s1"> --</span><span class="si">{</span><span class="n">sensitivity</span><span class="si">}</span><span class="s1"> -p </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">threads</span><span class="p">)</span><span class="si">}</span><span class="s1"> -o </span><span class="si">{</span><span class="n">outname</span><span class="si">}</span><span class="s1"> --outfmt 6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qlen &#39;</span><span class="p">)</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
        <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="sa">F</span><span class="s1">&#39;diamond blastp -d </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s2">&quot;db&quot;</span><span class="p">,</span><span class="n">db</span><span class="o">+</span><span class="s2">&quot;.dmnd&quot;</span><span class="p">))</span><span class="si">}</span><span class="s1"> -q </span><span class="si">{</span><span class="sa">F</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="si">}</span><span class="s1"> --</span><span class="si">{</span><span class="n">sensitivity</span><span class="si">}</span><span class="s1"> -p </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">threads</span><span class="p">)</span><span class="si">}</span><span class="s1"> -o </span><span class="si">{</span><span class="n">outname</span><span class="si">}</span><span class="s1"> --outfmt 6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qlen &#39;</span><span class="p">],</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">time: </span><span class="si">{</span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="si">}</span><span class="s1">s&#39;</span><span class="p">)</span></div>



<div class="viewcode-block" id="bdbp_diamond">
<a class="viewcode-back" href="../../../../modules/core.html#specimen.hqtb.core.bidirectional_blast.bdbp_diamond">[docs]</a>
<span class="k">def</span> <span class="nf">bdbp_diamond</span><span class="p">(</span><span class="nb">dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">template_name</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">input_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">template_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">input_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sensitivity</span><span class="o">=</span><span class="s1">&#39;sensitive&#39;</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="mi">2</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Perform bidirectional blastp using DIAMOND.</span>

<span class="sd">    Args:</span>
<span class="sd">        - dir (str): </span>
<span class="sd">            Path to the directory parent to in/out.</span>
<span class="sd">        - template_name (str): </span>
<span class="sd">            Name of the template genome.</span>
<span class="sd">        - input_name (str): </span>
<span class="sd">            Name of the input genome.</span>
<span class="sd">        - template_path (str): </span>
<span class="sd">            Path to the CDS FASTA-file of the template.</span>
<span class="sd">        - input_path (str): </span>
<span class="sd">            Path to the CDS FASTA-file of the input.</span>
<span class="sd">        - sensitivity (str, optional): </span>
<span class="sd">            Sensitivity mode for DIAMOND. </span>
<span class="sd">            Defaults to &#39;sensitive&#39;.</span>
<span class="sd">        - threads (int, optional): </span>
<span class="sd">            Number of threads to use when running DIAMOND. </span>
<span class="sd">            Defaults to 2.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># -----------------------</span>
    <span class="c1"># create output directory</span>
    <span class="c1"># -----------------------</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s2">&quot;db&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;Creating new directory </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s2">&quot;db&quot;</span><span class="p">))</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">FileExistsError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Given directory already has required structure.&#39;</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s2">&quot;DIAMONDblastp&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;Creating new directory </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s2">&quot;DIAMONDblastp&quot;</span><span class="p">))</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">FileExistsError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Given directory already has required structure.&#39;</span><span class="p">)</span>

    <span class="c1"># ---------------------</span>
    <span class="c1"># make DIAMOND database</span>
    <span class="c1"># ---------------------</span>

    <span class="n">create_diamond_db</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span> <span class="n">template_name</span><span class="p">,</span> <span class="n">template_path</span><span class="p">,</span> <span class="n">threads</span><span class="p">)</span>
    <span class="n">create_diamond_db</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span> <span class="n">input_name</span><span class="p">,</span> <span class="n">input_path</span><span class="p">,</span> <span class="n">threads</span><span class="p">)</span>

    <span class="c1"># ----------------------</span>
    <span class="c1"># perform diamond blastp</span>
    <span class="c1"># ----------------------</span>

    <span class="n">run_diamond_blastp</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span> <span class="n">template_name</span><span class="p">,</span> <span class="n">input_name</span><span class="p">,</span> <span class="n">input_path</span><span class="p">,</span> <span class="n">sensitivity</span><span class="p">,</span> <span class="n">threads</span><span class="p">)</span>
    <span class="n">run_diamond_blastp</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span> <span class="n">input_name</span><span class="p">,</span> <span class="n">template_name</span><span class="p">,</span> <span class="n">template_path</span><span class="p">,</span> <span class="n">sensitivity</span><span class="p">,</span> <span class="n">threads</span><span class="p">)</span></div>



<div class="viewcode-block" id="extract_bestbdbp_hits">
<a class="viewcode-back" href="../../../../modules/core.html#specimen.hqtb.core.bidirectional_blast.extract_bestbdbp_hits">[docs]</a>
<span class="k">def</span> <span class="nf">extract_bestbdbp_hits</span><span class="p">(</span><span class="n">tvq</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">qvt</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">cov</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.25</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extract the best directional blastp hits from two tsv files,</span>
<span class="sd">    which were generate by :py:func`~specimen.core.bidirectional_blast.bdbp_diamond` </span>
<span class="sd">    generated or similar steps.</span>

<span class="sd">    Args:</span>
<span class="sd">        - tvq (str): </span>
<span class="sd">            Path to the template vs. query file.</span>
<span class="sd">        - qvt (str): </span>
<span class="sd">            Path to the query vs. template file.</span>
<span class="sd">        - name (str): </span>
<span class="sd">            Name (path) of the output file.</span>
<span class="sd">        - cov (float, optional): </span>
<span class="sd">            Cut-off value for the coverage.</span>
<span class="sd">            All hits with coverage &lt; cov will be excluded. </span>
<span class="sd">            Defaults to 0.25.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># default value for coverage (currently) based on Norsigian et al. 2020</span>

    <span class="c1"># read in data and filter for coverage and percentage identity values (PID)</span>
    <span class="n">col_names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;query_ID&#39;</span><span class="p">,</span> <span class="s1">&#39;subject_ID&#39;</span><span class="p">,</span> <span class="s1">&#39;PID&#39;</span><span class="p">,</span> <span class="s1">&#39;align_len&#39;</span><span class="p">,</span> <span class="s1">&#39;no_mismatch&#39;</span><span class="p">,</span> <span class="s1">&#39;no_gapopen&#39;</span><span class="p">,</span> <span class="s1">&#39;query_start&#39;</span><span class="p">,</span> <span class="s1">&#39;query_end&#39;</span><span class="p">,</span> <span class="s1">&#39;subject_start&#39;</span><span class="p">,</span> <span class="s1">&#39;subject_end&#39;</span><span class="p">,</span><span class="s1">&#39;E-value&#39;</span><span class="p">,</span><span class="s1">&#39;bitscore&#39;</span><span class="p">,</span><span class="s1">&#39;query_length&#39;</span><span class="p">,</span> <span class="s1">&#39;reciprocal&#39;</span><span class="p">]</span>
    <span class="c1"># template vs. query</span>
    <span class="n">bh_tvq</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">tvq</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">col_names</span><span class="p">)</span>
    <span class="n">bh_tvq</span><span class="p">[</span><span class="s1">&#39;coverage&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bh_tvq</span><span class="p">[</span><span class="s1">&#39;align_len&#39;</span><span class="p">]</span><span class="o">/</span><span class="n">bh_tvq</span><span class="p">[</span><span class="s1">&#39;query_length&#39;</span><span class="p">]</span>
    <span class="n">bh_tvq</span> <span class="o">=</span> <span class="n">bh_tvq</span><span class="p">[</span><span class="n">bh_tvq</span><span class="p">[</span><span class="s1">&#39;coverage&#39;</span><span class="p">]</span><span class="o">&gt;=</span><span class="n">cov</span><span class="p">]</span>
    <span class="c1"># query vs. template</span>
    <span class="n">bh_qvt</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">qvt</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">col_names</span><span class="p">)</span>
    <span class="n">bh_qvt</span><span class="p">[</span><span class="s1">&#39;coverage&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bh_qvt</span><span class="p">[</span><span class="s1">&#39;align_len&#39;</span><span class="p">]</span><span class="o">/</span><span class="n">bh_qvt</span><span class="p">[</span><span class="s1">&#39;query_length&#39;</span><span class="p">]</span>
    <span class="n">bh_qvt</span> <span class="o">=</span> <span class="n">bh_qvt</span><span class="p">[</span><span class="n">bh_qvt</span><span class="p">[</span><span class="s1">&#39;coverage&#39;</span><span class="p">]</span><span class="o">&gt;=</span><span class="n">cov</span><span class="p">]</span>
    <span class="c1"># compare dataframes to extract bidirectional best hits</span>
    <span class="n">list_best_hits</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="k">for</span> <span class="nb">id</span> <span class="ow">in</span> <span class="n">bh_qvt</span><span class="p">[</span><span class="s1">&#39;query_ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="c1"># get best hit from query vs template</span>
        <span class="n">all_hits</span> <span class="o">=</span> <span class="n">bh_qvt</span><span class="p">[</span><span class="n">bh_qvt</span><span class="p">[</span><span class="s1">&#39;query_ID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="nb">id</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">all_hits</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">best_hit</span> <span class="o">=</span> <span class="n">all_hits</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">all_hits</span><span class="p">[</span><span class="s1">&#39;PID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">idxmax</span><span class="p">()]</span>
        <span class="c1"># get best reciprocal hit, if it exists</span>
        <span class="n">all_re_hits</span> <span class="o">=</span> <span class="n">bh_tvq</span><span class="p">[</span><span class="n">bh_tvq</span><span class="p">[</span><span class="s1">&#39;query_ID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">best_hit</span><span class="p">[</span><span class="s1">&#39;subject_ID&#39;</span><span class="p">]]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">all_re_hits</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">best_re_hit</span> <span class="o">=</span> <span class="n">all_re_hits</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">all_re_hits</span><span class="p">[</span><span class="s1">&#39;PID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">idxmax</span><span class="p">()]</span>
        <span class="c1"># add to list of best hits, if they are the same</span>
        <span class="k">if</span> <span class="n">best_re_hit</span><span class="o">.</span><span class="n">subject_ID</span> <span class="o">==</span> <span class="n">best_hit</span><span class="o">.</span><span class="n">query_ID</span><span class="p">:</span>
            <span class="n">best_hit</span><span class="p">[</span><span class="s1">&#39;reciprocal&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">list_best_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">best_hit</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">best_hit</span><span class="p">[</span><span class="s1">&#39;reciprocal&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">list_best_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">best_hit</span><span class="p">)</span>
    <span class="c1"># write results into file</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">list_best_hits</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">col_names</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="p">)</span><span class="si">}</span><span class="s1"> blast best hits with the template genome&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="p">[</span><span class="n">out</span><span class="p">[</span><span class="s2">&quot;reciprocal&quot;</span><span class="p">]</span><span class="o">==</span><span class="mi">1</span><span class="p">])</span><span class="si">}</span><span class="s1"> reciprocal and </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">out</span><span class="p">[</span><span class="n">out</span><span class="p">[</span><span class="s2">&quot;reciprocal&quot;</span><span class="p">]</span><span class="o">==</span><span class="mi">0</span><span class="p">])</span><span class="si">}</span><span class="s1"> query blast best hits&#39;</span><span class="p">)</span>
    <span class="n">out</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">F</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">.tsv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="run">
<a class="viewcode-back" href="../../../../modules/core.html#specimen.hqtb.core.bidirectional_blast.run">[docs]</a>
<span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="n">template</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="nb">input</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dir</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">template_name</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">input_name</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> 
        <span class="n">temp_header</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">&#39;protein_id&#39;</span><span class="p">,</span> <span class="n">in_header</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">,</span> 
        <span class="n">threads</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> 
        <span class="n">extra_info</span><span class="p">:</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">,</span> <span class="s1">&#39;product&#39;</span><span class="p">,</span> <span class="s1">&#39;protein_id&#39;</span><span class="p">],</span> 
        <span class="n">sensitivity</span><span class="p">:</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;sensitive&#39;</span><span class="p">,</span> <span class="s1">&#39;more-sensitive&#39;</span><span class="p">,</span> <span class="s1">&#39;very-sensitive&#39;</span><span class="p">,</span> <span class="s1">&#39;ultra-sensitive&#39;</span><span class="p">]</span><span class="o">=</span><span class="s1">&#39;more-sensitive&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Run the bidirectional blast on a template and input genome (annotated).</span>

<span class="sd">    Args:</span>
<span class="sd">        - template (str): </span>
<span class="sd">            Path to the annotated genome file used as a template.</span>
<span class="sd">        - input (str): </span>
<span class="sd">            Path to the annotated genome file used as a input.</span>
<span class="sd">        - dir (str): </span>
<span class="sd">            Path to the output directory.</span>
<span class="sd">        - template_name (str, optional): </span>
<span class="sd">            Name of the annotated genome file used as a template.. </span>
<span class="sd">            Defaults to None.</span>
<span class="sd">        - input_name (str, optional): </span>
<span class="sd">            Name of the annotated genome file used as input.. </span>
<span class="sd">            Defaults to None.</span>
<span class="sd">        - temp_header (str, optional): </span>
<span class="sd">            Feature qualifier of the gbff (NCBI) / faa (PROKKA) of the template to use as header for the FASTA files.</span>
<span class="sd">            If None is given, sets it based on file extension (currently only implememted for gbff and faa).</span>
<span class="sd">            Defaults to &#39;protein_id&#39;.</span>
<span class="sd">        - in_header (str, optional): </span>
<span class="sd">            Feature qualifier of the gbff (NCBI) / faa (PROKKA) of the input to use as header for the FASTA files. </span>
<span class="sd">            If None is given, sets it based on file extension (currently only implememted for gbff and faa).</span>
<span class="sd">            Defaults to &#39;locus_tag&#39;.</span>
<span class="sd">        - threads (int, optional):  </span>
<span class="sd">            Number of threads to be used for DIAMOND. </span>
<span class="sd">            Defaults to 2.</span>
<span class="sd">        - extra_info (list[str], optional):  </span>
<span class="sd">            List of feature qualifiers to be extracted from the annotated genome files as additional information. </span>
<span class="sd">            Defaults to [&#39;locus_tag&#39;, &#39;product&#39;, &#39;protein_id&#39;].</span>
<span class="sd">        - sensitivity (Literal[&#39;sensitive&#39;, &#39;more-sensitive&#39;, &#39;very-sensitive&#39;, &#39;ultra-sensitive&#39;], optional): </span>
<span class="sd">            Sensitivity mode for DIAMOND blastp run.. </span>
<span class="sd">            Defaults to &#39;more-sensitive&#39;.</span>

<span class="sd">    Raises:</span>
<span class="sd">        - ValueError: Unknown file extension. Please set value for temp_header manually or check file.</span>
<span class="sd">        - ValueError: Unknown file extension. Please set value for in_header manually or check file.</span>
<span class="sd">        - ValueError: Unknown sensitive mode</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">total_time_s</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="c1"># -----------</span>
    <span class="c1"># check input</span>
    <span class="c1"># -----------</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">template_name</span><span class="p">:</span>
        <span class="n">template_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">template</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">input_name</span><span class="p">:</span>
        <span class="n">input_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="nb">input</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">temp_header</span><span class="p">:</span>
        <span class="k">if</span> <span class="s1">&#39;.gbff&#39;</span> <span class="o">==</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">template</span><span class="p">)[</span><span class="mi">1</span><span class="p">]:</span>
            <span class="n">temp_header</span> <span class="o">=</span> <span class="s1">&#39;protein_id&#39;</span>
        <span class="k">elif</span> <span class="s1">&#39;.faa&#39;</span> <span class="o">==</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">template</span><span class="p">)[</span><span class="mi">1</span><span class="p">]:</span>
            <span class="n">temp_header</span> <span class="o">=</span> <span class="s1">&#39;locus_tag&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Unknown file extension. Please set value for temp_header manually or check file.&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">in_header</span><span class="p">:</span>
        <span class="k">if</span> <span class="s1">&#39;.gbff&#39;</span> <span class="o">==</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="nb">input</span><span class="p">)[</span><span class="mi">1</span><span class="p">]:</span>
            <span class="n">in_header</span> <span class="o">=</span> <span class="s1">&#39;protein_id&#39;</span>
        <span class="k">elif</span> <span class="s1">&#39;.faa&#39;</span> <span class="o">==</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="nb">input</span><span class="p">)[</span><span class="mi">1</span><span class="p">]:</span>
            <span class="n">in_header</span> <span class="o">=</span> <span class="s1">&#39;locus_tag&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Unknown file extension. Please set value for in_header manually or check file.&#39;</span><span class="p">)</span>


    <span class="k">if</span> <span class="n">sensitivity</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;sensitive&#39;</span><span class="p">,</span><span class="s1">&#39;more-sensitive&#39;</span><span class="p">,</span><span class="s1">&#39;very-sensitive&#39;</span><span class="p">,</span><span class="s1">&#39;ultra-sensitive&#39;</span><span class="p">]:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;Unknown sensitive mode </span><span class="si">{</span><span class="n">sensitivity</span><span class="si">}</span><span class="s1">. Please choose one of the following: sensitive, more-sensitive, very-sensitive, ultra-sensitive&#39;</span><span class="p">)</span>

    <span class="c1"># ......................</span>
    <span class="c1"># @TODO</span>
    <span class="c1">#    when to use protein_id and when to use locus_tag as default (prokka vs ncbi)</span>
    <span class="c1"># ......................</span>

    <span class="c1"># -------------</span>
    <span class="c1"># start program</span>
    <span class="c1"># -------------</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">bidirectional blast</span><span class="se">\n</span><span class="s1">################################################################################</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># ---------------------</span>
    <span class="c1"># extract CDS sequences</span>
    <span class="c1"># ---------------------</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1"># ----------------------</span><span class="se">\n</span><span class="s1"># extract CDS into FASTA</span><span class="se">\n</span><span class="s1"># ----------------------</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s2">&quot;FASTA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;Creating new directory </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s2">&quot;FASTA&quot;</span><span class="p">))</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">FileExistsError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Given directory already has required structure.&#39;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;processing template genome...&#39;</span><span class="p">)</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="n">temp_fasta</span> <span class="o">=</span> <span class="n">extract_cds</span><span class="p">(</span><span class="n">template</span><span class="p">,</span> <span class="n">template_name</span><span class="p">,</span> <span class="nb">dir</span><span class="p">,</span> <span class="n">extra_info</span><span class="p">,</span> <span class="n">temp_header</span><span class="p">)</span>
    <span class="n">end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">total time: </span><span class="si">{</span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="si">}</span><span class="s1">s&#39;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;processing genome of interest...&#39;</span><span class="p">)</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="n">inpu_fasta</span> <span class="o">=</span> <span class="n">extract_cds</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span> <span class="n">input_name</span><span class="p">,</span> <span class="nb">dir</span><span class="p">,</span> <span class="n">extra_info</span><span class="p">,</span> <span class="n">in_header</span><span class="p">)</span>
    <span class="n">end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">total time: </span><span class="si">{</span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="si">}</span><span class="s1">s&#39;</span><span class="p">)</span>


    <span class="c1"># ----------------------------</span>
    <span class="c1"># perform bidirectional blastp</span>
    <span class="c1"># ----------------------------</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1"># ----------------------------</span><span class="se">\n</span><span class="s1"># perform bidirectional blastp</span><span class="se">\n</span><span class="s1"># ----------------------------</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># reciprical blast - diamond version</span>
    <span class="n">bdbp_diamond</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span> <span class="n">template_name</span><span class="p">,</span> <span class="n">input_name</span><span class="p">,</span> <span class="n">temp_fasta</span><span class="p">,</span> <span class="n">inpu_fasta</span><span class="p">,</span> <span class="n">sensitivity</span><span class="p">,</span> <span class="n">threads</span><span class="p">)</span>

    <span class="c1"># ----------------------------</span>
    <span class="c1"># find best bidirectional hits</span>
    <span class="c1"># ----------------------------</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1"># ----------------------------</span><span class="se">\n</span><span class="s1"># find best bidirectional hits</span><span class="se">\n</span><span class="s1"># ----------------------------</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="n">extract_bestbdbp_hits</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s1">&#39;DIAMONDblastp&#39;</span><span class="p">,</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">template_name</span><span class="si">}</span><span class="s1">_vs_</span><span class="si">{</span><span class="n">input_name</span><span class="si">}</span><span class="s1">.tsv&#39;</span><span class="p">),</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s1">&#39;DIAMONDblastp&#39;</span><span class="p">,</span><span class="sa">F</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">input_name</span><span class="si">}</span><span class="s1">_vs_</span><span class="si">{</span><span class="n">template_name</span><span class="si">}</span><span class="s1">.tsv&#39;</span><span class="p">),</span> <span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="sa">F</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">input_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">template_name</span><span class="si">}</span><span class="s1">_bbh&#39;</span><span class="p">))</span>
    <span class="n">end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">total time: </span><span class="si">{</span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="si">}</span><span class="s1">s&#39;</span><span class="p">)</span>

    <span class="n">total_time_e</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;total runtime: </span><span class="si">{</span><span class="n">total_time_e</span><span class="o">-</span><span class="n">total_time_s</span><span class="si">}</span><span class="s1">s&#39;</span><span class="p">)</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Carolin Brune.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>