<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>specimen.core package &mdash; SPECIMEN 1.0.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="specimen.core.refinement package" href="specimen.core.refinement.html" />
    <link rel="prev" title="specimen.classes package" href="specimen.classes.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            SPECIMEN
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="about-pipeline.html">About the SPECIMEN</a></li>
<li class="toctree-l1"><a class="reference internal" href="run-pipeline.html">Run the Pipeline</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="specimen.html">Contents of SPECIMEN</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="specimen.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="specimen.classes.html">specimen.classes package</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">specimen.core package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#subpackages">Subpackages</a></li>
<li class="toctree-l4"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-specimen.core.analysis">specimen.core.analysis module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-specimen.core.bidirectional_blast">specimen.core.bidirectional_blast module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-specimen.core.generate_draft_model">specimen.core.generate_draft_model module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-specimen.core.validation">specimen.core.validation module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-specimen.core">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="specimen.util.html">specimen.util package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="specimen.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="specimen.html#module-specimen.cmd_access">specimen.cmd_access module</a></li>
<li class="toctree-l2"><a class="reference internal" href="specimen.html#module-specimen.workflow">specimen.workflow module</a></li>
<li class="toctree-l2"><a class="reference internal" href="specimen.html#module-specimen">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="dev-notes.html">Notes for Developers</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">SPECIMEN</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="specimen.html">Contents of SPECIMEN</a></li>
      <li class="breadcrumb-item active">specimen.core package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/specimen.core.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="specimen-core-package">
<h1>specimen.core package<a class="headerlink" href="#specimen-core-package" title="Permalink to this heading"></a></h1>
<section id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this heading"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="specimen.core.refinement.html">specimen.core.refinement package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="specimen.core.refinement.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="specimen.core.refinement.html#module-specimen.core.refinement.annotation">specimen.core.refinement.annotation module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.annotation.kegg_reaction_to_kegg_pathway"><code class="docutils literal notranslate"><span class="pre">kegg_reaction_to_kegg_pathway()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.annotation.run"><code class="docutils literal notranslate"><span class="pre">run()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="specimen.core.refinement.html#module-specimen.core.refinement.cleanup">specimen.core.refinement.cleanup module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.cleanup.check_direction"><code class="docutils literal notranslate"><span class="pre">check_direction()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.cleanup.cobra_gapfill_wrapper"><code class="docutils literal notranslate"><span class="pre">cobra_gapfill_wrapper()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.cleanup.complete_BioMetaCyc"><code class="docutils literal notranslate"><span class="pre">complete_BioMetaCyc()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.cleanup.multiple_cobra_gapfill"><code class="docutils literal notranslate"><span class="pre">multiple_cobra_gapfill()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.cleanup.resolve_duplicate_metabolites"><code class="docutils literal notranslate"><span class="pre">resolve_duplicate_metabolites()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.cleanup.resolve_duplicate_reactions"><code class="docutils literal notranslate"><span class="pre">resolve_duplicate_reactions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.cleanup.resolve_duplicates"><code class="docutils literal notranslate"><span class="pre">resolve_duplicates()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.cleanup.run"><code class="docutils literal notranslate"><span class="pre">run()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.cleanup.single_cobra_gapfill"><code class="docutils literal notranslate"><span class="pre">single_cobra_gapfill()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="specimen.core.refinement.html#module-specimen.core.refinement.extension">specimen.core.refinement.extension module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.add_gene"><code class="docutils literal notranslate"><span class="pre">add_gene()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.add_reaction"><code class="docutils literal notranslate"><span class="pre">add_reaction()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.build_metabolite_kegg"><code class="docutils literal notranslate"><span class="pre">build_metabolite_kegg()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.build_metabolite_mnx"><code class="docutils literal notranslate"><span class="pre">build_metabolite_mnx()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.create_reac_id"><code class="docutils literal notranslate"><span class="pre">create_reac_id()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.extent_model"><code class="docutils literal notranslate"><span class="pre">extent_model()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.find_best_diamond_hits"><code class="docutils literal notranslate"><span class="pre">find_best_diamond_hits()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.get_metabolites_kegg"><code class="docutils literal notranslate"><span class="pre">get_metabolites_kegg()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.get_metabolites_mnx"><code class="docutils literal notranslate"><span class="pre">get_metabolites_mnx()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.get_ncbi_info"><code class="docutils literal notranslate"><span class="pre">get_ncbi_info()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.identify_missing_genes"><code class="docutils literal notranslate"><span class="pre">identify_missing_genes()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.isreaction_complete"><code class="docutils literal notranslate"><span class="pre">isreaction_complete()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.make_reaction_dict"><code class="docutils literal notranslate"><span class="pre">make_reaction_dict()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.map_BiGG_reactions"><code class="docutils literal notranslate"><span class="pre">map_BiGG_reactions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.map_Bigg_reactions_row"><code class="docutils literal notranslate"><span class="pre">map_Bigg_reactions_row()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.map_to_KEGG"><code class="docutils literal notranslate"><span class="pre">map_to_KEGG()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.map_to_KEGG_row"><code class="docutils literal notranslate"><span class="pre">map_to_KEGG_row()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.map_to_NCBI_efetch_row"><code class="docutils literal notranslate"><span class="pre">map_to_NCBI_efetch_row()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.run"><code class="docutils literal notranslate"><span class="pre">run()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="specimen.core.refinement.html#module-specimen.core.refinement.smoothing">specimen.core.refinement.smoothing module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.smoothing.adjust_BOF"><code class="docutils literal notranslate"><span class="pre">adjust_BOF()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.smoothing.check_for_cycles"><code class="docutils literal notranslate"><span class="pre">check_for_cycles()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.smoothing.perform_mcc"><code class="docutils literal notranslate"><span class="pre">perform_mcc()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.smoothing.run"><code class="docutils literal notranslate"><span class="pre">run()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="specimen.core.refinement.html#module-specimen.core.refinement">Module contents</a></li>
</ul>
</li>
</ul>
</div>
</section>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading"></a></h2>
</section>
<section id="module-specimen.core.analysis">
<span id="specimen-core-analysis-module"></span><h2>specimen.core.analysis module<a class="headerlink" href="#module-specimen.core.analysis" title="Permalink to this heading"></a></h2>
<p>Analyse a model (part 5 of the workflow).</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.analysis.analyse_growth">
<span class="sig-prename descclassname"><span class="pre">specimen.core.analysis.</span></span><span class="sig-name descname"><span class="pre">analyse_growth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_media</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">change_to_aerobic</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">change_to_anaerobic</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_casamino</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">external_media</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_min_medium</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">growth_rate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_aa_auxotrophies</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.analysis.analyse_growth" title="Permalink to this definition"></a></dt>
<dd><p>Analyse the growth of a model on different media.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The model to be tested.</p></li>
<li><p><strong>db_path</strong> (<em>string</em><em> (</em><em>optional</em><em>)</em>) – Path to a media database file.</p></li>
<li><p><strong>load_media</strong> (<em>list</em><em>, </em><em>elements are strings</em><em> (</em><em>optional</em><em>)</em>) – List of medium identifiers to be loaded from the database.</p></li>
<li><p><strong>change_to_aerobic</strong> – List of medium identifiers to be loaded from the database
and changed to fullfill the aerobic criteria.</p></li>
<li><p><strong>change_to_aerobic</strong> – List of medium identifiers to be loaded from the database
and changed to fullfill the anaerobic criteria.</p></li>
<li><p><strong>add_casamino</strong> (<em>list</em><em>, </em><em>elements are strings</em><em> (</em><em>optional</em><em>)</em>) – List of medium identifiers to be loaded from the database
Afterwards, add the casamino acids to the medium.</p></li>
<li><p><strong>external_media</strong> (<em>list</em><em>, </em><em>elements are strings</em><em> (</em><em>optional</em><em>)</em>) – List of paths to files containing exactly one medium description.
Offers the user the possibility to use personal media without adding them into a database.</p></li>
<li><p><strong>test_min_medium</strong> (<em>bool</em>) – Option to search for a minimal medium based on the
exchange reactions. This can take a while,if the number of exchange
reactions is large. Default is False. Requires the :growth_rate:
parameter to be higher than 0.0.</p></li>
<li><p><strong>growth_rate</strong> (<em>float</em><em>, </em><em>should be &gt; 0.0.</em>) – Growth rate that should be reached by the minimal medium.
Default is 0.5. Rule of thumb: higher values lead to faster runtime.</p></li>
<li><p><strong>test_aa_auxotrophies</strong> (<em>bool</em>) – Enable the auxotrophy test.
Default is True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The report of the analysis</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="specimen.classes.html#specimen.classes.reports.GrowthAnalysisReport" title="specimen.classes.reports.GrowthAnalysisReport">classes.reports.GrowthAnalysisReport</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.analysis.auxotrophies_simulation">
<span class="sig-prename descclassname"><span class="pre">specimen.core.analysis.</span></span><span class="sig-name descname"><span class="pre">auxotrophies_simulation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">media_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minimal</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.analysis.auxotrophies_simulation" title="Permalink to this definition"></a></dt>
<dd><p>Perform auxotrophy test for amino acids on a model and a list of media.</p>
<p>Tests, if the model growths on the media and if and with what fluxes the
20 proteinogenic amino acids are produced by temporarily adding a
sink reaction for each of the amino acids to the model as the objective function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The model to be tested.</p></li>
<li><p><strong>media_list</strong> (<em>list</em><em>, </em><em>elements</em><em> of </em><em>type classes.medium.Medium</em>) – list of media to be tested.</p></li>
<li><p><strong>minimal</strong> (<a class="reference internal" href="specimen.classes.html#specimen.classes.medium.Medium" title="specimen.classes.medium.Medium"><em>classes.medium.Medium</em></a><em>, </em><em>optional</em>) – A previously generated minimal medium.</p></li>
<li><p><strong>show_names</strong> (<em>bool</em>) – Use the names of the amino acids as axis labels instead of the BiGG IDs.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The results of the tests.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>nested dict, {&lt;medium.name&gt;:{&lt;aa.name&gt;:growth_rate}}</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.analysis.growth_simulation">
<span class="sig-prename descclassname"><span class="pre">specimen.core.analysis.</span></span><span class="sig-name descname"><span class="pre">growth_simulation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">media_list</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.analysis.growth_simulation" title="Permalink to this definition"></a></dt>
<dd><p>Perform growth simulation for a list of media.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The model for simulation.</p></li>
<li><p><strong>media_list</strong> (<em>list</em><em> of </em><a class="reference internal" href="specimen.classes.html#specimen.classes.medium.Medium" title="specimen.classes.medium.Medium"><em>classes.medium.Medium</em></a>) – List of media to be tested.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The results.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict, classes.medium.name:growth_rate</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.analysis.kegg_pathway_analysis">
<span class="sig-prename descclassname"><span class="pre">specimen.core.analysis.</span></span><span class="sig-name descname"><span class="pre">kegg_pathway_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.analysis.kegg_pathway_analysis" title="Permalink to this definition"></a></dt>
<dd><p>Analyse the pathways that are covered by the model.</p>
<p>The analysis is based on the KEGG pathway classification and the available
KEGG pathway identifiers present in the model.
Note: one reaction can have multiple pathway identifiers associated with it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>model</strong> (<em>cobra.Model</em>) – The GEM to be analysed</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The results of the analysis.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="specimen.classes.html#specimen.classes.reports.PathwayAnalysisReport" title="specimen.classes.reports.PathwayAnalysisReport">PathwayAnalysisReport</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.analysis.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.analysis.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc_model_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc_based_on</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'id'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_media</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['LB',</span> <span class="pre">'M9']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">change_to_aerobic</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">change_to_anaerobic</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_casamino</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">external_media</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_min_medium</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">growth_rate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_aa_auxotrophies</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pathway</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.analysis.run" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-specimen.core.bidirectional_blast">
<span id="specimen-core-bidirectional-blast-module"></span><h2>specimen.core.bidirectional_blast module<a class="headerlink" href="#module-specimen.core.bidirectional_blast" title="Permalink to this heading"></a></h2>
<p>Perform a bidirectional blastp using DIAMOND on an input and a template (annotated genomes).</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.bdbp_diamond">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">bdbp_diamond</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">template_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">template_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensitivity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sensitive'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.bidirectional_blast.bdbp_diamond" title="Permalink to this definition"></a></dt>
<dd><p>Perform bidirectional blastp using DIAMOND.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dir</strong> (<em>string</em>) – path to the working directory</p></li>
<li><p><strong>template_name</strong> (<em>string</em>) – name of the template genome</p></li>
<li><p><strong>input_name</strong> (<em>string</em>) – name of the input genome</p></li>
<li><p><strong>template_path</strong> (<em>string</em>) – path of the template genome fasta file (for the CDS)</p></li>
<li><p><strong>input_path</strong> (<em>string</em>) – path of the input genome fasta file (for the CDS)</p></li>
<li><p><strong>sensitivity</strong> (<em>string</em>) – sensitivity mode of DIAMOND, default is mid-sensitive</p></li>
<li><p><strong>threads</strong> (<em>int</em>) – number of threads to be used for computation, default is 2</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.create_diamond_db">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">create_diamond_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.bidirectional_blast.create_diamond_db" title="Permalink to this definition"></a></dt>
<dd><p>Create a DIAMOND database for a given protein FASTA file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dir</strong> (<em>string</em>) – path to the data directory</p></li>
<li><p><strong>name</strong> (<em>string</em>) – name of the genome</p></li>
<li><p><strong>path</strong> (<em>string</em>) – path to the FASTA file</p></li>
<li><p><strong>threads</strong> (<em>int</em>) – number of threads to be used</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.extract_bestbdbp_hits">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">extract_bestbdbp_hits</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tvq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qvt</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cov</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.25</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.bidirectional_blast.extract_bestbdbp_hits" title="Permalink to this definition"></a></dt>
<dd><p>Extract the best directional blastp hits from two tsv files,
which were generate by bdbp_diamond() generated or similar steps.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tvq</strong> (<em>string</em>) – path to the template_vs_query file.</p></li>
<li><p><strong>qvt</strong> (<em>string</em>) – path to the query_vs_template file.</p></li>
<li><p><strong>name</strong> (<em>string</em>) – name (path) of the output file.</p></li>
<li><p><strong>cov</strong> (<em>float</em>) – cutoff value for the coverage. The default if 0.25, which means all hits with coverage &lt;0.25 will be excluded.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.extract_cds">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">extract_cds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collect_info</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">identifier</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.bidirectional_blast.extract_cds" title="Permalink to this definition"></a></dt>
<dd><p>Extract the CDS from a genbank file (annotated genome). Produces a FASTA
file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file</strong> (<em>string</em>) – filename to extract CDS from</p></li>
<li><p><strong>dir</strong> (<em>string</em>) – directory for the generated files</p></li>
<li><p><strong>name</strong> (<em>string</em>) – name of the genome</p></li>
<li><p><strong>collect_info</strong> (<em>list</em>) – feature identifiers to collect information from</p></li>
<li><p><strong>identifier</strong> (<em>string</em>) – feature identifier to use as a header for the FASTA files</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">template</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">template_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp_header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'protein_id'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'locus_tag'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_info</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['locus_tag',</span> <span class="pre">'product',</span> <span class="pre">'protein_id']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensitivity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'more-sensitive'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.bidirectional_blast.run" title="Permalink to this definition"></a></dt>
<dd><p>Run the bidirectional blast on a template and inpute genome (annotated).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>template</strong> (<em>string</em>) – Path to the annotated genome file used as a template.</p></li>
<li><p><strong>input</strong> (<em>string</em>) – Path to the annotated genome file used as a input.</p></li>
<li><p><strong>dir</strong> (<em>string</em>) – Path to the output directory.</p></li>
<li><p><strong>template_name</strong> (<em>string</em><em>, </em><em>optional</em>) – Name of the annotated genome file used as a template.</p></li>
<li><p><strong>input_name</strong> (<em>string</em><em>, </em><em>optional</em>) – Name of the annotated genome file used as input.</p></li>
<li><p><strong>temp_header</strong> (<em>string</em><em>, </em><em>optional</em>) – Feature qualifier of the gbff (NCBI) / faa (PROKKA) of the template to use as header for the FASTA files.
Default is None. If None is given, sets it based on file extension (currently only implememted for gbff and faa).</p></li>
<li><p><strong>in_header</strong> (<em>string</em><em>, </em><em>optional</em>) – Feature qualifier of the gbff (NCBI) / faa (PROKKA) of the input to use as header for the FASTA files.
Default is None. If None is given, sets it based on file extension (currently only implememted for gbff and faa).</p></li>
<li><p><strong>threads</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of threads to be used for DIAMOND.
Default is 2.</p></li>
<li><p><strong>extra_info</strong> (<em>list</em><em>, </em><em>optional</em>) – List of feature qualifiers to be extracted from the annotated genome files as additional information.
Default is [‘locus_tag’, ‘product’, ‘protein_id’].</p></li>
<li><p><strong>sensitivity</strong> (<em>string</em>) – Sensitivity mode for DIAMOND blastp run.
Can be sensitive, more-sensitive, very-sensitive or ultra-sensitive.
Default is ‘more-sensitive’.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.run_diamond_blastp">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">run_diamond_blastp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fasta_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensitivity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.bidirectional_blast.run_diamond_blastp" title="Permalink to this definition"></a></dt>
<dd><p>Run DIAMOND blastp for a given database name and FASTA - relies on the structure
created by this file (bidirectional_blast.py).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dir</strong> (<em>string</em>) – parent directory of the place to save the files to.</p></li>
<li><p><strong>db</strong> (<em>string</em>) – name of the genome used as the database.</p></li>
<li><p><strong>query</strong> (<em>string</em>) – name of the genome used as the query.</p></li>
<li><p><strong>fasta_path</strong> (<em>string</em>) – path of the fasta file (for the CDS)</p></li>
<li><p><strong>sensitivity</strong> (<em>string</em>) – sensitivity mode that should be used for blastp.</p></li>
<li><p><strong>threads</strong> (<em>int</em>) – number of threads should be used for running DIAMOND</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.core.generate_draft_model">
<span id="specimen-core-generate-draft-model-module"></span><h2>specimen.core.generate_draft_model module<a class="headerlink" href="#module-specimen.core.generate_draft_model" title="Permalink to this heading"></a></h2>
<p>Generate a draft model from a template.</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.check_unchanged">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">check_unchanged</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">draft</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bbh</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.generate_draft_model.check_unchanged" title="Permalink to this definition"></a></dt>
<dd><p>Check the genes names (more correctly, the IDs) for still existing original col_names.
Depending on the case, decide if to keep or remove them.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>draft</strong> (<em>cobra.Model</em>) – The draft model currently in the making.</p></li>
<li><p><strong>bbh</strong> (<em>pd.DataFrame</em>) – The table from the bidirectional_blast.py script containing the bidirectional blastp best hits information</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The edited draft model.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cobra.Model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.edit_template_identifiers">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">edit_template_identifiers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edit</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.generate_draft_model.edit_template_identifiers" title="Permalink to this definition"></a></dt>
<dd><p>Edit the subject IDs to fit the gene IDs of the template model.
Requires further extention, if needed edits are not included.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.DataFrame</em>) – the data frame containing the bidirectional blastp best hits information.</p></li>
<li><p><strong>edit</strong> (<em>string</em><em>, </em><em>currently implemented: no</em><em>, </em><em>dot-to-underscore</em>) – type of edit that is to be performed</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the edited data</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.gen_draft_model">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">gen_draft_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bbh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'default'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.generate_draft_model.gen_draft_model" title="Permalink to this definition"></a></dt>
<dd><p>Generate a draft model from a template model and the results of a bidirectional blastp (blast best hits) table
and save it as a new model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The template model.</p></li>
<li><p><strong>bbh</strong> (<em>pd.DataFrame</em>) – The bidirectional blastp best hits table.</p></li>
<li><p><strong>name</strong> (<em>string</em>) – The name of the generated model.</p></li>
<li><p><strong>dir</strong> (<em>string</em>) – Path to the directory to save the new model in.</p></li>
<li><p><strong>edit</strong> (<em>string</em><em>, </em><em>currently implemented are options no and dot-to-underscore</em>) – Allows for editing the subject_ID od the bbh table to be edited</p></li>
<li><p><strong>medium</strong> (<em>string</em>) – Name of the to be loaded from a database or ‘default’ = the one
from the template model.</p></li>
<li><p><strong>db_path</strong> (<em>string</em>) – Path to a medium database file. Optional.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.pid_filter">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">pid_filter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.generate_draft_model.pid_filter" title="Permalink to this definition"></a></dt>
<dd><p>Filter the data based on PID threshold. Entries above the given value are kept.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.DataFrame</em>) – The data from bidirectional_blast.py containing at least a ‘PID’ column.</p></li>
<li><p><strong>pid</strong> (<em>double</em><em>, </em><em>should be given in percentage e.g. 80.0.</em>) – PID threshold value.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The filtered data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.remove_absent_genes">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">remove_absent_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.generate_draft_model.remove_absent_genes" title="Permalink to this definition"></a></dt>
<dd><p>Remove a list of genes from a given model.
Note: genes that are not found in the model are skipped.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – A model to delete genes from. A copy will be created before deleting.</p></li>
<li><p><strong>genes</strong> (<em>list</em><em>, of </em><em>strings</em>) – Gene identifiers of genes that should be deleted.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A new model with the given genes deleted, if found in the original model.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cobra.Model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.rename_found_homologs">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">rename_found_homologs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">draft</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bbh</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.generate_draft_model.rename_found_homologs" title="Permalink to this definition"></a></dt>
<dd><p>Rename the genes in the model correnspondingly to the homologous ones found in the query.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>draft</strong> (<em>cobra.Model</em>) – The draft model with the to-be-renamed genes.</p></li>
<li><p><strong>bbh</strong> (<em>pd.DataFrame</em>) – The table from the bidirectional_blast.py script containing the bidirectional blastp best hits information</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The draft model with renamed genes.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cobra.Model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">template</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bpbbh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edit_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'no'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">80.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'default'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memote</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.generate_draft_model.run" title="Permalink to this definition"></a></dt>
<dd><p>Generate a draft model from a blastp best hits tsv file and a template model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>template</strong> (<em>string</em>) – Path to the file containing the template model.</p></li>
<li><p><strong>bpbbh</strong> (<em>string</em>) – Path to the blastp bidirectional best hits.</p></li>
<li><p><strong>dir</strong> (<em>string</em>) – Path to output directory.</p></li>
<li><p><strong>edit_names</strong> (<em>string</em>) – Edit the identifier of the FASTA files to match the names in the model.
Default is no. Currently implemented edits: dot-to-underscore’.</p></li>
<li><p><strong>pid</strong> (<em>float</em>) – Threshold value for determining, if a gene is counted as present or absent.
The default is 80.0 (percentage).</p></li>
<li><p><strong>name</strong> (<em>string</em><em>, </em><em>optional</em>) – Name of the output.</p></li>
<li><p><strong>medium</strong> (<em>string</em>) – Set the medium for the new model. If not set, will use the one from the template.
If given the keyword “exchanges”, will open all exchange reaction and use them as the medium.
If given together with db_path, the corresponding medium is loaded.</p></li>
<li><p><strong>db_path</strong> (<em>string</em>) – Path to a medium database file. If specified together with “medium”,
the name from medium will be searched and loaded from the database as the new medium.</p></li>
<li><p><strong>memote</strong> (<em>bool</em>) – Option to run memote after creating the draft model.
Default is False (not running memote).</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ValueError</span></code>: ‘Edit_names value {edit_names} not in list of allowed values: no, dot-to-underscore’</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ValueError</span></code>: ‘Model type {model_type} not in list of allowed values: sbml, json, matlab’</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.core.validation">
<span id="specimen-core-validation-module"></span><h2>specimen.core.validation module<a class="headerlink" href="#module-specimen.core.validation" title="Permalink to this heading"></a></h2>
<p>Validate a model (part 4 of the workflow).</p>
<p>Implemented tests in include:
- cobra/sbml check using cobrapy</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.validation.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.validation.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tests</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_all</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.validation.run" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-specimen.core">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-specimen.core" title="Permalink to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="specimen.classes.html" class="btn btn-neutral float-left" title="specimen.classes package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="specimen.core.refinement.html" class="btn btn-neutral float-right" title="specimen.core.refinement package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Carolin Brune.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>