<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>specimen.core package &mdash; SPECIMEN 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=559046d4" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
      <link rel="stylesheet" type="text/css" href="_static/custom_theme.css?v=0eca129c" />

  
    <link rel="shortcut icon" href="_static/LogoSPECIMEN.png"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=8d563738"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
        <script src="_static/clipboard.min.js?v=a7894cd8"></script>
        <script src="_static/copybutton.js?v=7859de39"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="specimen.core.refinement package" href="specimen.core.refinement.html" />
    <link rel="prev" title="specimen.classes package" href="specimen.classes.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html">
            
              <img src="_static/LogoSPECIMEN.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="about-pipeline.html">About the SPECIMEN</a></li>
<li class="toctree-l1"><a class="reference internal" href="run-pipeline.html">Run the Pipeline</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="specimen.html">Contents of SPECIMEN</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="specimen.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="specimen.classes.html">specimen.classes package</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">specimen.core package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#subpackages">Subpackages</a></li>
<li class="toctree-l4"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-specimen.core.analysis">specimen.core.analysis module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-specimen.core.bidirectional_blast">specimen.core.bidirectional_blast module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-specimen.core.generate_draft_model">specimen.core.generate_draft_model module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-specimen.core.validation">specimen.core.validation module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-specimen.core">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="specimen.util.html">specimen.util package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="specimen.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="specimen.html#module-specimen.cmd_access">specimen.cmd_access module</a></li>
<li class="toctree-l2"><a class="reference internal" href="specimen.html#module-specimen.workflow">specimen.workflow module</a></li>
<li class="toctree-l2"><a class="reference internal" href="specimen.html#module-specimen">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="dev-notes.html">Notes for Developers</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">SPECIMEN</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="specimen.html">Contents of SPECIMEN</a></li>
      <li class="breadcrumb-item active">specimen.core package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/specimen.core.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="specimen-core-package">
<h1>specimen.core package<a class="headerlink" href="#specimen-core-package" title="Link to this heading"></a></h1>
<section id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Link to this heading"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="specimen.core.refinement.html">specimen.core.refinement package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="specimen.core.refinement.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="specimen.core.refinement.html#module-specimen.core.refinement.annotation">specimen.core.refinement.annotation module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.annotation.kegg_reaction_to_kegg_pathway"><code class="docutils literal notranslate"><span class="pre">kegg_reaction_to_kegg_pathway()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.annotation.run"><code class="docutils literal notranslate"><span class="pre">run()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="specimen.core.refinement.html#module-specimen.core.refinement.cleanup">specimen.core.refinement.cleanup module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.cleanup.check_direction"><code class="docutils literal notranslate"><span class="pre">check_direction()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.cleanup.cobra_gapfill_wrapper"><code class="docutils literal notranslate"><span class="pre">cobra_gapfill_wrapper()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.cleanup.multiple_cobra_gapfill"><code class="docutils literal notranslate"><span class="pre">multiple_cobra_gapfill()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.cleanup.run"><code class="docutils literal notranslate"><span class="pre">run()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.cleanup.single_cobra_gapfill"><code class="docutils literal notranslate"><span class="pre">single_cobra_gapfill()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="specimen.core.refinement.html#module-specimen.core.refinement.extension">specimen.core.refinement.extension module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.add_gene"><code class="docutils literal notranslate"><span class="pre">add_gene()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.add_reaction"><code class="docutils literal notranslate"><span class="pre">add_reaction()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.build_metabolite_kegg"><code class="docutils literal notranslate"><span class="pre">build_metabolite_kegg()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.build_metabolite_mnx"><code class="docutils literal notranslate"><span class="pre">build_metabolite_mnx()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.extent_model"><code class="docutils literal notranslate"><span class="pre">extent_model()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.find_best_diamond_hits"><code class="docutils literal notranslate"><span class="pre">find_best_diamond_hits()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.get_metabolites_kegg"><code class="docutils literal notranslate"><span class="pre">get_metabolites_kegg()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.get_metabolites_mnx"><code class="docutils literal notranslate"><span class="pre">get_metabolites_mnx()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.get_ncbi_info"><code class="docutils literal notranslate"><span class="pre">get_ncbi_info()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.identify_missing_genes"><code class="docutils literal notranslate"><span class="pre">identify_missing_genes()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.isreaction_complete"><code class="docutils literal notranslate"><span class="pre">isreaction_complete()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.map_BiGG_reactions"><code class="docutils literal notranslate"><span class="pre">map_BiGG_reactions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.map_BiGG_reactions_row"><code class="docutils literal notranslate"><span class="pre">map_BiGG_reactions_row()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.map_to_KEGG"><code class="docutils literal notranslate"><span class="pre">map_to_KEGG()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.map_to_KEGG_row"><code class="docutils literal notranslate"><span class="pre">map_to_KEGG_row()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.map_to_NCBI_efetch_row"><code class="docutils literal notranslate"><span class="pre">map_to_NCBI_efetch_row()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.extension.run"><code class="docutils literal notranslate"><span class="pre">run()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="specimen.core.refinement.html#module-specimen.core.refinement.smoothing">specimen.core.refinement.smoothing module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.smoothing.adjust_BOF"><code class="docutils literal notranslate"><span class="pre">adjust_BOF()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.smoothing.perform_mcc"><code class="docutils literal notranslate"><span class="pre">perform_mcc()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.refinement.html#specimen.core.refinement.smoothing.run"><code class="docutils literal notranslate"><span class="pre">run()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="specimen.core.refinement.html#module-specimen.core.refinement">Module contents</a></li>
</ul>
</li>
</ul>
</div>
</section>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="module-specimen.core.analysis">
<span id="specimen-core-analysis-module"></span><h2>specimen.core.analysis module<a class="headerlink" href="#module-specimen.core.analysis" title="Link to this heading"></a></h2>
<p>Analyse a model (part 5 of the workflow).</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.analysis.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.analysis.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">media_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'BiGG'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'BiGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc_model_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc_based_on</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'id'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'id'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_aa_auxotrophies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pathway</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.analysis.run" title="Link to this definition"></a></dt>
<dd><p>SPECIMEN Step 5: Analyse the generated model.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>model_path (str): Path to the model.</p></li>
<li><p>dir (str): Path to the output directory.</p></li>
<li><dl class="simple">
<dt>media_path (str, optional): Path to a media config file. </dt><dd><p>Using this enables growth simulation.
Defaults to None.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (Literal[‘BiGG’], optional): Namespace to work on. </dt><dd><p>Defaults to ‘BiGG’.</p>
</dd>
</dl>
</li>
<li><p>pc_model_path (str, optional): Path to a core-pan model. Defaults to None.</p></li>
<li><dl class="simple">
<dt>pc_based_on (Literal[‘id’], optional): How to compare the model to the core-pan model. </dt><dd><p>Defaults to ‘id’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>test_aa_auxotrophies (bool, optional): Option to enable the amino acid</dt><dd><p>auxotrophy simulation. Defaults to True.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>pathway (bool, optional): Optional to enable KEGG pathway analysis. </dt><dd><p>Defaults to True.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.core.bidirectional_blast">
<span id="specimen-core-bidirectional-blast-module"></span><h2>specimen.core.bidirectional_blast module<a class="headerlink" href="#module-specimen.core.bidirectional_blast" title="Link to this heading"></a></h2>
<p>Perform a bidirectional blastp using DIAMOND on an input and a template (annotated genomes).</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.bdbp_diamond">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">bdbp_diamond</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">template_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">template_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensitivity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sensitive'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.bidirectional_blast.bdbp_diamond" title="Link to this definition"></a></dt>
<dd><p>Perform bidirectional blastp using DIAMOND.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>dir (str): Path to the directory parent to in/out.</p></li>
<li><p>template_name (str): Name of the template genome.</p></li>
<li><p>input_name (str): name of the input genome.</p></li>
<li><p>template_path (str): Path to the CDS FASTA-file of the template.</p></li>
<li><p>input_path (str): Path to the CDS FASTA-file of the input.</p></li>
<li><dl class="simple">
<dt>sensitivity (str, optional): Sensitivity mode for DIAMOND. </dt><dd><p>Defaults to ‘sensitive’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>threads (int, optional): Number of threads to use when running DIAMOND. </dt><dd><p>Defaults to 2.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.create_diamond_db">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">create_diamond_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.bidirectional_blast.create_diamond_db" title="Link to this definition"></a></dt>
<dd><p>Create a DIAMOND database for a given protein FASTA file.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>dir (str): Path to the data directory.</p></li>
<li><p>name (str): Name of the genome/database.</p></li>
<li><p>path (str): Path to the FASTA-file.</p></li>
<li><p>threads (int): Number of threads to use.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.extract_bestbdbp_hits">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">extract_bestbdbp_hits</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tvq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qvt</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cov</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.25</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.bidirectional_blast.extract_bestbdbp_hits" title="Link to this definition"></a></dt>
<dd><p>Extract the best directional blastp hits from two tsv files,
which were generate by bdbp_diamond() generated or similar steps.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>tvq (str): Path to the template vs. query file.</p></li>
<li><p>qvt (str): Path to the query vs. template file.</p></li>
<li><p>name (str): Name (path) of the output file.</p></li>
<li><dl class="simple">
<dt>cov (float, optional): Cut-off value for the coverage.</dt><dd><p>All hits with coverage &lt; cov will be excluded. 
Defaults to 0.25.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.extract_cds">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">extract_cds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collect_info</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">identifier</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="headerlink" href="#specimen.core.bidirectional_blast.extract_cds" title="Link to this definition"></a></dt>
<dd><p>Extract the CDS from a genbank file (annotated genome). 
Produces a FASTA-file.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>file (str): File to extract CDS from.</p></li>
<li><p>name (str): Name of the genome.</p></li>
<li><p>dir (str): Directory for the ouput.</p></li>
<li><p>collect_info (list): Feature identifiers to collect information from.</p></li>
<li><p>identifier (str): Feature identifier to use of the header of the FASTA.</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><p>str: Name of the FASTA-file</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">template</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">template_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp_header</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'protein_id'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_header</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'locus_tag'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_info</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['locus_tag',</span> <span class="pre">'product',</span> <span class="pre">'protein_id']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensitivity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'sensitive'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'more-sensitive'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'very-sensitive'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'ultra-sensitive'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'more-sensitive'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.bidirectional_blast.run" title="Link to this definition"></a></dt>
<dd><p>Run the bidirectional blast on a template and input genome (annotated).</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>template (str): Path to the annotated genome file used as a template.</p></li>
<li><p>input (str): Path to the annotated genome file used as a input.</p></li>
<li><p>dir (str): Path to the output directory.</p></li>
<li><dl class="simple">
<dt>template_name (str, optional): Name of the annotated genome file used as a template.. </dt><dd><p>Defaults to None.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>input_name (str, optional): Name of the annotated genome file used as input.. </dt><dd><p>Defaults to None.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>temp_header (str, optional): Feature qualifier of the gbff (NCBI) / faa (PROKKA) of the template to use as header for the FASTA files.</dt><dd><p>If None is given, sets it based on file extension (currently only implememted for gbff and faa).
Defaults to ‘protein_id’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>in_header (str, optional): Feature qualifier of the gbff (NCBI) / faa (PROKKA) of the input to use as header for the FASTA files. </dt><dd><p>If None is given, sets it based on file extension (currently only implememted for gbff and faa).
Defaults to ‘locus_tag’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>threads (int, optional):  Number of threads to be used for DIAMOND. </dt><dd><p>Defaults to 2.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>extra_info (list[str], optional):  List of feature qualifiers to be extracted from the annotated genome files as additional information. </dt><dd><p>Defaults to [‘locus_tag’, ‘product’, ‘protein_id’].</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>sensitivity (Literal[‘sensitive’, ‘more-sensitive’, ‘very-sensitive’, ‘ultra-sensitive’], optional): </dt><dd><p>Sensitivity mode for DIAMOND blastp run.. 
Defaults to ‘more-sensitive’.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Raises:</dt><dd><p>ValueError: Unknown file extension. Please set value for temp_header manually or check file.
ValueError: Unknown file extension. Please set value for in_header manually or check file.
ValueError: Unknown sensitive mode</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.run_diamond_blastp">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">run_diamond_blastp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fasta_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensitivity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.bidirectional_blast.run_diamond_blastp" title="Link to this definition"></a></dt>
<dd><p>Run DIAMOND blastp for a given database name and FASTA - relies on the structure
created by this file (bidirectional_blast.py).</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>dir (str): Parent directory of the place to save the files to.</p></li>
<li><p>db (str): Name of the genome/database used as the database.</p></li>
<li><p>query (str): Name of the genome used as the query.</p></li>
<li><p>fasta_path (str): Path to the FASTA-file containing the CDS.</p></li>
<li><p>sensitivity (str): Sensitivity mode to use for DIAMOND blastp.</p></li>
<li><p>threads (int): Number of threads that will be used for running DIAMOND</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.core.generate_draft_model">
<span id="specimen-core-generate-draft-model-module"></span><h2>specimen.core.generate_draft_model module<a class="headerlink" href="#module-specimen.core.generate_draft_model" title="Link to this heading"></a></h2>
<p>Generate a draft model from a template.</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.check_unchanged">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">check_unchanged</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">draft</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bbh</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Model</span></span></span><a class="headerlink" href="#specimen.core.generate_draft_model.check_unchanged" title="Link to this definition"></a></dt>
<dd><p>Check the genes names (more correctly, the IDs) for still existing original col_names.
Depending on the case, decide if to keep or remove them.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>draft (cobra.Model): The draft model currently in the making.</p></li>
<li><p>bbh (pd.DataFrame): The table from the bidirectional_blast.py script containing the bidirectional blastp best hits information.</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><p>cobra.Model: The model after the check and possible removal of genes.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.edit_template_identifiers">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">edit_template_identifiers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'no'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'dot-to-underscore'</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="headerlink" href="#specimen.core.generate_draft_model.edit_template_identifiers" title="Link to this definition"></a></dt>
<dd><p>Edit the subject IDs to fit the gene IDs of the template model.
Requires further extention, if needed edits are not included.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>data (pd.DataFrame): The data frame containing the bidirectional blastp best hits information.</p></li>
<li><dl class="simple">
<dt>edit (Literal[‘no’,’dot-to-underscore’]): Type of edit to perform.</dt><dd><p>Currently possible options: no, dot-to-underscore.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><p>pd.DataFrame: The (un)edited DataFrame.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.gen_draft_model">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">gen_draft_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bbh</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'no'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'dot-to-underscore'</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'default'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'BiGG'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'BiGG'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Model</span></span></span><a class="headerlink" href="#specimen.core.generate_draft_model.gen_draft_model" title="Link to this definition"></a></dt>
<dd><p>Generate a draft model from a template model and the results of a bidirectional blastp (blast best hits) table
and save it as a new model.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>model (cobra.Model): The template model.</p></li>
<li><p>bbh (pd.DataFrame): The bidirectional blastp best hits table.</p></li>
<li><p>name (str): Name of the newly generated model.</p></li>
<li><p>dir (str): Path to the directory to save the new model in.</p></li>
<li><dl class="simple">
<dt>edit (Literal[‘no’,’dot-to-underscore’):  Type of edit to perform.</dt><dd><p>Currently possible options: no, dot-to-underscore.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>medium (str, optional):  Name of the to be loaded from the refineGEMs database or ‘default’ = the one</dt><dd><p>from the template model. If given the keyword ‘exchanges’, will use all exchange reactions in the model as a medium.
Defaults to ‘default’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (Literal[‘BiGG’], optional): Namespace of the model. </dt><dd><p>Defaults to ‘BiGG’.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><p>cobra.Model: The generated draft model.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.pid_filter">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">pid_filter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="headerlink" href="#specimen.core.generate_draft_model.pid_filter" title="Link to this definition"></a></dt>
<dd><p>Filter the data based on PID threshold. Entries above the given value are kept.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>data (pd.DataFrame): The data from bidirectional_blast.py containing at least a ‘PID’ column.</p></li>
<li><p>pid (float): PID threshold value, given in percentage e.g. 80.0.</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><p>pd.DataFrame: The filtered data.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.remove_absent_genes">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">remove_absent_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Model</span></span></span><a class="headerlink" href="#specimen.core.generate_draft_model.remove_absent_genes" title="Link to this definition"></a></dt>
<dd><p>Remove a list of genes from a given model.
Note: genes that are not found in the model are skipped.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>model (cobra.Model): A template model to delete genes from. </dt><dd><p>A copy will be created before deleting.</p>
</dd>
</dl>
</li>
<li><p>genes (list[str]): Gene identifiers of genes that should be deleted.</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><p>cobra.Model: A new model with the given genes deleted, if found in the original model.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.rename_found_homologs">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">rename_found_homologs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">draft</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bbh</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Model</span></span></span><a class="headerlink" href="#specimen.core.generate_draft_model.rename_found_homologs" title="Link to this definition"></a></dt>
<dd><p>Rename the genes in the model correnspondingly to the homologous ones found in the query.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>draft (cobra.Model): The draft model with the to-be-renamed genes.</p></li>
<li><p>bbh (pd.DataFrame): The table from the bidirectional_blast.py script containing the bidirectional blastp best hits information</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><p>cobra.Model: The draft model with renamed genes.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">template</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bpbbh</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edit_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'no'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'dot-to-underscore'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'no'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">80.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'default'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'BiGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memote</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.generate_draft_model.run" title="Link to this definition"></a></dt>
<dd><p>Generate a draft model from a blastp best hits tsv file and a template model.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>template (str): Path to the file containing the template model.</p></li>
<li><p>bpbbh (str): Path to the blastp bidirectional best hits.</p></li>
<li><p>dir (str): Path to output directory.</p></li>
<li><dl class="simple">
<dt>edit_names (Literal[‘no’,’dot-to-underscore’, optional):  Type of edit to perform.</dt><dd><p>Currently possible options: no, dot-to-underscore. 
Defaults to ‘no’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>pid (float, optional): Threshold value for determining, if a gene is counted as present or absent. </dt><dd><p>Given in percentage, e.g. 80.0 = 80%.
Defaults to 80.0.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>name (Union[str,None], optional): Name of the output model. </dt><dd><p>If not given, takes name from filename. 
Defaults to None.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>medium (str, optional):  Name of the to be loaded from the refineGEMs database or ‘default’ = the one</dt><dd><p>from the template model. If given the keyword ‘exchanges’, will use all exchange reactions in the model as a medium.
Defaults to ‘default’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (str, optional): Namespace of the model. </dt><dd><p>Defaults to ‘BiGG’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>memote (bool, optional): Option to run memote after creating the draft model. </dt><dd><p>Defaults to False.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Raises:</dt><dd><p>ValueError: ‘Edit_names value {edit_names} not in list of allowed values: no, dot-to-underscore’</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.core.validation">
<span id="specimen-core-validation-module"></span><h2>specimen.core.validation module<a class="headerlink" href="#module-specimen.core.validation" title="Link to this heading"></a></h2>
<p>Validate a model (part 4 of the workflow).</p>
<p>Implemented tests in include:
- cobra/sbml check using cobrapy</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.validation.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.validation.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tests</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'cobra'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_all</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.validation.run" title="Link to this definition"></a></dt>
<dd><p>SPECIMEN Step 4: Validate the model.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>dir (str): Path to the output directory.</p></li>
<li><p>model_path (str): Path to the model to be validated</p></li>
<li><dl class="simple">
<dt>tests (None | Literal[‘cobra’], optional): Tests to perform. </dt><dd><p>Defaults to None.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>run_all (bool, optional): Run al available tests. If True, overwrites</dt><dd><p>the previous parameter.
Defaults to True.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.core">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-specimen.core" title="Link to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="specimen.classes.html" class="btn btn-neutral float-left" title="specimen.classes package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="specimen.core.refinement.html" class="btn btn-neutral float-right" title="specimen.core.refinement package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Carolin Brune.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>