<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>specimen.core.refinement package &mdash; SPECIMEN 1.0.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="specimen.util package" href="specimen.util.html" />
    <link rel="prev" title="specimen.core package" href="specimen.core.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            SPECIMEN
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="about-pipeline.html">About the SPECIMEN</a></li>
<li class="toctree-l1"><a class="reference internal" href="run-pipeline.html">Run the Pipeline</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="specimen.html">Contents of SPECIMEN</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="specimen.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="specimen.classes.html">specimen.classes package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="specimen.core.html">specimen.core package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="specimen.core.html#subpackages">Subpackages</a></li>
<li class="toctree-l4"><a class="reference internal" href="specimen.core.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="specimen.core.html#module-specimen.core.analysis">specimen.core.analysis module</a></li>
<li class="toctree-l4"><a class="reference internal" href="specimen.core.html#module-specimen.core.bidirectional_blast">specimen.core.bidirectional_blast module</a></li>
<li class="toctree-l4"><a class="reference internal" href="specimen.core.html#module-specimen.core.generate_draft_model">specimen.core.generate_draft_model module</a></li>
<li class="toctree-l4"><a class="reference internal" href="specimen.core.html#module-specimen.core.validation">specimen.core.validation module</a></li>
<li class="toctree-l4"><a class="reference internal" href="specimen.core.html#module-specimen.core">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="specimen.util.html">specimen.util package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="specimen.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="specimen.html#module-specimen.cmd_access">specimen.cmd_access module</a></li>
<li class="toctree-l2"><a class="reference internal" href="specimen.html#module-specimen.workflow">specimen.workflow module</a></li>
<li class="toctree-l2"><a class="reference internal" href="specimen.html#module-specimen">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="dev-notes.html">Notes for Developers</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">SPECIMEN</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="specimen.html">Contents of SPECIMEN</a></li>
          <li class="breadcrumb-item"><a href="specimen.core.html">specimen.core package</a></li>
      <li class="breadcrumb-item active">specimen.core.refinement package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/specimen.core.refinement.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="specimen-core-refinement-package">
<h1>specimen.core.refinement package<a class="headerlink" href="#specimen-core-refinement-package" title="Permalink to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading"></a></h2>
</section>
<section id="module-specimen.core.refinement.annotation">
<span id="specimen-core-refinement-annotation-module"></span><h2>specimen.core.refinement.annotation module<a class="headerlink" href="#module-specimen.core.refinement.annotation" title="Permalink to this heading"></a></h2>
<p>Further annotate a model.</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.annotation.kegg_reaction_to_kegg_pathway">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.annotation.</span></span><span class="sig-name descname"><span class="pre">kegg_reaction_to_kegg_pathway</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">viaEC</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">viaRC</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.annotation.kegg_reaction_to_kegg_pathway" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve the KEGG pathways for existing KEGG reaction IDs, if
they have yet to be added. Depending on the given options, only the
reactions is searched or additional searches are started using the
EC number and reactions class if the first search was unsuccesful.
NOTE: EC number and reaction class cover a broader set of pathways and might
not be entirely accurate for the given reaction.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The model to be annotated with KEGG pathway.</p></li>
<li><p><strong>viaEC</strong> (<em>bool</em>) – Option to search for KEGG pathway ID using the EC number if
previous searches were unsuccesful. Default is False.</p></li>
<li><p><strong>viaRC</strong> – Option to search for KEGG pathway ID using the reaction class if
previous searches were unsuccesful. Default is False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.annotation.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.annotation.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kegg_viaEC</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kegg_viaRC</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memote</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.annotation.run" title="Permalink to this definition"></a></dt>
<dd><p>Further annotate a given model.</p>
<p>Currently add annotations for:
- SBO using SBOannotator</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>string</em>) – Path to the model (sbml)</p></li>
<li><p><strong>dir</strong> (<em>string</em>) – Path to the output directory.</p></li>
<li><p><strong>kegg_viaEC</strong> (<em>bool</em>) – Option to search for KEGG pathway ID using the EC number if
previous searches were unsuccesful. Default is False.</p></li>
<li><p><strong>kegg_viaRC</strong> – Option to search for KEGG pathway ID using the reaction class if
previous searches were unsuccesful. Default is False.</p></li>
<li><p><strong>memote</strong> (<em>bool</em><em>, </em><em>optional</em>) – Option to run memote after the annotation on the model.
Default is False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.core.refinement.cleanup">
<span id="specimen-core-refinement-cleanup-module"></span><h2>specimen.core.refinement.cleanup module<a class="headerlink" href="#module-specimen.core.refinement.cleanup" title="Permalink to this heading"></a></h2>
<p>Perform part 3 of the refinement: clean-up.</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.cleanup.check_direction">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.cleanup.</span></span><span class="sig-name descname"><span class="pre">check_direction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.cleanup.check_direction" title="Permalink to this definition"></a></dt>
<dd><p>Check the direction of newly created reactions (01-extention) by searching for matching MetaCyc,
KEGG and MetaNetX IDs as well as EC number in a downloaded BioCyc (MetaCyc)
database table (need to contain at least the following columns:
Reactions (MetaCyc ID),EC-Number,KEGG reaction,METANETX,Reaction-Direction.</p>
<p>note: checks only creations that do not contain the notes[‘creation’] == ‘via template’</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The GEM containing the reactions to be check for direction.</p></li>
<li><p><strong>data_file</strong> (<em>string</em>) – Path to the MetabCyc (BioCyc) smart table.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The updated model.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cobra.Model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.cleanup.cobra_gapfill_wrapper">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.cleanup.</span></span><span class="sig-name descname"><span class="pre">cobra_gapfill_wrapper</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">universal</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunk_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">growth_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.cleanup.cobra_gapfill_wrapper" title="Permalink to this definition"></a></dt>
<dd><p>Wrapper for single_cobra_gapfill().</p>
<p>Either use the full set of reactions in universal model by setting iteration to
0 or None or use them in randomized chunks for faster runtime (especially useful
on laptops). Note: when using the second option, be aware that this does not
test all reaction combinations exhaustively (heuristic approach!!!).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The model to perform gapfilling on.</p></li>
<li><p><strong>universal</strong> (<em>cobra.model</em>) – A model with reactions to be potentially used for the gapfilling.</p></li>
<li><p><strong>medium</strong> (<a class="reference internal" href="specimen.classes.html#specimen.classes.medium.Medium" title="specimen.classes.medium.Medium"><em>medium.Medium</em></a>) – A medium the model should grow on.</p></li>
<li><p><strong>iterations</strong> (<em>int</em>) – Number of iterations for the heuristic version.
Default is 3, is 0 or None is given, uses full set of reactions.</p></li>
<li><p><strong>chunk_size</strong> (<em>int</em>) – Number of reactions to be used for gapfilling at the same time.
Default is 10000. If None or 0 is given, use full set, not heuristic.</p></li>
<li><p><strong>growth_threshold</strong> – Minimal rate for the model to be considered growing.</p></li>
</ul>
</dd>
<dt class="field-even">Type<span class="colon">:</span></dt>
<dd class="field-even"><p>float</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The updated (gapfilled) model.</p>
</dd>
<dt class="field-even">Rytpe<span class="colon">:</span></dt>
<dd class="field-even"><p>cobra.Model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.cleanup.complete_BioMetaCyc">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.cleanup.</span></span><span class="sig-name descname"><span class="pre">complete_BioMetaCyc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.cleanup.complete_BioMetaCyc" title="Permalink to this definition"></a></dt>
<dd><p>Check for existing MetaCyc and BioCyc annotations for metabolites and
reactions and generate them from the other if one of the two is missing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>model</strong> (<em>cobra.Model</em>) – A genome-scale model to be checked for complete BioCyc/MetaCyc annotations.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The updated model.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cobra.Model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.cleanup.multiple_cobra_gapfill">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.cleanup.</span></span><span class="sig-name descname"><span class="pre">multiple_cobra_gapfill</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">universal</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">media_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">growth_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunk_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.cleanup.multiple_cobra_gapfill" title="Permalink to this definition"></a></dt>
<dd><p>Perform single_cobra_gapfill() on a list of models.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The model to perform gapfilling on.</p></li>
<li><p><strong>universal</strong> (<em>cobra.model</em>) – A model with reactions to be potentially used for the gapfilling.</p></li>
<li><p><strong>media</strong> (<em>list</em><em> of </em><a class="reference internal" href="specimen.classes.html#specimen.classes.medium.Medium" title="specimen.classes.medium.Medium"><em>medium.Medium</em></a>) – A list of media the model should grow on.</p></li>
<li><p><strong>growth_threshold</strong> – Minimal rate for the model to be considered growing.</p></li>
<li><p><strong>iterations</strong> (<em>int</em>) – Number of iterations for the heuristic version.
Default is 3, is 0 or None is given, uses full set of reactions.</p></li>
<li><p><strong>chunk_size</strong> (<em>int</em>) – Number of reactions to be used for gapfilling at the same time.
Default is 10000. If None or 0 is given, use full set, not heuristic.</p></li>
</ul>
</dd>
<dt class="field-even">Type<span class="colon">:</span></dt>
<dd class="field-even"><p>float</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The gapfilled model, of possible.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>cobra.model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.cleanup.resolve_duplicate_metabolites">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.cleanup.</span></span><span class="sig-name descname"><span class="pre">resolve_duplicate_metabolites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">based_on</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'metanetx.chemical'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.cleanup.resolve_duplicate_metabolites" title="Permalink to this definition"></a></dt>
<dd><p>Resolve duplicate metabolites in a model. Metabolites are considered
duplicate if they share the same annotations (same or nan).
Note: Depending on the starting database, the results might differ.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The model to be checked for duplicate metabolites.</p></li>
<li><p><strong>based_on</strong> – Starting point of the comparison.</p></li>
<li><p><strong>replace</strong> (<em>bool</em><em>, </em><em>default is True</em>) – Option to replace the duplicate metabolites or only report them.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The updated model.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cobra.Model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.cleanup.resolve_duplicate_reactions">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.cleanup.</span></span><span class="sig-name descname"><span class="pre">resolve_duplicate_reactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">based_on</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reaction'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_reac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.cleanup.resolve_duplicate_reactions" title="Permalink to this definition"></a></dt>
<dd><p>Resolve and remove duplicate reaction based on their reaction equation
and matching database identifiers. Only if all match or are nan will one of
the reactions be removed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The model to be checked for duplicate reactions.</p></li>
<li><p><strong>based_on</strong> (<em>string</em>) – Starting point of the comparison for duplicates. Default is ‘reaction’. Currently not advisable to change this parameter.</p></li>
<li><p><strong>remove_reac</strong> (<em>bool</em><em>, </em><em>default True.</em>) – Option to keep or remove found duplicates.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The updated model.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cobra.Model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.cleanup.resolve_duplicates">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.cleanup.</span></span><span class="sig-name descname"><span class="pre">resolve_duplicates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_reac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_meta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'default'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace_dupl_meta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_unused_meta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_dupl_reac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.cleanup.resolve_duplicates" title="Permalink to this definition"></a></dt>
<dd><p>Resolve and remove (optional) duplicate metabolites and reactions in the model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The model to be checked for duplicates.</p></li>
<li><p><strong>check_reac</strong> (<em>bool</em><em>, </em><em>default is True</em><em> (</em><em>check reactions</em><em>)</em><em>.</em>) – Option to check for duplicate reactions.</p></li>
<li><p><strong>check_meta</strong> (<em>bool</em><em>, </em><em>default is True</em><em> (</em><em>check metabolites</em><em>)</em><em>.</em>) – Option to check for duplicate metabolites.</p></li>
<li><p><strong>replace_dupl_meta</strong> (<em>bool</em><em>, </em><em>default is True</em><em> (</em><em>Replace metabolites with duplicate</em><em>)</em><em>.</em>) – Option to replace duplicate metabolites.</p></li>
<li><p><strong>remove_unused_meta</strong> (<em>bool</em><em>, </em><em>default is False</em><em> (</em><em>remove unused metabolites</em><em>)</em><em>.</em>) – Option to remove unused metabolites (metabolites that are not taking part in any reaction).</p></li>
<li><p><strong>remove_dupl_reac</strong> (<em>bool</em><em>, </em><em>default is True</em><em> (</em><em>remove duplicate reactions</em><em>)</em><em>.</em>) – Option to remove duplicate reactions.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The updated model.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cobra.Model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.cleanup.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.cleanup.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biocyc_db</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_dupl_reac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_dupl_meta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'default'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_unused_meta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_dupl_reac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_dupl_meta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">universal</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">media_db</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_media</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">external_media</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">change_to_aerobic</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">change_to_anaerobic</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_casamino</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">growth_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunk_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memote</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.cleanup.run" title="Permalink to this definition"></a></dt>
<dd><p>Perform the second refinement step, cleanup, on a model.</p>
<p>The second refinement step resolves the following issues:
- (optional) checking direction of reactions with BioCyc
- complete BioCyc/MetaCyc annotation inconsistencies
- find and/or resolve duplicates (reactions and metabolites)
- gapfilling using cobra</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>string</em>) – The Path to an sbml model.</p></li>
<li><p><strong>dir</strong> (<em>string</em>) – Path to the directory of the output.</p></li>
<li><p><strong>biocyc_db</strong> (<em>string</em><em>, </em><em>optional</em>) – Path to the BioCyc/MetaCyc reaction database file.
Default is None, which leads to skipping the direction check.</p></li>
<li><p><strong>check_dupl_reac</strong> (<em>bool</em>) – Option to check for duplicate reactions.
Default is False</p></li>
<li><p><strong>check_dupl_meta</strong> (<em>bool</em>) – Option to check for duplicate metabolites.
Default is ‘default’ (check based on MetaNetX first).
Further options include ‘skip’ and ‘exhaustive’ (check for all possibilities).</p></li>
<li><p><strong>remove_unused_meta</strong> (<em>bool</em>) – Option to remove unused metabolites from the model.
Default is False.</p></li>
<li><p><strong>remove_dupl_reac</strong> (<em>bool</em>) – Option to remove the duplicate reactions.
Deafult is False. True is only applied, if check_dupl_reac is also True.</p></li>
<li><p><strong>remove_dupl_meta</strong> (<em>bool</em>) – Option to remove the duplicate metabolites.
Deafult is False. True is only applied, if check_dupl_meta is also True.</p></li>
</ul>
</dd>
</dl>
<p>&#64;TODO missing docstrings</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>iterations</strong> (<em>int</em>) – Number of iterations for the heuristic version.
Default is 3, is 0 or None is given, uses full set of reactions.</p></li>
<li><p><strong>chunk_size</strong> (<em>int</em>) – Number of reactions to be used for gapfilling at the same time.
Default is 10000. If None or 0 is given, use full set, not heuristic.</p></li>
<li><p><strong>memote</strong> (<em>bool</em>) – Option to run memote on the cleaned model.
Default is False. Requires Memote to be installed and runnable from the command line.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.cleanup.single_cobra_gapfill">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.cleanup.</span></span><span class="sig-name descname"><span class="pre">single_cobra_gapfill</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">universal</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">growth_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.cleanup.single_cobra_gapfill" title="Permalink to this definition"></a></dt>
<dd><p>Attempt gapfilling (with cobra) for a given model to allow growth on a given
medium.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The model to perform gapfilling on.</p></li>
<li><p><strong>universal</strong> (<em>cobra.model</em>) – A model with reactions to be potentially used for the gapfilling.</p></li>
<li><p><strong>medium</strong> (<a class="reference internal" href="specimen.classes.html#specimen.classes.medium.Medium" title="specimen.classes.medium.Medium"><em>medium.Medium</em></a>) – A medium the model should grow on.</p></li>
<li><p><strong>growth_threshold</strong> – Minimal rate for the model to be considered growing.</p></li>
</ul>
</dd>
<dt class="field-even">Type<span class="colon">:</span></dt>
<dd class="field-even"><p>float</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of reactions to be added to the model to allow growth
or True, if the model already grows.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list or bool</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.core.refinement.extension">
<span id="specimen-core-refinement-extension-module"></span><h2>specimen.core.refinement.extension module<a class="headerlink" href="#module-specimen.core.refinement.extension" title="Permalink to this heading"></a></h2>
<p>Part one of the third step of the pipeline: refinement - extension.</p>
<p>Extents the model by mapping missing genes to reactions via NCBI, KEGG,
MetaNetX and BiGG.</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.add_gene">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">add_gene</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reaction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">first</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.add_gene" title="Permalink to this definition"></a></dt>
<dd><p>Add a new gene to a genome-scale metabolic cobra model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The model.</p></li>
<li><p><strong>reaction</strong> (<em>string</em>) – The reaction id to add the gene to.</p></li>
<li><p><strong>row</strong> (<em>pd.Series</em>) – A single row of the output table of map_BiGG_reactions().</p></li>
<li><p><strong>first</strong> (<em>bool</em><em>, </em><em>true if gene is first to be added.</em>) – Shows, if gene is the first gene to be added to the reaction.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The updated model.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cobra.Model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.add_reaction">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">add_reaction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reac_xref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reac_prop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chem_xref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chem_prop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigg_metabolites</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_dna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_rna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.add_reaction" title="Permalink to this definition"></a></dt>
<dd><p>Add a reaction to a given genome-scale metabolic model.
If possible, the reaction is added via MetaNetX way. If the reaction cannot
be found within the MetaNetX namespace, it will be added via KEGG.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – A GEM.</p></li>
<li><p><strong>row</strong> (<em>pd.Series</em>) – A single row of the output table of map_BiGG_reactions().</p></li>
<li><p><strong>reac_xref</strong> (<em>pd.DataFrame</em>) – The chem_xref table from MetaNetX</p></li>
<li><p><strong>reac_prop</strong> (<em>pd.DataFrame</em>) – The chem_prop table from MetaNetX</p></li>
<li><p><strong>chem_xref</strong> (<em>pd.DataFrame</em>) – The chem_xref table from MetaNetX</p></li>
<li><p><strong>chem_prop</strong> (<em>pd.DataFrame</em>) – The chem_prop table from MetaNetX</p></li>
<li><p><strong>bigg_metabolites</strong> (<em>pd.DataFrame</em>) – The BiGG compound namespace table.</p></li>
<li><p><strong>exclude_dna</strong> (<em>bool</em><em>, </em><em>default is True.</em>) – Tag to include or exclude DNA reactions.</p></li>
<li><p><strong>exclude_rna</strong> (<em>bool</em><em>, </em><em>default is True.</em>) – Tag to include or exclude RNA reactions.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The updated model.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cobra.Model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.build_metabolite_kegg">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">build_metabolite_kegg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kegg_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_metabolites</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_kegg_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigg_metabolites</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.build_metabolite_kegg" title="Permalink to this definition"></a></dt>
<dd><p>Create or retrieve (from model) a metabolite based on its KEGG ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kegg_id</strong> (<em>string</em>) – The KEGG.compound ID of the metabolite in question.</p></li>
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The model.</p></li>
<li><p><strong>model_metabolites</strong> (<em>list</em>) – List of the formulas of all metabolites in the model.</p></li>
<li><p><strong>model_kegg_ids</strong> (<em>list</em>) – List of all annotated KEGG Compound IDs in the model.</p></li>
<li><p><strong>bigg_metabolites</strong> (<em>pd.DataFrame</em>) – The BiGG compound namespace table.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The retrieved or newly build metabolite.</p>
</dd>
<dt class="field-odd">Rytpe<span class="colon">:</span></dt>
<dd class="field-odd"><p>cobra.Metabolite</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.build_metabolite_mnx">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">build_metabolite_mnx</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metabolite</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_metabolites</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnx_chem_prop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnx_chem_xref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigg_metabolites</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.build_metabolite_mnx" title="Permalink to this definition"></a></dt>
<dd><p>Create or retrieve (from model) a metabolite based on its MetaNetX ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>metabolite</strong> (<em>string</em>) – The MetaNetX ID of the metabolite.</p></li>
<li><p><strong>model</strong> (<em>cobra.model</em>) – The underlying genome-scale metabolic model.</p></li>
<li><p><strong>model_metabolites</strong> (<em>list</em>) – List of the formulas of all metabolites in the model.</p></li>
<li><p><strong>mnx_chem_xref</strong> (<em>pd.DataFrame</em>) – The chem_xref table from MetaNetX</p></li>
<li><p><strong>mnx_chem_prop</strong> (<em>pd.DataFrame</em>) – The chem_prop table from MetaNetX</p></li>
<li><p><strong>bigg_metabolites</strong> (<em>pd.DataFrame</em>) – The BiGG compound namespace table.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The retrieved or newly build metabolite.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cobra.Metabolite</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.create_reac_id">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">create_reac_id</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.create_reac_id" title="Permalink to this definition"></a></dt>
<dd><p>From a given reaction name, create an ID for a model.</p>
<p>&#64;TODO (if needed):</p>
<ul class="simple">
<li><p>check, if created name already exists</p></li>
<li><p>find a more sensible way to create an ID</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>name</strong> (<em>string</em>) – The name of the reaction.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The generated ID</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>string</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.extent_model">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">extent_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chem_prop_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chem_xref_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reac_prop_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reac_xref_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigg_metabolites_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_dna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_rna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.extent_model" title="Permalink to this definition"></a></dt>
<dd><p>Add reactions, metabolites and genes to a model based on the output of map_to_bigg().</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>table</strong> (<em>pd.DataFrame</em><em>, </em><em>output</em><em> of </em><em>map_to_bigg</em>) – The table with the information to be added to the model.</p></li>
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The genome-scale metabolic model to be extended</p></li>
<li><p><strong>chem_prop_file</strong> (<em>string</em>) – Path to the MetaNetX chem_prop file.</p></li>
<li><p><strong>chem_xref_file</strong> (<em>string</em>) – Path to the MetaNetX chem_xref file.</p></li>
<li><p><strong>reac_prop_file</strong> (<em>string</em>) – Path to the MetaNetX reac_prop file.</p></li>
<li><p><strong>reac_xref_file</strong> (<em>string</em>) – Path to the MetaNetX reac_xref file.</p></li>
<li><p><strong>bigg_metabolites_file</strong> (<em>string</em>) – Path to the BiGG Metabolites namespace file (rewritten).</p></li>
<li><p><strong>exclude_dna</strong> (<em>bool</em><em>, </em><em>default is True.</em>) – Tag to include or exclude DNA reactions.</p></li>
<li><p><strong>exclude_rna</strong> (<em>bool</em><em>, </em><em>default is True.</em>) – Tag to include or exclude RNA reactions.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The extended model.</p>
</dd>
<dt class="field-odd">Rytpe<span class="colon">:</span></dt>
<dd class="field-odd"><p>cobra.Model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.find_best_diamond_hits">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">find_best_diamond_hits</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pid</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.find_best_diamond_hits" title="Permalink to this definition"></a></dt>
<dd><p>Identify the best hits from the DIAMOND run.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file</strong> (<em>string</em>) – File name (path) of the “missing_genes” file.</p></li>
<li><p><strong>pid</strong> (<em>float</em><em>, </em><em>should be between 0.0 and 100.0</em><em> (</em><em>given in percentage</em><em>)</em>) – Threshold value for the PID .</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Table of the locus tags of the query and the NCBI accession version numbers of the RefSeqs.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.get_metabolites_kegg">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">get_metabolites_kegg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">equation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chem_xref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chem_prop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigg_metabolites</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.get_metabolites_kegg" title="Permalink to this definition"></a></dt>
<dd><p>Based on a given KEGG equation and a model, get or
create metabolite entires in/for the model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – A GEM.</p></li>
<li><p><strong>equation</strong> (<em>string</em>) – The equation from KEGG</p></li>
<li><p><strong>chem_xref</strong> (<em>pd.DataFrame</em>) – The chem_xref table from MetaNetX</p></li>
<li><p><strong>chem_prop</strong> (<em>pd.DataFrame</em>) – The chem_prop table from MetaNetX</p></li>
<li><p><strong>bigg_metabolites</strong> (<em>pd.DataFrame</em>) – The BiGG compound namespace table.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictonary of metabolites and stoichiometric factors.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.get_metabolites_mnx">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">get_metabolites_mnx</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">equation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnx_chem_xref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnx_chem_prop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigg_metabolites</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.get_metabolites_mnx" title="Permalink to this definition"></a></dt>
<dd><p>Based on a given MetaNetX equation and a model, get or
create metabolite entires in/for the model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – A GEM.</p></li>
<li><p><strong>equation</strong> (<em>string</em>) – The equation from MetaNetX</p></li>
<li><p><strong>mnx_chem_xref</strong> (<em>pd.DataFrame</em>) – The chem_xref table from MetaNetX</p></li>
<li><p><strong>mnx_chem_prop</strong> (<em>pd.DataFrame</em>) – The chem_prop table from MetaNetX</p></li>
<li><p><strong>bigg_metabolites</strong> (<em>pd.DataFrame</em>) – The BiGG compound namespace table.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictonary of metabolites and stoichiometric factors.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.get_ncbi_info">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">get_ncbi_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncbi_map_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.get_ncbi_info" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Retrieve information from NCBI via mapping to precomputed files.</dt><dd><p>If no mapping exists, try via Entrez - implemented, but does not work well
-&gt; problems with connecting successfully to NCBI.
&#64;TODO if time, rewrite using the Bioppython package, see if it works better</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>table</strong> (<em>pd.DataFrame</em><em>, </em><em>must have the column ncbi_accession_version</em>) – A table containing the output of “find_best_diamond_hits”.</p></li>
<li><p><strong>ncbi_map_file</strong> (<em>string</em>) – A csv file containing the ncbi mapping.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A table with the added information.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.identify_missing_genes">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">identify_missing_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fasta_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.identify_missing_genes" title="Permalink to this definition"></a></dt>
<dd><p>identify missing genes in a model based on the draft model and a list of genes.
Save the missing genes’ sequences in a new FASTA file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_path</strong> (<em>string</em><em>, </em><em>csv file should have a header</em>) – Path to the csv file containing the list of genes.</p></li>
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The draft model.</p></li>
<li><p><strong>id</strong> (<em>string</em>) – Name of the column in gene_path, that equal the ids in the model.</p></li>
<li><p><strong>fasta_path</strong> (<em>string</em>) – Path to the FASTA file containing all genes for the genome.</p></li>
<li><p><strong>dir</strong> (<em>string</em>) – Path to the directory to save the new FASTA in.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A table with the missing genes information.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.isreaction_complete">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">isreaction_complete</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_dna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_rna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.isreaction_complete" title="Permalink to this definition"></a></dt>
<dd><p>Check, if a reaction is complete and ready to be added to the model.
Additionally, it is possible to check for DNA and RNA reations
and set them to be excluded or included.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>reac</strong> (<em>cobra.Reaction</em>) – The reaction to be checked.</p></li>
<li><p><strong>exclude_dna</strong> (<em>bool</em><em>, </em><em>default is True.</em>) – Tag to include or exclude DNA reactions.</p></li>
<li><p><strong>exclude_rna</strong> (<em>bool</em><em>, </em><em>default is True.</em>) – Tag to include or exclude RNA reactions.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the check is successful, else false.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.make_reaction_dict">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">make_reaction_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.make_reaction_dict" title="Permalink to this definition"></a></dt>
<dd><p>For a given model, create a dictionary of reaction IDs and their
KEGG IDs, if annotated.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>model</strong> (<em>cobra.Model</em>) – A genome-scale metabolic model.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The reaction.id - kegg.reaction.id mapping.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.map_BiGG_reactions">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">map_BiGG_reactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.map_BiGG_reactions" title="Permalink to this definition"></a></dt>
<dd><p>Map the output of map_to_KEGG() to a BiGG namespace file (rewritten-type, see auxilliaries).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>table_file</strong> (<em>string</em>) – The path to the saved table from running map_to_KEGG().</p></li>
<li><p><strong>file</strong> (<em>string</em>) – The path to the BiGG namespace (reactions!) file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The table with an additional column for the mapping to BiGG reactions.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.map_Bigg_reactions_row">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">map_Bigg_reactions_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.map_Bigg_reactions_row" title="Permalink to this definition"></a></dt>
<dd><p>Map a single entry from the table in map_BiGG_reactions() to the BiGG reaction namespace.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>row</strong> (<em>pd.Series</em>) – A single row of the table.</p></li>
<li><p><strong>namespace</strong> (<em>pd.DataFrame</em>) – The BiGG reaction namespace table.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The edited row.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.Series</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.map_to_KEGG">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">map_to_KEGG</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">working_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">manual_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_summary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.map_to_KEGG" title="Permalink to this definition"></a></dt>
<dd><p>Map entries of the table from get_ncbi_info() to KEGG.
Get KEGG.reaction ID from locus tag, if available, otherwise try EC number.
Entries of the finalised table are saved depending whether they were assigned
a KEGG.reaction ID (further use) or not (possible starting point for manual curation).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_table</strong> (<em>pd.DataFrame</em>) – The input table containing the output of get_ncbi_info().</p></li>
<li><p><strong>genome_summary</strong> (<em>string</em><em>, </em><em>default None</em>) – Path to a information file for the reference genomes to check if they are in KEGG.</p></li>
<li><p><strong>working_dir</strong> (<em>string</em>) – Path to the working directory - place to save genes with a KEGG.reaction ID.</p></li>
<li><p><strong>manual_dir</strong> (<em>string</em>) – Path to a directory to save the genes without KEGG assignment for possible manual curation.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The path to the file containing the genes with KEGG assignment.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>string</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.map_to_KEGG_row">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">map_to_KEGG_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loci_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.map_to_KEGG_row" title="Permalink to this definition"></a></dt>
<dd><p>Map a single entry of the table in map_to_KEGG() to KEGG.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>row</strong> (<em>pd.Series</em>) – A single row of the table.</p></li>
<li><p><strong>loci_list</strong> (<em>dict</em><em>, of </em><em>strings</em><em>, </em><em>default None</em><em> (</em><em>searches all entries</em><em>)</em>) – Dictionary of starting letters of locus tags belong to organisms, that are listed in KEGG
and the corresponding organism three-letter code</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The edited row.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.Series</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.map_to_NCBI_efetch_row">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">map_to_NCBI_efetch_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.map_to_NCBI_efetch_row" title="Permalink to this definition"></a></dt>
<dd><p>Map a single entry from the table in get_ncbi_info() to NCBI using EntrezDirect.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>row</strong> (<em>pd.Series</em><em>(</em><em>)</em>) – A single row of the table.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The edited row.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.Series()</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">draft</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fasta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigg_reac</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigg_meta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnx_chem_prop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnx_chem_xref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnx_reac_prop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnx_reac_xref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncbi_map</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncbi_dat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'locus_tag'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensitivity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'more-sensitive'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coverage</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">95.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">90.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_dna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_rna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memote</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.extension.run" title="Permalink to this definition"></a></dt>
<dd><p>Create a draft model.</p>
<p>Explaination missing ….</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>draft</strong> (<em>string</em>) – Path to the draft model</p></li>
<li><p><strong>gene_list</strong> (<em>string</em>) – Path to a csv file containing information on all the genes found in the annotated genome.</p></li>
<li><p><strong>fasta</strong> (<em>string</em>) – Path to the (protein) FASTA file containing the CDS sequences</p></li>
<li><p><strong>db</strong> (<em>string</em>) – Path to the database used for running DIAMOND.</p></li>
<li><p><strong>dir</strong> (<em>string</em>) – Path to the directory for the output (directories).</p></li>
<li><p><strong>bigg_reac</strong> (<em>string</em>) – Path to the BiGG reaction namespace file (rewritten version).</p></li>
<li><p><strong>bigg_meta</strong> (<em>string</em>) – Path to the BiGG metabolite namespace file (rewritten version).</p></li>
<li><p><strong>mnx_chem_prop</strong> (<em>string</em>) – Path to the MetaNetX chem_prop namespace file.</p></li>
<li><p><strong>mnx_chem_xref</strong> (<em>string</em>) – Path to the MetaNetX chem_xref namespace file.</p></li>
<li><p><strong>mnx_reac_prop</strong> (<em>string</em>) – Path to the MetaNetX reac_prop namespace file.</p></li>
<li><p><strong>mnx_reac_xref</strong> (<em>string</em>) – path to the MetaNetX reac_xref namespace file.</p></li>
<li><p><strong>ncbi_map</strong> (<em>string</em>) – Path to the ncbi information mapping file. Optional, but recommended.</p></li>
<li><p><strong>ncbi_dat</strong> (<em>string</em>) – Path to the ncbi database information file. Optional, but recommended.</p></li>
<li><p><strong>id</strong> (<em>string</em>) – Name of the column of the csv file that contains the entries that were used as gene identifiers in the draft model.
Default is ‘locus_tag’.</p></li>
<li><p><strong>sensitivity</strong> (<em>string</em><em>, </em><em>optional</em>) – Sensitivity mode for DIAMOND blastp run.
Can be sensitive, more-sensitive, very-sensitive or ultra-sensitive.
Default is more-sensitive.</p></li>
<li><p><strong>coverage</strong> (<em>float</em><em>, </em><em>optional</em>) – Threshold value for the query coverage for DIAMOND.
Default is 95.0.</p></li>
<li><p><strong>pid</strong> (<em>float</em><em>, </em><em>optional</em>) – PID (percentage identity value) to filter the blast hist by.
Default is 90.0, only hits equal or above the given value are kept.’)</p></li>
<li><p><strong>threads</strong> (<em>int</em>) – Number of threads to be used.
Default is 2.</p></li>
<li><p><strong>exclude_dna</strong> (<em>bool</em>) – Exclude reactions with DNA in their name when added.
Default is True.</p></li>
<li><p><strong>exclude_rna</strong> (<em>bool</em>) – Exclude reactions with RNA in their name when added.
Default is True.</p></li>
<li><p><strong>memote</strong> (<em>bool</em>) – Use memote on the extended model.
Default is False</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.core.refinement.smoothing">
<span id="specimen-core-refinement-smoothing-module"></span><h2>specimen.core.refinement.smoothing module<a class="headerlink" href="#module-specimen.core.refinement.smoothing" title="Permalink to this heading"></a></h2>
<p>Perform part 4 of the refinement: smoothing.</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.smoothing.adjust_BOF">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.smoothing.</span></span><span class="sig-name descname"><span class="pre">adjust_BOF</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dna_weight_fraction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weight_frac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.smoothing.adjust_BOF" title="Permalink to this definition"></a></dt>
<dd><p>Adjust the model’s BOF using BOFdat. Currently implemented are step 1
dna coefficients and step 2.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome</strong> (<em>string</em>) – Path to the genome (e.g. .fna) FASTA file.</p></li>
<li><p><strong>model_file</strong> – Path to the sbml (.xml) file of the model.</p></li>
<li><p><strong>model_file</strong> – string</p></li>
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The loaded genome-scale metabolic model (from the string above).</p></li>
<li><p><strong>dna_weight_fraction</strong> (<em>float</em>) – DNA weight fraction for BOF step 1.</p></li>
<li><p><strong>weight_frac</strong> (<em>float</em>) – Weight fraction for the second step of BOFdat (enzymes and ions)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The updated BOF reaction.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>string</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.smoothing.check_for_cycles">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.smoothing.</span></span><span class="sig-name descname"><span class="pre">check_for_cycles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.smoothing.check_for_cycles" title="Permalink to this definition"></a></dt>
<dd><p>Check a given metabolic model for futile cycles and energy generating cycles.
As the correction of those is highly dependant on biological evidence, the correction
has to be performed manually after using this function to check the existence and possible
reasons for cycles.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>model</strong> (<em>cobra.Model</em>) – The genome–scale metabolic model.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.smoothing.perform_mcc">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.smoothing.</span></span><span class="sig-name descname"><span class="pre">perform_mcc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">apply</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.smoothing.perform_mcc" title="Permalink to this definition"></a></dt>
<dd><p>Use the MassChargeCuration tool on the model. Either apply it directly
or on a copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The model to use the tool on.</p></li>
<li><p><strong>dir</strong> (<em>string</em>) – Path of the directory to save MCC output in.</p></li>
<li><p><strong>apply</strong> (<em>bool</em><em>, </em><em>true if MCC should be used directly on the model.</em>) – Option, if MCC should be used directly on the model or not.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The model (updated or not).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cobra.Model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.smoothing.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.smoothing.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mcc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'skip'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dna_weight_frac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.023</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ion_weight_frac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memote</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.core.refinement.smoothing.run" title="Permalink to this definition"></a></dt>
<dd><p>Perform the fourth step of the refinement, smoothing, on a model.</p>
<p>The fourth step of the refinment, smoothing, includes:
- Mass-Charge curation
- checking for futile and energy generating cycles
- adjusting Biomass objective function parameters based on genome</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome</strong> (<em>string</em>) – Path to the genome FASTA (e.g. .fna) file of your genome.</p></li>
<li><p><strong>model</strong> (<em>string</em>) – Path to the model. Should be sbml.</p></li>
<li><p><strong>dir</strong> (<em>string</em>) – Path of the output directory.</p></li>
<li><p><strong>mcc</strong> (<em>string</em>) – Option for the Mass-Charge curation.
Can be ‘skip’, ‘apply’ (directly use MCC on model) or ‘extra’ (only generate information).
Default is ‘skip’.</p></li>
<li><p><strong>dna_weight_frac</strong> (<em>float</em>) – DNA weight fraction to use for BOFdat.
Default is 0.023 for Klebsiella pneumoniae based on Liao et al.</p></li>
<li><p><strong>ion_weight_frac</strong> (<em>float</em>) – Ion weight fraction to use for BOFdat.
Default is the BOFdat default.</p></li>
<li><p><strong>memote</strong> (<em>bool</em>) – Option to run memote after the refinement.
Default is False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.core.refinement">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-specimen.core.refinement" title="Permalink to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="specimen.core.html" class="btn btn-neutral float-left" title="specimen.core package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="specimen.util.html" class="btn btn-neutral float-right" title="specimen.util package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Carolin Brune.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>