<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>specimen.classes package &mdash; SPECIMEN 1.0.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="specimen.core package" href="specimen.core.html" />
    <link rel="prev" title="Contents of SPECIMEN" href="specimen.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            SPECIMEN
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="about-pipeline.html">About the SPECIMEN</a></li>
<li class="toctree-l1"><a class="reference internal" href="run-pipeline.html">Run the Pipeline</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="specimen.html">Contents of SPECIMEN</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="specimen.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">specimen.classes package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-specimen.classes.medium">specimen.classes.medium module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-specimen.classes.reports">specimen.classes.reports module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-specimen.classes">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="specimen.core.html">specimen.core package</a></li>
<li class="toctree-l3"><a class="reference internal" href="specimen.util.html">specimen.util package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="specimen.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="specimen.html#module-specimen.cmd_access">specimen.cmd_access module</a></li>
<li class="toctree-l2"><a class="reference internal" href="specimen.html#module-specimen.workflow">specimen.workflow module</a></li>
<li class="toctree-l2"><a class="reference internal" href="specimen.html#module-specimen">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="dev-notes.html">Notes for Developers</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">SPECIMEN</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="specimen.html">Contents of SPECIMEN</a></li>
      <li class="breadcrumb-item active">specimen.classes package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/specimen.classes.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="specimen-classes-package">
<h1>specimen.classes package<a class="headerlink" href="#specimen-classes-package" title="Permalink to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading"></a></h2>
</section>
<section id="module-specimen.classes.medium">
<span id="specimen-classes-medium-module"></span><h2>specimen.classes.medium module<a class="headerlink" href="#module-specimen.classes.medium" title="Permalink to this heading"></a></h2>
<p>Classes to import, handle, manipulate and save media.</p>
<dl class="py class">
<dt class="sig sig-object py" id="specimen.classes.medium.Compound">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">specimen.classes.medium.</span></span><span class="sig-name descname"><span class="pre">Compound</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">formula</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exchange_flux</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.Compound" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The Compound objects describe the compounds of a Medium object.</p>
<p>The name, BiGG ID, molecule formula and exchange_flux in the medium are saved as attributes
of the Compound objects. While not directly needed for the object, when using a
Compound object as a part of a medium object, at least the formula attribute
should be set.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>string</em>) – The name of the compound.</p></li>
<li><p><strong>bigg</strong> (<em>string</em>) – The BiGG ID (without compartment e.g. ‘_e’) of the compound.</p></li>
<li><p><strong>formula</strong> (<em>string</em>) – The molecule formula (without charges, e.g. NH4 for ammonia).</p></li>
<li><p><strong>exchange_flux</strong> (<em>double</em>) – The exchange_flux of the compound in the medium.</p></li>
</ul>
</dd>
<dt class="field-even">Variables<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>name</strong> (<em>string</em>) – The name of the compound.</p></li>
<li><p><strong>bigg</strong> (<em>string</em>) – The BiGG ID (without compartment e.g. ‘_e’) of the compound.</p></li>
<li><p><strong>formula</strong> (<em>string</em>) – The molecule formula (without charges, e.g. NH4 for ammonia).</p></li>
<li><p><strong>exchange_flux</strong> (<em>double</em>) – The exchange_flux of the compound in the medium.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.medium.Compound.bigg_to_ex">
<span class="sig-name descname"><span class="pre">bigg_to_ex</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.Compound.bigg_to_ex" title="Permalink to this definition"></a></dt>
<dd><p>Tranform the BiGG ID into the BiGG ID of corresponding the exchange reactions.
If not BiGG ID is found, tries to use the formula instead.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ValueError</span></code>: ‘Both bigg and formula attributes are not set, conversion not possible.’</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The BiGG ID of the corresponding exchange reactions.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>string</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.medium.Compound.get_elements">
<span class="sig-name descname"><span class="pre">get_elements</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.Compound.get_elements" title="Permalink to this definition"></a></dt>
<dd><p>Count the chemical elements of the Compound.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dictionary of the chemical elements and their counts.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="specimen.classes.medium.Medium">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">specimen.classes.medium.</span></span><span class="sig-name descname"><span class="pre">Medium</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">description</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compounds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.Medium" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The Medium class objects contain the name and composition of a medium
for genome-scale metabolic modelling.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>string</em>) – Name of the medium.</p></li>
<li><p><strong>compounds</strong> (<em>list</em><em> of </em><em>Compounds</em>) – Name (ideally BiGG identifier) and exchange_fluxs of the compounds of the medium.</p></li>
</ul>
</dd>
<dt class="field-even">Variables<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>name</strong> (<em>string</em>) – Name of the medium.</p></li>
<li><p><strong>compounds</strong> (<em>dict</em><em>, </em><em>string</em><em> (</em><em>identifer</em><em>) </em><em>as key and float</em><em> (</em><em>exchange_flux in mM</em><em>) </em><em>as value.</em>) – Name (ideally BiGG identifier) and exchange_fluxs of the compounds of the medium.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.medium.Medium.add_compound">
<span class="sig-name descname"><span class="pre">add_compound</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.Medium.add_compound" title="Permalink to this definition"></a></dt>
<dd><p>Add a compound to the medium.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>c</strong> (<a class="reference internal" href="#specimen.classes.medium.Compound" title="specimen.classes.medium.Compound"><em>Compound</em></a>) – The compound to be added.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ValueError</span></code>: ‘Neither bigg id or formula found. Cannot add to medium.’</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.medium.Medium.combine">
<span class="sig-name descname"><span class="pre">combine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.Medium.combine" title="Permalink to this definition"></a></dt>
<dd><p>Combine the medium with another medium.</p>
<p>Add all elements of the second medium, which keys are not in the first
medium to the first one.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The combined medium (new object).</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#specimen.classes.medium.Medium" title="specimen.classes.medium.Medium">Medium</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.medium.Medium.export_to_cobra">
<span class="sig-name descname"><span class="pre">export_to_cobra</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">default_c</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10.0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.Medium.export_to_cobra" title="Permalink to this definition"></a></dt>
<dd><p>Export the medium into a cobra model medium format.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The medium in a format that can be directly set as a medium of a model.</p>
</dd>
<dt class="field-even">Rytpe<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.medium.Medium.get_compound_with_attribute">
<span class="sig-name descname"><span class="pre">get_compound_with_attribute</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">formula</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exchange_flux</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.Medium.get_compound_with_attribute" title="Permalink to this definition"></a></dt>
<dd><p>Get all compounds that match certain attribute(s).</p>
<p>Search for compounds that have the given name/bigg/formula/exchange_flux.
Default for all attributes is None, which means a skip for this attribute</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>string</em>) – The name of the compound.</p></li>
<li><p><strong>bigg</strong> (<em>string</em>) – The BiGG ID (without compartment e.g. ‘_e’) of the compound.</p></li>
<li><p><strong>formula</strong> (<em>string</em>) – The molecule formula (without charges, e.g. NH4 for ammonia).</p></li>
<li><p><strong>exchange_flux</strong> (<em>double</em>) – The exchange_flux of the compound in the medium.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The compounds that match the given attributes (all!).
Return an empty list if no match is found.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list, of Compound objects</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.medium.Medium.get_source_of">
<span class="sig-name descname"><span class="pre">get_source_of</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">element</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.Medium.get_source_of" title="Permalink to this definition"></a></dt>
<dd><p>Get the names of the compounds that are a source for a given element,
meaning that the given element is found in the formula.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>element</strong> (<em>string</em>) – The element to get the sources of.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The names of the compounds that are a source of the given element.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list of strings</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.medium.Medium.has_compound">
<span class="sig-name descname"><span class="pre">has_compound</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.Medium.has_compound" title="Permalink to this definition"></a></dt>
<dd><p>Checks, if a compound is part of the medium.</p>
<p>The check can either be performed with a Compound object or a name of a compound.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>c</strong> (<a class="reference internal" href="#specimen.classes.medium.Compound" title="specimen.classes.medium.Compound"><em>Compound</em></a><em> or </em><em>string</em>) – Either name or Compound to search the Medium for.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">TypeError</span></code>: ‘’Unknown type for c (compound)’</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.medium.Medium.is_aerobic">
<span class="sig-name descname"><span class="pre">is_aerobic</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.Medium.is_aerobic" title="Permalink to this definition"></a></dt>
<dd><p>Check if the medium is aerobic = contains O2.</p>
<p>Checks the medium for the name ‘o2’ and the formula ‘O2’</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The result of the check (True if medium is aerobic).</p>
</dd>
<dt class="field-even">Rytpe<span class="colon">:</span></dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.medium.Medium.make_aerobic">
<span class="sig-name descname"><span class="pre">make_aerobic</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10.0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.Medium.make_aerobic" title="Permalink to this definition"></a></dt>
<dd><p>Make the medium aerobic (if it isn’t already) by adding O2.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>c</strong> (<em>double</em>) – The exchange_flux of O2.
Default is 10.0</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.medium.Medium.make_anaerobic">
<span class="sig-name descname"><span class="pre">make_anaerobic</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.Medium.make_anaerobic" title="Permalink to this definition"></a></dt>
<dd><p>Make the medium anaerobic by removing elemental oxygen from it.</p>
<p>Removes oxygen by searching for compounds with the formula ‘O2’.</p>
<dl class="simple">
<dt>&#64;TODO</dt><dd><p>add option:
remove all additional oxygen sources, that are not needed for other elements?</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.medium.Medium.remove_compound">
<span class="sig-name descname"><span class="pre">remove_compound</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.Medium.remove_compound" title="Permalink to this definition"></a></dt>
<dd><p>Remove a compound from the medium.</p>
<p>The removal can either be performed with a Compound object or the name of a compound.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>c</strong> (<a class="reference internal" href="#specimen.classes.medium.Compound" title="specimen.classes.medium.Compound"><em>Compound</em></a><em> or </em><em>string</em>) – Either name or Compound to search the Medium for.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.medium.Medium.set_source_of">
<span class="sig-name descname"><span class="pre">set_source_of</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">element</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_all</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.Medium.set_source_of" title="Permalink to this definition"></a></dt>
<dd><p>Set the source of an element to a given new source (Compound).</p>
<p>This function replaces either all old sources with the new one (remove_all=True)
or just the compounds that are not the exclusive source for any ofther element
(remove_all=False).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>element</strong> (<em>string</em>) – The chemical element.</p></li>
<li><p><strong>new_source</strong> (<a class="reference internal" href="#specimen.classes.medium.Compound" title="specimen.classes.medium.Compound"><em>Compound</em></a>) – The new source for the element.</p></li>
<li><p><strong>remove_all</strong> (<em>bool</em>) – Option to eitherr remove all possible sources
of the element or only those that serve NOT as the exclusive source
for another element.
Default is the latter option (set to False).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.medium.Medium.validate_for_model">
<span class="sig-name descname"><span class="pre">validate_for_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.Medium.validate_for_model" title="Permalink to this definition"></a></dt>
<dd><p>Validate a medium to be used for a cobra model.</p>
<p>Checks, if all exchange reactions created from the medium are present in
the model. If not,</p>
<ul class="simple">
<li><p>add=False: removes corresponding compounds from the medium</p></li>
<li><p>&#64;TODO add=True: add the missing reaction (and metabolite to the model)</p></li>
</ul>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.classes.medium.add_medium_to_db">
<span class="sig-prename descclassname"><span class="pre">specimen.classes.medium.</span></span><span class="sig-name descname"><span class="pre">add_medium_to_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">medium</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.add_medium_to_db" title="Permalink to this definition"></a></dt>
<dd><p>Add a medium to a media database file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>medium</strong> (<a class="reference internal" href="#specimen.classes.medium.Medium" title="specimen.classes.medium.Medium"><em>Medium</em></a>) – The medium to be added.</p></li>
<li><p><strong>db_path</strong> (<em>string</em>) – Path to the database.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.classes.medium.db_to_table">
<span class="sig-prename descclassname"><span class="pre">specimen.classes.medium.</span></span><span class="sig-name descname"><span class="pre">db_to_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">db</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.db_to_table" title="Permalink to this definition"></a></dt>
<dd><p>Transform a medium database (dictionary of name:Medium) into a DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>db</strong> (<em>dict</em><em>, </em><em>{name:Medium}</em>) – The medium dictionary.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The table.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.classes.medium.from_table">
<span class="sig-prename descclassname"><span class="pre">specimen.classes.medium.</span></span><span class="sig-name descname"><span class="pre">from_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.from_table" title="Permalink to this definition"></a></dt>
<dd><p>Construct a medium from a table.</p>
<p>The table needs to have the following columns:
medium,description,compound,bigg_id,formula, exchange_flux</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>table</strong> (<em>pd.DataFrame</em>) – The table.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The constructed medium.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#specimen.classes.medium.Medium" title="specimen.classes.medium.Medium">Medium</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.classes.medium.import_medium_from_cobra">
<span class="sig-prename descclassname"><span class="pre">specimen.classes.medium.</span></span><span class="sig-name descname"><span class="pre">import_medium_from_cobra</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.import_medium_from_cobra" title="Permalink to this definition"></a></dt>
<dd><p>Import a medium from a cobra model into a Medium ob object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>model</strong> (<em>cobra.Model</em>) – The cobra model.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The imported medium.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#specimen.classes.medium.Medium" title="specimen.classes.medium.Medium">Medium</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.classes.medium.load_media_db">
<span class="sig-prename descclassname"><span class="pre">specimen.classes.medium.</span></span><span class="sig-name descname"><span class="pre">load_media_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.load_media_db" title="Permalink to this definition"></a></dt>
<dd><p>Load a media database into a dictionary of media (name:medium).</p>
<p>The database should be a csv file with ‘;’ as seperators.
The first line should be the following header:
medium;description;compound;bigg_id;formula;exchange_flux</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>path</strong> (<em>string</em>) – The path to the database file.
If None is given, reads the in-build media database.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The media found in the database as a dictionary of names and Media.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict, {string:Medium}</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.classes.medium.model_minimal_medium">
<span class="sig-prename descclassname"><span class="pre">specimen.classes.medium.</span></span><span class="sig-name descname"><span class="pre">model_minimal_medium</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">objective</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'flux'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">growth_rate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">open_exchanges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.model_minimal_medium" title="Permalink to this definition"></a></dt>
<dd><p>Get the minimal medium based on different objectives:
- ‘flux’:      find the minimal fluxes based in current medium.
- ‘medium’:    find the minimal number of compounds for the current medium.
- ‘exchanges’: find the minimal number of compounds in a medium based on all avaiblae exchange reactions in the model.
Note: there may be multiple solution for the minimisation, but only 1 will be returned</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The model to minimize the medium for.</p></li>
<li><p><strong>objective</strong> (<em>string</em>) – Objective for the minimisation task. Options listed above.</p></li>
<li><p><strong>growth_rate</strong> (<em>double</em><em>, </em><em>only for objectives medium and exchanges</em>) – Minimum growth rate the model has to archieve.</p></li>
<li><p><strong>open_exchanges</strong> (<em>bool</em>) – If set to True assigns large upper bound to all import reactions.
Author’s note: setting is to true leads to an endlessly running program for me.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ValueError</span></code>: ‘Unknown objective for minimisation.’</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The medium that is a solution for the minimisation task.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#specimen.classes.medium.Medium" title="specimen.classes.medium.Medium">Medium</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.classes.medium.save_db">
<span class="sig-prename descclassname"><span class="pre">specimen.classes.medium.</span></span><span class="sig-name descname"><span class="pre">save_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.medium.save_db" title="Permalink to this definition"></a></dt>
<dd><p>Save a medium database (dictionary of name:Medium) into a csv file
(seperator is ‘;’).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>db</strong> (<em>dict</em><em>, </em><em>{name:Medium}</em>) – The medium dictionary.</p></li>
<li><p><strong>out_path</strong> (<em>string</em>) – Path to the file for the output.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.classes.reports">
<span id="specimen-classes-reports-module"></span><h2>specimen.classes.reports module<a class="headerlink" href="#module-specimen.classes.reports" title="Permalink to this heading"></a></h2>
<p>Classes to generate, handle, manipulate and save reports.</p>
<dl class="py class">
<dt class="sig sig-object py" id="specimen.classes.reports.GrowthAnalysisReport">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">specimen.classes.reports.</span></span><span class="sig-name descname"><span class="pre">GrowthAnalysisReport</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">growth_sim_results</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minimal_medium</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">auxo_results</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.GrowthAnalysisReport" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>GrowthAnalysisReport saved the results of analyse_growth() and allows for saving and visualising the results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>growth_sim_results</strong> (<em>dict</em><em>, </em><em>medium name : growth rate</em><em> (</em><em>optional</em><em>)</em>) – The results of  <code class="xref py py-func docutils literal notranslate"><span class="pre">specimen.classes.medium.growth_simulation()</span></code></p></li>
<li><p><strong>minimal_medium</strong> (<a class="reference internal" href="#specimen.classes.medium.Medium" title="specimen.classes.medium.Medium"><em>medium.Medium</em></a><em> (</em><em>optional</em><em>)</em>) – The minimal medium for the underlying model.</p></li>
<li><p><strong>auxo_results</strong> (<em>nested dict</em>) – The results of auxotrophies_simulation().</p></li>
</ul>
</dd>
<dt class="field-even">Variables<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>growth_rates</strong> (<em>dict</em><em>, </em><em>medium name : growth rate</em><em> (</em><em>optional</em><em>)</em>) – The results of  medium.growth_simulation()</p></li>
<li><p><strong>minimal_medium</strong> (<a class="reference internal" href="#specimen.classes.medium.Medium" title="specimen.classes.medium.Medium"><em>medium.Medium</em></a><em> (</em><em>optional</em><em>)</em>) – The minimal medium for the underlying model.</p></li>
<li><p><strong>auxotrophy_results</strong> (<em>nested dict</em>) – The results of auxotrophies_simulation().</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.reports.GrowthAnalysisReport.auxotrophies_table">
<span class="sig-name descname"><span class="pre">auxotrophies_table</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.GrowthAnalysisReport.auxotrophies_table" title="Permalink to this definition"></a></dt>
<dd><p>Tranform the auxotrophy_results attribute into a table format.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the table (amino acids vs medium names)</p>
</dd>
<dt class="field-even">Rytpe<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.reports.GrowthAnalysisReport.growth_rates_table">
<span class="sig-name descname"><span class="pre">growth_rates_table</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.GrowthAnalysisReport.growth_rates_table" title="Permalink to this definition"></a></dt>
<dd><p>Turn the dictionary of growth rates into a DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The DataFrame.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.reports.GrowthAnalysisReport.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.GrowthAnalysisReport.save" title="Permalink to this definition"></a></dt>
<dd><p>Save the report.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dir</strong> (<em>string</em>) – Path to a directory to save the output to.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.reports.GrowthAnalysisReport.visualise_auxotrophies">
<span class="sig-name descname"><span class="pre">visualise_auxotrophies</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.GrowthAnalysisReport.visualise_auxotrophies" title="Permalink to this definition"></a></dt>
<dd><p>Visualise the results of the auxotrophy tests as a heatmap.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The generate heatmap.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>plt.Figure or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.reports.GrowthAnalysisReport.visualise_growth">
<span class="sig-name descname"><span class="pre">visualise_growth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">unit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'h'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.GrowthAnalysisReport.visualise_growth" title="Permalink to this definition"></a></dt>
<dd><p>Visualise thegrowth of the simulation results.
Either the growth rates themselves can be visualised per hour
or per min or the doubling times in minutes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>unit</strong> (<em>string</em><em>, </em><em>choices</em><em> [</em><em>'h'</em><em>,</em><em>'min'</em><em>,</em><em>'dt'</em><em>]</em>) – Determines the type of visualisation that is prodoced.
Can be ‘h’, ‘min’ or ‘dt’. First two ones are the unit for the
growth rate visualisation, the last one makes the doubling time graphic.
Default is ‘h’.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The visualisation.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>plt.fig</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="specimen.classes.reports.ModelStatisticsReport">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">specimen.classes.reports.</span></span><span class="sig-name descname"><span class="pre">ModelStatisticsReport</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">total_reac</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">total_meta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">total_gene</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reac_origin_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reac_with_gpr</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.ModelStatisticsReport" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>ModelStatisticsReport objects are used to store statistical information
about a model. Additionally, this class provides functionalities to report
said information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model_id</strong> (<em>string</em>) – The model’s ID (its name).</p></li>
<li><p><strong>total_reac</strong> (<em>int</em>) – Total number of reactions in the model.</p></li>
<li><p><strong>total_gene</strong> (<em>int</em>) – Total number of genes in the model.</p></li>
<li><p><strong>total_meta</strong> (<em>int</em>) – Total number of metabolites in the model.</p></li>
<li><p><strong>reac_origin_counts</strong> (<em>dict</em>) – Descriptors and corresponding count for where the reactions where build from.
Should be a dictionary of string and ints.</p></li>
<li><p><strong>reac_with_gpr</strong> (<em>int</em>) – Number of reactions that are associated with a gene reaction rule (GPR)</p></li>
</ul>
</dd>
<dt class="field-even">Variables<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>model_id</strong> (<em>string</em>) – The model’s ID the statistic is about.</p></li>
<li><p><strong>reac</strong> (<em>dict</em><em> (</em><em>string as above</em><em>, </em><em>int</em><em> or </em><em>dict as value</em><em>)</em>) – Information about the reaction objects, including “total” count, “origin” count and number of reactions “with gpr”.</p></li>
<li><p><strong>meta</strong> (<em>dict</em><em> (</em><em>string as above</em><em>, </em><em>int as value</em><em>)</em>) – Information about the metabolite objects, including “total” count.</p></li>
<li><p><strong>gene</strong> (<em>dict</em><em> (</em><em>string as above</em><em>, </em><em>int as value</em><em>)</em>) – Information about the gene objects, including “total” count.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.reports.ModelStatisticsReport.get_statistics">
<span class="sig-name descname"><span class="pre">get_statistics</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.ModelStatisticsReport.get_statistics" title="Permalink to this definition"></a></dt>
<dd><p>Generate a string got the statistics.</p>
<p>&#64;TODO find a better format, maybe html?</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the statistics</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>string</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.reports.ModelStatisticsReport.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.ModelStatisticsReport.save" title="Permalink to this definition"></a></dt>
<dd><p>Save the report.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dir</strong> (<em>string</em>) – Path to a directory to save the output to.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="specimen.classes.reports.PanCoreAnalysisReport">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">specimen.classes.reports.</span></span><span class="sig-name descname"><span class="pre">PanCoreAnalysisReport</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">core_reac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pan_reac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">novel_reac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.PanCoreAnalysisReport" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>PanCoreAnalysisReport objects are used to store, handle and present the
information from the pan-core analysis (comparison of a model and a pan-core model).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The model the test was performed on.</p></li>
<li><p><strong>core_reac</strong> (<em>list</em><em>, </em><em>elements are cobra.Reaction</em>) – List of reactions characterised as core.</p></li>
<li><p><strong>pan_reac</strong> (<em>list</em><em>, </em><em>elements are cobra.Reaction</em>) – List of reactions characterised as pan.</p></li>
<li><p><strong>novel_reac</strong> (<em>list</em><em>, </em><em>elements are cobra.Reaction</em>) – List of reactions characterised as core.</p></li>
</ul>
</dd>
<dt class="field-even">Variables<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>model</strong> (<em>cobra.Model</em>) – The model the test was performed on.</p></li>
<li><p><strong>core_reac</strong> (<em>list</em><em>, </em><em>elements are cobra.Reaction</em>) – List of reactions characterised as core.</p></li>
<li><p><strong>pan_reac</strong> (<em>list</em><em>, </em><em>elements are cobra.Reaction</em>) – List of reactions characterised as pan.</p></li>
<li><p><strong>novel_reac</strong> (<em>list</em><em>, </em><em>elements are cobra.Reaction</em>) – List of reactions characterised as core.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.reports.PanCoreAnalysisReport.get_reac_counts">
<span class="sig-name descname"><span class="pre">get_reac_counts</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.PanCoreAnalysisReport.get_reac_counts" title="Permalink to this definition"></a></dt>
<dd><p>Return a dictionary of the counts of the reactions types (core, pan, novel).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.reports.PanCoreAnalysisReport.isValid">
<span class="sig-name descname"><span class="pre">isValid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">check</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reaction-count'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.PanCoreAnalysisReport.isValid" title="Permalink to this definition"></a></dt>
<dd><p>Check if a certaing part of the analysis is valid.</p>
<dl class="simple">
<dt>Currently possible checks:</dt><dd><dl class="simple">
<dt>reaction-count<span class="classifier">check if the number of reactions in the model</span></dt><dd><p>equal the sum of the novel, pan and core reactions</p>
</dd>
</dl>
</dd>
<dt>&#64;TODO</dt><dd><p>implements more checks</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>check</strong> (<em>string</em>) – Part of the report to check if it is valid.
Default is ‘reaction-count’. Options are in the function description.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ValueError</span></code>: ‘Unknown string for parameter check: ‘</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>True, if test was successful.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.reports.PanCoreAnalysisReport.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.PanCoreAnalysisReport.save" title="Permalink to this definition"></a></dt>
<dd><p>Save the results inside a PanCoreAnalysisReport object.</p>
<p>The function creates a new folder ‘pan-core-analysis’
inside the given directory and creates the following documents:</p>
<ul class="simple">
<li><p>table_reactions.tsv : reactions ID mapped to their labels</p></li>
<li><p>visualise_reactions : donut chart of the values above</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dir</strong> (<em>string</em>) – Path to a directory to save the output to.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.reports.PanCoreAnalysisReport.visualise_reactions">
<span class="sig-name descname"><span class="pre">visualise_reactions</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.PanCoreAnalysisReport.visualise_reactions" title="Permalink to this definition"></a></dt>
<dd><p>Visualise the results of the pan-core analysis for the reactions as a donut chart.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The generated visualisation object (donut chart).</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>matplotlib.figure.Figure</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="specimen.classes.reports.PathwayAnalysisReport">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">specimen.classes.reports.</span></span><span class="sig-name descname"><span class="pre">PathwayAnalysisReport</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">total_reac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kegg_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kegg_global</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kegg_over</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kegg_rest</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.PathwayAnalysisReport" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>PathwayAnalysisReport saves the results of kegg_pathway_analysis() (see core.analysis)
and allows for saving and visualising the results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>total_reac</strong> (<em>int</em><em> (</em><em>optional</em><em>)</em>) – Total number of reactions in the model</p></li>
<li><p><strong>kegg_count</strong> (<em>int</em><em> (</em><em>optional</em><em>)</em>) – Number of reactions with KEGG pathway annotation.</p></li>
<li><p><strong>kegg_global</strong> (<em>dict</em><em>(</em><em>str</em><em>,</em><em>int</em><em>)</em>) – Identifiers and counts for the global level.</p></li>
<li><p><strong>kegg_over</strong> (<em>dict</em><em>(</em><em>str</em><em>,</em><em>int</em><em>) </em><em>(</em><em>optional</em><em>)</em>) – Identifiers and counts for the overview level.</p></li>
<li><p><strong>kegg_rest</strong> (<em>dict</em><em>(</em><em>str</em><em>,</em><em>int</em><em>) </em><em>(</em><em>optional</em><em>)</em>) – Identifiers and counts not match by overview and global level.</p></li>
</ul>
</dd>
<dt class="field-even">Variables<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>total_reac</strong> (<em>int</em>) – Total number of reactions in the model</p></li>
<li><p><strong>kegg_count</strong> (<em>int</em>) – Number of reactions with KEGG pathway annotation.</p></li>
<li><p><strong>kegg_global</strong> (<em>dict</em><em>(</em><em>str</em><em>,</em><em>int</em><em>)</em>) – Identifiers and counts for the global level.</p></li>
<li><p><strong>kegg_over</strong> (<em>dict</em><em>(</em><em>str</em><em>,</em><em>int</em><em>)</em>) – Identifiers and counts for the overview level.</p></li>
<li><p><strong>kegg_paths</strong> (<em>dict</em><em>(</em><em>str</em><em>,</em><em>int</em><em>)</em>) – Identifiers and counts not match by overview and global level.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.reports.PathwayAnalysisReport.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.PathwayAnalysisReport.save" title="Permalink to this definition"></a></dt>
<dd><p>Save the content of the report as plots.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dir</strong> (<em>string</em>) – Path to a directory to save the output to.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.reports.PathwayAnalysisReport.visualise_kegg_counts">
<span class="sig-name descname"><span class="pre">visualise_kegg_counts</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.PathwayAnalysisReport.visualise_kegg_counts" title="Permalink to this definition"></a></dt>
<dd><p>Visualise the amounts of reaction with and without
KEGG pathway annotation.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The generated plot.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>matplotlib.pyplot.figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="specimen.classes.reports.PathwayAnalysisReport.visualise_kegg_pathway">
<span class="sig-name descname"><span class="pre">visualise_kegg_pathway</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">plot_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'global'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'id'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#specimen.classes.reports.PathwayAnalysisReport.visualise_kegg_pathway" title="Permalink to this definition"></a></dt>
<dd><p>Visualise the KEGG pathway identifiers present.</p>
<p>Depending on the :plot_type:, different levels of pathway identifiers
are plotted:</p>
<ul class="simple">
<li><p>global: check and plot only the global pathway identifiers</p></li>
<li><p>overview: check and plot only the overview pathway identifiers</p></li>
<li><p>high: check and plot all identifiers grouped by their high level pathway identifiers. This option uses label=name, independedly of the input</p></li>
<li><p>all: check and plot all identifiers</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>plot_type</strong> (<em>string</em>) – Set the type of plot. Can be global, overview, high or existing.
Default is global.</p></li>
<li><p><strong>id</strong> (<em>string</em>) – Set the pathway label to identifier (ID) or actual name.
Default is ‘id’, other option is ‘name’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The generated plot.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.pyplot.figure</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-specimen.classes">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-specimen.classes" title="Permalink to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="specimen.html" class="btn btn-neutral float-left" title="Contents of SPECIMEN" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="specimen.core.html" class="btn btn-neutral float-right" title="specimen.core package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Carolin Brune.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>