<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>specimen.hqtb.core.refinement.extension &mdash; SPECIMEN 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../../../_static/pygments.css?v=559046d4" />
      <link rel="stylesheet" type="text/css" href="../../../../../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../../../../../_static/copybutton.css?v=76b2166b" />
      <link rel="stylesheet" type="text/css" href="../../../../../_static/custom_theme.css?v=ef5873f9" />

  
    <link rel="shortcut icon" href="../../../../../_static/LogoSPECIMEN.png"/>
  <!--[if lt IE 9]>
    <script src="../../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../../../../_static/documentation_options.js?v=8d563738"></script>
        <script src="../../../../../_static/doctools.js?v=888ff710"></script>
        <script src="../../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script src="../../../../../_static/clipboard.min.js?v=a7894cd8"></script>
        <script src="../../../../../_static/copybutton.js?v=7859de39"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../../index.html">
            
              <img src="../../../../../_static/LogoSPECIMEN.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Content</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../overview-pipes.html">Overview of the Pipelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../specimen.html">Contents of SPECIMEN</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../help.html">Help &amp; FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../dev-notes.html">Notes for Developers</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../../index.html">SPECIMEN</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">specimen.hqtb.core.refinement.extension</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for specimen.hqtb.core.refinement.extension</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Part one of the third step of the pipeline: refinement - extension.</span>

<span class="sd">Extents the model by mapping missing genes to reactions via NCBI, KEGG,</span>
<span class="sd">MetaNetX and BiGG.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">__author__</span> <span class="o">=</span> <span class="s1">&#39;Carolin Brune&#39;</span>

<span class="c1">################################################################################</span>
<span class="c1"># requirements</span>
<span class="c1">################################################################################</span>

<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">urllib.error</span>

<span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">from</span> <span class="nn">tqdm.auto</span> <span class="kn">import</span> <span class="n">tqdm</span>
<span class="n">tqdm</span><span class="o">.</span><span class="n">pandas</span><span class="p">()</span>
<span class="kn">import</span> <span class="nn">cobra</span>

<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Literal</span>

<span class="kn">from</span> <span class="nn">Bio.KEGG</span> <span class="kn">import</span> <span class="n">REST</span>
<span class="kn">from</span> <span class="nn">Bio.KEGG</span> <span class="kn">import</span> <span class="n">Enzyme</span><span class="p">,</span> <span class="n">Compound</span>

<span class="kn">from</span> <span class="nn">refinegems.utility.io</span> <span class="kn">import</span> <span class="n">kegg_reaction_parser</span><span class="p">,</span> <span class="n">load_a_table_from_database</span>
<span class="kn">from</span> <span class="nn">refinegems.utility.entities</span> <span class="kn">import</span> <span class="n">create_random_id</span><span class="p">,</span> <span class="n">get_reaction_annotation_dict</span><span class="p">,</span> <span class="n">match_id_to_namespace</span>
<span class="kn">from</span> <span class="nn">refinegems.utility.connections</span> <span class="kn">import</span> <span class="n">run_memote</span>

<span class="c1"># further required programs:</span>
<span class="c1">#        - DIAMOND, tested with version 0.9.14 (works only for certain sensitivity mode)</span>
<span class="c1">#                   tested with version 2.1.8 (works for all sensitivity modes for that version)</span>

<span class="c1">################################################################################</span>
<span class="c1"># functions</span>
<span class="c1">################################################################################</span>

<span class="c1"># ----------------------</span>
<span class="c1"># identify missing genes</span>
<span class="c1"># ----------------------</span>
<div class="viewcode-block" id="identify_missing_genes">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.identify_missing_genes">[docs]</a>
<span class="k">def</span> <span class="nf">identify_missing_genes</span><span class="p">(</span><span class="n">gene_path</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span><span class="n">cobra</span><span class="o">.</span><span class="n">Model</span><span class="p">,</span> 
                           <span class="nb">id</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">fasta_path</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dir</span><span class="p">:</span><span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Identify missing genes in a model based on the draft model and a list of genes.</span>
<span class="sd">    Save the missing genes&#39; sequences in a new FASTA file.</span>


<span class="sd">    Args:</span>
<span class="sd">        - gene_path (str): </span>
<span class="sd">            Path to the csv file containing the list of genes.</span>
<span class="sd">        - model (cobra.Model): </span>
<span class="sd">            The draft model.</span>
<span class="sd">        - id (str): </span>
<span class="sd">            Name of the column in gene_path, that equal the ids in the model.</span>
<span class="sd">        - fasta_path (str): </span>
<span class="sd">            Path to the FASTA file containing all genes for the genome.</span>
<span class="sd">        - dir (str): </span>
<span class="sd">            Path to the directory to save the new FASTA in.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: </span>
<span class="sd">            A table with the missing genes&#39; information.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># get list of all genes</span>
    <span class="n">gene_list_all</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gene_path</span><span class="p">)</span>
    <span class="c1"># filter list by genes in model</span>
    <span class="n">gene_list_model</span> <span class="o">=</span> <span class="p">[</span><span class="n">g</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">genes</span><span class="p">]</span>
    <span class="n">gene_list_missing</span> <span class="o">=</span> <span class="n">gene_list_all</span><span class="p">[</span><span class="n">gene_list_all</span><span class="p">[</span><span class="nb">id</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">gene_list_model</span><span class="p">)</span> <span class="o">==</span> <span class="kc">False</span><span class="p">]</span>
    <span class="c1"># extract wanted sequences from the FASTA and save new FASTA</span>
    <span class="n">fasta</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)</span>
    <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">((</span><span class="n">seq</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">fasta</span> <span class="k">if</span> <span class="n">seq</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">gene_list_missing</span><span class="p">[</span><span class="nb">id</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()),</span> <span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s1">&#39;missing_genes.fasta&#39;</span><span class="p">),</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">gene_list_missing</span></div>



<div class="viewcode-block" id="find_best_diamond_hits">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.find_best_diamond_hits">[docs]</a>
<span class="k">def</span> <span class="nf">find_best_diamond_hits</span><span class="p">(</span><span class="n">file</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">pid</span><span class="p">:</span><span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Identify the best hits from the DIAMOND run.</span>

<span class="sd">    Args:</span>
<span class="sd">        - file (str): </span>
<span class="sd">            File name (path) of the &quot;missing_genes&quot; file.</span>
<span class="sd">        - pid (float): </span>
<span class="sd">            Threshold value for the PID. </span>
<span class="sd">            Should be between 0.0 and 100.0 (given in percentage)</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: </span>
<span class="sd">            Table of the locus tags of the query and the NCBI accession version numbers of the RefSeqs.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># load diamond results</span>
    <span class="n">diamond_results</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">diamond_results</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;query_ID&#39;</span><span class="p">,</span> <span class="s1">&#39;subject_ID&#39;</span><span class="p">,</span> <span class="s1">&#39;PID&#39;</span><span class="p">,</span> <span class="s1">&#39;align_len&#39;</span><span class="p">,</span> <span class="s1">&#39;no_mismatch&#39;</span><span class="p">,</span> <span class="s1">&#39;no_gapopen&#39;</span><span class="p">,</span> <span class="s1">&#39;query_start&#39;</span><span class="p">,</span> <span class="s1">&#39;query_end&#39;</span><span class="p">,</span> <span class="s1">&#39;subject_start&#39;</span><span class="p">,</span> <span class="s1">&#39;subject_end&#39;</span><span class="p">,</span><span class="s1">&#39;E-value&#39;</span><span class="p">,</span><span class="s1">&#39;bitscore&#39;</span><span class="p">]</span>
    <span class="c1"># filter by PID</span>
    <span class="n">diamond_results</span> <span class="o">=</span> <span class="n">diamond_results</span><span class="p">[</span><span class="n">diamond_results</span><span class="p">[</span><span class="s1">&#39;PID&#39;</span><span class="p">]</span><span class="o">&gt;=</span><span class="n">pid</span><span class="p">]</span>
    <span class="c1"># get best hit for each gene in list</span>
    <span class="c1"># ...............................................</span>
    <span class="c1"># @TODO</span>
    <span class="c1"># NOTE: best hit = FIRST best hit for a locus_tag</span>
    <span class="c1">#       mainly a runtime optimisation</span>
    <span class="c1">#       however can lead to information loss</span>
    <span class="c1"># ...............................................</span>
    <span class="n">best_hits</span> <span class="o">=</span> <span class="n">diamond_results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">diamond_results</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;query_ID&#39;</span><span class="p">)[</span><span class="s1">&#39;PID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">idxmax</span><span class="p">()]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">]]</span>
    <span class="n">best_hits</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">,</span><span class="s1">&#39;ncbi_accession_version&#39;</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">best_hits</span></div>


<span class="c1"># -----------------------------</span>
<span class="c1"># map to additional information</span>
<span class="c1"># -----------------------------</span>

<span class="c1"># mapping zo NCBI</span>
<span class="c1"># ---------------</span>
<div class="viewcode-block" id="map_to_NCBI_efetch_row">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.map_to_NCBI_efetch_row">[docs]</a>
<span class="k">def</span> <span class="nf">map_to_NCBI_efetch_row</span><span class="p">(</span><span class="n">row</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Map a single entry from the table in :py:func:`~specimen.core.refinement.extension.get_ncbi_info` </span>
<span class="sd">    to NCBI using EntrezDirect.</span>

<span class="sd">    Args:</span>
<span class="sd">        - row (pd.Series): </span>
<span class="sd">            A single row of the table.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.Series: </span>
<span class="sd">            The edited row.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">#  EC number</span>
    <span class="c1"># fetch_ec = subprocess.run([&#39;efetch&#39;,&#39;-db&#39;,&#39;protein&#39;,&#39;-id&#39;,row[&quot;ncbi_accession_version&quot;],&#39;-format&#39;,&#39;gpc&#39;,&#39;|&#39;,&#39;xtract&#39;,&#39;-insd&#39;,&#39;Protein&#39;,&#39;EC_number&#39;], stdout=subprocess.PIPE, shell=False, encoding=&#39;utf-8&#39;)</span>
    <span class="n">fetch_ec</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="s1">&#39;efetch&#39;</span><span class="p">,</span><span class="s1">&#39;-db&#39;</span><span class="p">,</span><span class="s1">&#39;protein&#39;</span><span class="p">,</span><span class="s1">&#39;-id&#39;</span><span class="p">,</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;ncbi_accession_version&quot;</span><span class="p">],</span><span class="s1">&#39;-format&#39;</span><span class="p">,</span><span class="s1">&#39;gpc&#39;</span><span class="p">],</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">capture_output</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">fetch_ec</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="s1">&#39;xtract&#39;</span><span class="p">,</span><span class="s1">&#39;-insd&#39;</span><span class="p">,</span><span class="s1">&#39;Protein&#39;</span><span class="p">,</span><span class="s1">&#39;EC_number&#39;</span><span class="p">],</span> <span class="nb">input</span><span class="o">=</span><span class="n">fetch_ec</span><span class="o">.</span><span class="n">stdout</span><span class="p">,</span> <span class="n">capture_output</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">fetch_ec</span> <span class="o">=</span> <span class="n">fetch_ec</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">fetch_ec</span><span class="p">:</span>
        <span class="n">row</span><span class="p">[</span><span class="s1">&#39;EC number&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fetch_ec</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">row</span><span class="p">[</span><span class="s1">&#39;EC number&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;-&#39;</span>

    <span class="c1"># locus_tag_ref old_locus_tag GeneID</span>
    <span class="n">fetch_rest</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="s1">&#39;esearch&#39;</span><span class="p">,</span><span class="s1">&#39;-db&#39;</span><span class="p">,</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span><span class="s1">&#39;-query&#39;</span><span class="p">,</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;ncbi_accession_version&quot;</span><span class="p">]],</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">capture_output</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">fetch_rest</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="s1">&#39;efetch&#39;</span><span class="p">,</span><span class="s1">&#39;-format&#39;</span><span class="p">,</span><span class="s1">&#39;docsum&#39;</span><span class="p">],</span> <span class="nb">input</span><span class="o">=</span><span class="n">fetch_rest</span><span class="o">.</span><span class="n">stdout</span><span class="p">,</span> <span class="n">capture_output</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">fetch_rest</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="s1">&#39;xtract&#39;</span><span class="p">,</span><span class="s1">&#39;-pattern&#39;</span><span class="p">,</span><span class="s1">&#39;DocumentSummary&#39;</span><span class="p">,</span><span class="s1">&#39;-element&#39;</span><span class="p">,</span><span class="s1">&#39;Id,OtherAliases&#39;</span><span class="p">],</span> <span class="nb">input</span><span class="o">=</span><span class="n">fetch_rest</span><span class="o">.</span><span class="n">stdout</span><span class="p">,</span> <span class="n">capture_output</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">fetch_rest</span> <span class="o">=</span> <span class="n">fetch_rest</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">fetch_rest</span><span class="p">:</span>
        <span class="c1"># since there is no way to determine, which entry is the correct one,</span>
        <span class="c1"># if multiple entries are returned, always take the first one.</span>
        <span class="c1"># Assuming, if multiple are returned, they represent the same gene in different subjects.</span>
        <span class="n">fetch_rest</span> <span class="o">=</span> <span class="n">fetch_rest</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">id</span><span class="p">,</span> <span class="n">aliases</span> <span class="o">=</span> <span class="n">fetch_rest</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">row</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">id</span>
        <span class="n">aliases</span> <span class="o">=</span> <span class="p">[</span><span class="n">a</span><span class="o">.</span><span class="n">strip</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">aliases</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)]</span>
        <span class="n">row</span><span class="p">[</span><span class="s1">&#39;locus_tag_ref&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">aliases</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">aliases</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">row</span><span class="p">[</span><span class="s1">&#39;old_locus_tag&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">aliases</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="c1"># further names get ignored</span>
            <span class="c1"># .........................................</span>
            <span class="c1"># possible problem:</span>
            <span class="c1"># assumes, that aliases 1 and 2 are always</span>
            <span class="c1"># locus_tag_ref and old_locus_tag</span>
            <span class="c1"># .........................................</span>

    <span class="k">return</span> <span class="n">row</span></div>



<span class="c1"># @TODO</span>
<div class="viewcode-block" id="get_ncbi_info">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.get_ncbi_info">[docs]</a>
<span class="k">def</span> <span class="nf">get_ncbi_info</span><span class="p">(</span><span class="n">table</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">ncbi_map_file</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Retrieve information from NCBI via mapping to precomputed files.</span>

<span class="sd">       If no mapping exists, try via Entrez - implemented, but does not work well</span>

<span class="sd">       -&gt; problems with connecting successfully to NCBI.</span>

<span class="sd">       @TODO if time, rewrite using the Bioppython package, see if it works better</span>

<span class="sd">    Args:</span>
<span class="sd">        - table (pd.DataFrame):  </span>
<span class="sd">            A table containing the output of :py:func:`specimen.core.refinement.extension.find_best_diamond_hits`.</span>
<span class="sd">            Must have the column &#39;ncbi_accession_version&#39;.</span>
<span class="sd">        - ncbi_map_file (str, optional): </span>
<span class="sd">            A csv file containing the ncbi mapping. </span>
<span class="sd">            Defaults to None.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame:  </span>
<span class="sd">            A table with the added information.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># case 1: if precomputed mapping was given</span>
    <span class="k">if</span> <span class="n">ncbi_map_file</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t\t</span><span class="s1">Retrieving mapping from precomputed file.&#39;</span><span class="p">)</span>
        <span class="c1"># load pre-compiles ncbi mapping</span>
        <span class="n">ncbi_mapping</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">ncbi_map_file</span><span class="p">)</span>

        <span class="c1"># merge with input info table:</span>
        <span class="c1"># locus_tag ncbi_accession_version locus_tag_ref old_locus_tag GeneID EC number</span>
        <span class="c1"># ......................................................................</span>
        <span class="c1"># @TODO // @NOTE</span>
        <span class="c1">#    using first() takes the first mapped accession version organism</span>
        <span class="c1">#    instead of using all</span>
        <span class="c1">#    while testing, all matches returned same EC number, which is why</span>
        <span class="c1">#    this is used to get a better runtime</span>
        <span class="c1">#    however, if the locus tag play a more significant role,</span>
        <span class="c1">#    it might be better to remove the &#39;.first().reset_index()&#39;</span>
        <span class="c1"># ......................................................................</span>
        <span class="n">table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">ncbi_mapping</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;ncbi_accession_version&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">first</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

    <span class="c1"># case 2: no mapping exists</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t\t</span><span class="s1">No precomputed mapping. Retrieving information directly from NCBI.</span><span class="se">\n\t\t</span><span class="s1">This may take a while.&#39;</span><span class="p">)</span>
        <span class="n">table</span><span class="p">[</span><span class="s1">&#39;locus_tag_ref&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>
        <span class="n">table</span><span class="p">[</span><span class="s1">&#39;old_locus_tag&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>
        <span class="n">table</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>
        <span class="n">table</span><span class="p">[</span><span class="s1">&#39;EC number&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>
        <span class="n">table</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">progress_apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">map_to_NCBI_efetch_row</span><span class="p">(</span><span class="n">row</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">table</span></div>


<span class="c1"># mapping to KEGG</span>
<span class="c1"># ---------------</span>
<span class="c1"># @TODO</span>
<div class="viewcode-block" id="map_to_KEGG_row">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.map_to_KEGG_row">[docs]</a>
<span class="k">def</span> <span class="nf">map_to_KEGG_row</span><span class="p">(</span><span class="n">row</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span> <span class="n">loci_list</span><span class="p">:</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Map a single entry of the table in :py:func:`~specimen.core.refinement.extension.map_to_KEGG` to KEGG.</span>

<span class="sd">    Args:</span>
<span class="sd">        - row (pd.Series): </span>
<span class="sd">            A single row of the table.</span>
<span class="sd">        - loci_list (list[str], optional):</span>
<span class="sd">            List of starting letters of locus tags belong to organisms, that are listed in KEGG</span>
<span class="sd">            and the corresponding organism three-letter code. </span>
<span class="sd">            Defaults to None (= search all entries).</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.Series: </span>
<span class="sd">            The mapped row.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># --------------------------------</span>
    <span class="c1"># decide search based on loci list</span>
    <span class="c1"># --------------------------------</span>
    <span class="c1"># @TODO if no list is given, all locus tags will be searched</span>
    <span class="c1"># if no list is given, all locus tags will be searched</span>
    <span class="n">search_locus</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="c1"># if a list of locus tags in KEGG is given</span>
    <span class="k">if</span> <span class="n">loci_list</span><span class="p">:</span>
        <span class="c1"># use it to search KEGG more efficiently</span>
        <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">x</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;locus_tag_ref&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">loci_list</span><span class="p">):</span>
            <span class="c1"># search only if loci might indeed be in KEGG</span>
            <span class="n">search_locus</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="c1">#@TODO this assumes that the locus tag is of the format XXX_YYY</span>
            <span class="c1"># can this be written more general?</span>
            <span class="n">organism_abbrv</span> <span class="o">=</span> <span class="n">loci_list</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;locus_tag_ref&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]]</span><span class="o">+</span><span class="s1">&#39;:&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># if locus not in list, do not search</span>
            <span class="n">search_locus</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="c1"># check, if ...</span>
    <span class="c1"># ----------------------------------------------</span>
    <span class="c1"># 1) ... organism (= locus tag start) is in KEGG</span>
    <span class="c1"># ----------------------------------------------</span>
    <span class="k">if</span> <span class="n">search_locus</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># search locus tag in KEGG</span>
            <span class="n">request</span> <span class="o">=</span> <span class="n">REST</span><span class="o">.</span><span class="n">kegg_get</span><span class="p">(</span><span class="n">organism_abbrv</span><span class="o">+</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;locus_tag_ref&#39;</span><span class="p">])</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">request</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;ORTHOLOGY&#39;</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;\[EC:(.*?)\]&#39;</span><span class="p">,</span> <span class="n">line</span><span class="p">):</span>
                        <span class="c1"># @TODO</span>
                        <span class="c1"># will this result in problems with multiple EC numbers ?</span>
                        <span class="n">ec</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;\[EC:(.*?)\]&#39;</span><span class="p">,</span><span class="n">line</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                        <span class="n">row</span><span class="p">[</span><span class="s1">&#39;EC number&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ec</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                        <span class="n">row</span><span class="p">[</span><span class="s1">&#39;KEGG.notes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;updated EC with KEGG: True&#39;</span>
        <span class="k">except</span> <span class="n">urllib</span><span class="o">.</span><span class="n">error</span><span class="o">.</span><span class="n">HTTPError</span><span class="p">:</span>
            <span class="c1"># since locus tag was not found, try ...</span>
            <span class="c1"># 1. : the old one (if it exists)</span>
            <span class="n">found</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="n">not_found_log</span> <span class="o">=</span> <span class="sa">F</span><span class="s1">&#39;HTTPError: </span><span class="si">{</span><span class="n">organism_abbrv</span><span class="o">+</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;locus_tag_ref&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span>
            <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;old_locus_tag&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;-&#39;</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;old_locus_tag&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;locus_tag_ref&#39;</span><span class="p">]:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">request</span> <span class="o">=</span> <span class="n">REST</span><span class="o">.</span><span class="n">kegg_get</span><span class="p">(</span><span class="n">organism_abbrv</span><span class="o">+</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;old_locus_tag&#39;</span><span class="p">])</span>
                    <span class="n">found</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">request</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">):</span>
                        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;ORTHOLOGY&#39;</span><span class="p">):</span>
                            <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;\[EC:(.*?)\]&#39;</span><span class="p">,</span> <span class="n">line</span><span class="p">):</span>
                                <span class="n">ec</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;\[EC:(.*?)\]&#39;</span><span class="p">,</span><span class="n">line</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                                <span class="n">row</span><span class="p">[</span><span class="s1">&#39;EC number&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ec</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                                <span class="n">row</span><span class="p">[</span><span class="s1">&#39;KEGG.notes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;updated EC with KEGG: True&#39;</span>
                <span class="k">except</span> <span class="n">urllib</span><span class="o">.</span><span class="n">error</span><span class="o">.</span><span class="n">HTTPError</span><span class="p">:</span>
                    <span class="c1"># if there is still a &#39;not found&#39; error:</span>
                    <span class="n">not_found_log</span> <span class="o">+=</span> <span class="sa">F</span><span class="s1">&#39;, </span><span class="si">{</span><span class="n">organism_abbrv</span><span class="o">+</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;old_locus_tag&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span>
                <span class="k">except</span> <span class="ne">ConnectionResetError</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;ConnectionResetError: </span><span class="si">{</span><span class="n">organism_abbrv</span><span class="o">+</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;old_locus_tag&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
                <span class="k">except</span> <span class="n">urllib</span><span class="o">.</span><span class="n">error</span><span class="o">.</span><span class="n">URLError</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;URLError: </span><span class="si">{</span><span class="n">organism_abbrv</span><span class="o">+</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;old_locus_tag&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="c1"># 2. : check if RS after _, remove it and try again</span>
            <span class="k">if</span> <span class="s1">&#39;_RS&#39;</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;locus_tag_ref&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">found</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;old_locus_tag&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;locus_tag_ref&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_RS&#39;</span><span class="p">,</span><span class="s1">&#39;_&#39;</span><span class="p">):</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">edited_tag</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;locus_tag_ref&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_RS&#39;</span><span class="p">,</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>
                    <span class="n">request</span> <span class="o">=</span> <span class="n">REST</span><span class="o">.</span><span class="n">kegg_get</span><span class="p">(</span><span class="n">organism_abbrv</span><span class="o">+</span><span class="n">edited_tag</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">request</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">):</span>
                        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;ORTHOLOGY&#39;</span><span class="p">):</span>
                            <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;\[EC:(.*?)\]&#39;</span><span class="p">,</span> <span class="n">line</span><span class="p">):</span>
                                <span class="n">ec</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;\[EC:(.*?)\]&#39;</span><span class="p">,</span><span class="n">line</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                                <span class="n">row</span><span class="p">[</span><span class="s1">&#39;EC number&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ec</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                                <span class="n">row</span><span class="p">[</span><span class="s1">&#39;KEGG.notes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;updated EC with KEGG: True&#39;</span>
                    <span class="n">found</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">except</span> <span class="n">urllib</span><span class="o">.</span><span class="n">error</span><span class="o">.</span><span class="n">HTTPError</span><span class="p">:</span>
                    <span class="c1"># if there is still a &#39;not found&#39; error:</span>
                    <span class="n">not_found_log</span> <span class="o">+=</span> <span class="sa">F</span><span class="s1">&#39;, </span><span class="si">{</span><span class="n">organism_abbrv</span><span class="o">+</span><span class="n">edited_tag</span><span class="si">}</span><span class="s1">&#39;</span>
                <span class="k">except</span> <span class="ne">ConnectionResetError</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;ConnectionResetError: </span><span class="si">{</span><span class="n">organism_abbrv</span><span class="o">+</span><span class="n">edited_tag</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
                <span class="k">except</span> <span class="n">urllib</span><span class="o">.</span><span class="n">error</span><span class="o">.</span><span class="n">URLError</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;URLError: </span><span class="si">{</span><span class="n">organism_abbrv</span><span class="o">+</span><span class="n">edited_tag</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">found</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">not_found_log</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">ConnectionResetError</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;ConnectionResetError: </span><span class="si">{</span><span class="n">organism_abbrv</span><span class="o">+</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;locus_tag_ref&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">except</span> <span class="n">urllib</span><span class="o">.</span><span class="n">error</span><span class="o">.</span><span class="n">URLError</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;URLError: </span><span class="si">{</span><span class="n">organism_abbrv</span><span class="o">+</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;locus_tag_ref&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># -------------------------------------------------</span>
    <span class="c1"># 2) ... EC number is in KEGG and leads to reaction</span>
    <span class="c1"># -------------------------------------------------</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="s1">&#39;-&#39;</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;EC number&#39;</span><span class="p">]:</span>
        <span class="c1"># .......................................</span>
        <span class="c1"># @TODO</span>
        <span class="c1">#    how to handle multiple EC numbers ?</span>
        <span class="c1">#    idea: split row but how ?</span>
        <span class="c1"># .......................................</span>
        <span class="k">if</span> <span class="s1">&#39; &#39;</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;EC number&#39;</span><span class="p">]:</span>
            <span class="n">row</span><span class="p">[</span><span class="s1">&#39;EC number&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;EC number&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;NOTE: Dropped EC numbers </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;EC number&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]</span><span class="si">}</span><span class="s1"> for </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;locus_tag&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">.&#39;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">request</span> <span class="o">=</span> <span class="n">REST</span><span class="o">.</span><span class="n">kegg_get</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;ec:</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;EC number&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">record</span> <span class="o">=</span> <span class="n">Enzyme</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">request</span><span class="p">)</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">reaction</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">record</span><span class="o">.</span><span class="n">reaction</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
                <span class="c1"># no reaction associated</span>
                <span class="k">return</span> <span class="n">row</span>

            <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">reaction</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="c1"># exactly one reaction (block) ...</span>
                <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;\[RN:(.*?)\]&#39;</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">reaction</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
                    <span class="c1"># ... with an ID in KEGG</span>
                    <span class="n">row</span><span class="p">[</span><span class="s1">&#39;KEGG.reaction&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;\[RN:(.*?)\]&#39;</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">reaction</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># multiple reaction(s) blocks found, no clear &#39;answer&#39;, which is the correct one</span>
                <span class="c1"># ......................</span>
                <span class="c1"># @TODO</span>
                <span class="c1"># current solution: skip</span>
                <span class="c1"># ......................</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;Warning: multiple entries found for </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;EC number&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">, assignment skipped&#39;</span><span class="p">)</span>
        <span class="k">except</span> <span class="n">urllib</span><span class="o">.</span><span class="n">error</span><span class="o">.</span><span class="n">HTTPError</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;HTTPError: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;EC number&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">ConnectionResetError</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;ConnectionResetError: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;EC number&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">except</span> <span class="n">urllib</span><span class="o">.</span><span class="n">error</span><span class="o">.</span><span class="n">URLError</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;URLError: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;EC number&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">row</span></div>



<div class="viewcode-block" id="map_to_KEGG">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.map_to_KEGG">[docs]</a>
<span class="k">def</span> <span class="nf">map_to_KEGG</span><span class="p">(</span><span class="n">gene_table</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span><span class="n">working_dir</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span>
                <span class="n">manual_dir</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span><span class="n">genome_summary</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Path</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Map entries of the table from :py:func:`~specimen.core.refinement.extension.get_ncbi_info` to KEGG.</span>
<span class="sd">    Get KEGG.reaction ID from locus tag, if available, otherwise try EC number.</span>
<span class="sd">    Entries of the finalised table are saved depending whether they were assigned</span>
<span class="sd">    a KEGG.reaction ID (further use) or not (possible starting point for manual curation).</span>

<span class="sd">    Args:</span>
<span class="sd">        - gene_table (pd.DataFrame): </span>
<span class="sd">            The input table containing the output of get_ncbi_info().</span>
<span class="sd">        - working_dir (str): </span>
<span class="sd">            Path to the working directory - place to save genes with a KEGG.reaction ID.</span>
<span class="sd">        - manual_dir (str): </span>
<span class="sd">            Path to a directory to save the genes without KEGG assignment for possible manual curation.</span>
<span class="sd">        - genome_summary (str, optional): </span>
<span class="sd">            Path to a information file for the reference genomes to check if they are in KEGG. </span>
<span class="sd">            Defaults to None.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Path: </span>
<span class="sd">            The path to the file containing the genes with KEGG assignment.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">genome_summary</span><span class="p">:</span>
        <span class="c1"># identify locus tags that belong to genomes,</span>
        <span class="c1"># that are listed in KEGG</span>
        <span class="n">genome_info</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">genome_summary</span><span class="p">)</span>
        <span class="c1"># loci_in_KEGG = genome_info.loc[genome_info[&#39;KEGG.organism&#39;].notna()][&#39;locus_tag (start)&#39;].tolist()</span>
        <span class="n">loci_in_KEGG</span> <span class="o">=</span> <span class="n">genome_info</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">genome_info</span><span class="p">[</span><span class="s1">&#39;KEGG.organism&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">notna</span><span class="p">()][[</span><span class="s1">&#39;locus_tag (start)&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG.organism&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;locus_tag (start)&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()[</span><span class="s1">&#39;KEGG.organism&#39;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">loci_in_KEGG</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># map to KEGG</span>
    <span class="n">gene_table</span><span class="p">[</span><span class="s1">&#39;KEGG.reaction&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>
    <span class="n">gene_table</span><span class="p">[</span><span class="s1">&#39;KEGG.notes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;updated EC with KEGG: False&#39;</span>
    <span class="n">gene_table</span> <span class="o">=</span> <span class="n">gene_table</span><span class="o">.</span><span class="n">progress_apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">map_to_KEGG_row</span><span class="p">(</span><span class="n">row</span><span class="p">,</span><span class="n">loci_in_KEGG</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># save information</span>
    <span class="c1"># with KEGG.reaction for adding to model</span>
    <span class="c1"># .....................................................................</span>
    <span class="c1"># @TODO</span>
    <span class="c1">#    is this filter too hard? only EC is a bit risky</span>
    <span class="c1">#    how can one get more hits ?</span>
    <span class="c1">#    -&gt; best idea would probably be get more EC numbers to check</span>
    <span class="c1"># .....................................................................</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t\t</span><span class="s1">genes mapped to KEGG + EC: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_table</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene_table</span><span class="p">[</span><span class="s2">&quot;KEGG.reaction&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">notna</span><span class="p">()])</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">gene_table</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene_table</span><span class="p">[</span><span class="s1">&#39;KEGG.reaction&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">notna</span><span class="p">()]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">working_dir</span><span class="p">,</span><span class="s1">&#39;genes_mapped.csv&#39;</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># without for possible manual curation</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t\t</span><span class="s1">genes NOT mapped to KEGG + EC: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_table</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene_table</span><span class="p">[</span><span class="s2">&quot;KEGG.reaction&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()])</span><span class="si">}</span><span class="s1">, saved for manual curation&#39;</span><span class="p">)</span>
    <span class="n">gene_table</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene_table</span><span class="p">[</span><span class="s1">&#39;KEGG.reaction&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">manual_dir</span><span class="p">,</span><span class="s1">&#39;genes_no_reaction.csv&#39;</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">Path</span><span class="p">(</span><span class="n">working_dir</span><span class="p">,</span><span class="s1">&#39;genes_mapped.csv&#39;</span><span class="p">)</span></div>



<span class="c1"># mapping to BiGG</span>
<span class="c1"># ---------------</span>

<span class="c1"># @TODO</span>
<div class="viewcode-block" id="map_BiGG_reactions_row">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.map_BiGG_reactions_row">[docs]</a>
<span class="k">def</span> <span class="nf">map_BiGG_reactions_row</span><span class="p">(</span><span class="n">row</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span> <span class="n">namespace</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Map a single entry from the table in :py:func:`~specimen.core.refinement.extension.map_BiGG_reactions` to the BiGG reaction namespace.</span>

<span class="sd">    @TODO</span>

<span class="sd">    NOTE: </span>

<span class="sd">        only works for cases, where KEGG.reaction in row contains EXACTLY one entry</span>
<span class="sd">        in the rare case that multiple reactions belong to one enzyme, they are omitted</span>
<span class="sd">        in this search</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        - row (pd.Series): </span>
<span class="sd">            A single row of the table.</span>
<span class="sd">        - namespace (pd.DataFrame): </span>
<span class="sd">            The BiGG reaction namespace table.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.Series: </span>
<span class="sd">            The edited row.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="c1"># match by EC number AND KEGG id</span>
    <span class="n">matches</span> <span class="o">=</span> <span class="n">namespace</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">namespace</span><span class="p">[</span><span class="s1">&#39;EC Number&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;EC number&#39;</span><span class="p">])</span> <span class="o">&amp;</span> <span class="n">namespace</span><span class="p">[</span><span class="s1">&#39;KEGG Reaction&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;KEGG.reaction&#39;</span><span class="p">])]</span>

    <span class="c1"># case 1 : no matches</span>
    <span class="k">if</span> <span class="n">matches</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">row</span>

    <span class="c1"># case 2 : exactly one match</span>
    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">matches</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">row</span><span class="p">[</span><span class="s1">&#39;bigg_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">matches</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># case 3 : multiple matches</span>
    <span class="c1">#          often due to reaction being in different compartments</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">row</span><span class="p">[</span><span class="s1">&#39;bigg_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">matches</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">row</span></div>



<span class="c1"># @TEST : fitted to refinegems</span>
<span class="c1"># @CHECK : connections, e.g. input is now a param short </span>
<div class="viewcode-block" id="map_BiGG_reactions">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.map_BiGG_reactions">[docs]</a>
<span class="k">def</span> <span class="nf">map_BiGG_reactions</span><span class="p">(</span><span class="n">table_file</span><span class="p">:</span><span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Map the output of :py:func:`~specimen.core.refinement.extension.map_to_KEGG`</span>
<span class="sd">    to a BiGG namespace file (rewritten-type, see auxilliaries).</span>

<span class="sd">    Args:</span>
<span class="sd">        - table_file (str): </span>
<span class="sd">            The path to the saved table from running map_to_KEGG().</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: </span>
<span class="sd">            The table with an additional column for the mapping to BiGG reactions.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">r_namespace</span> <span class="o">=</span> <span class="n">load_a_table_from_database</span><span class="p">(</span><span class="s1">&#39;bigg_reactions&#39;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>

    <span class="n">table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">table_file</span><span class="p">)</span>
    <span class="n">table</span><span class="p">[</span><span class="s1">&#39;bigg_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>

    <span class="n">table</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">map_BiGG_reactions_row</span><span class="p">(</span><span class="n">row</span><span class="p">,</span><span class="n">r_namespace</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">table</span></div>


<span class="c1"># ------------</span>
<span class="c1"># extent model</span>
<span class="c1"># ------------</span>

<span class="c1"># @TODO</span>
<div class="viewcode-block" id="isreaction_complete">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.isreaction_complete">[docs]</a>
<span class="k">def</span> <span class="nf">isreaction_complete</span><span class="p">(</span><span class="n">reac</span><span class="p">:</span><span class="n">cobra</span><span class="o">.</span><span class="n">Reaction</span><span class="p">,</span> 
                        <span class="n">exclude_dna</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exclude_rna</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check, if a reaction is complete and ready to be added to the model.</span>
<span class="sd">    Additionally, it is possible to check for DNA and RNA reations</span>
<span class="sd">    and set them to be excluded or included.</span>

<span class="sd">    Args:</span>
<span class="sd">        - reac (cobra.Reaction): </span>
<span class="sd">            The reaction to be checked.</span>
<span class="sd">        - exclude_dna (bool, optional): </span>
<span class="sd">            Tag to include or exclude DNA reactions. </span>
<span class="sd">            Defaults to True.</span>
<span class="sd">        - exclude_rna (bool, optional): </span>
<span class="sd">            Tag to include or exclude RNA reactions. </span>
<span class="sd">            Defaults to True.</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool: </span>
<span class="sd">            True if the check is successful, else false.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># ................</span>
    <span class="c1"># @TODO</span>
    <span class="c1"># extendable</span>
    <span class="c1"># ................</span>

    <span class="c1"># check reaction</span>
    <span class="k">if</span> <span class="n">exclude_dna</span> <span class="ow">and</span> <span class="s1">&#39;DNA&#39;</span> <span class="ow">in</span> <span class="n">reac</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span>
    <span class="k">if</span> <span class="n">exclude_rna</span> <span class="ow">and</span> <span class="s1">&#39;RNA&#39;</span> <span class="ow">in</span> <span class="n">reac</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span>

    <span class="c1"># check metabolites</span>
    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">reac</span><span class="o">.</span><span class="n">metabolites</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span> <span class="ow">or</span> <span class="n">pd</span><span class="o">.</span><span class="n">isnull</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">id</span><span class="p">):</span>
            <span class="k">return</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span> <span class="ow">or</span> <span class="n">pd</span><span class="o">.</span><span class="n">isnull</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">name</span><span class="p">):</span>
            <span class="k">return</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">formula</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span> <span class="ow">or</span> <span class="n">pd</span><span class="o">.</span><span class="n">isnull</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">formula</span><span class="p">):</span>
            <span class="k">return</span> <span class="kc">False</span>

    <span class="k">return</span> <span class="kc">True</span></div>



<span class="c1"># build genes</span>
<span class="c1"># -----------</span>

<div class="viewcode-block" id="add_gene">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.add_gene">[docs]</a>
<span class="k">def</span> <span class="nf">add_gene</span><span class="p">(</span><span class="n">model</span><span class="p">:</span><span class="n">cobra</span><span class="o">.</span><span class="n">Model</span><span class="p">,</span> <span class="n">reaction</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">row</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span> 
             <span class="n">first</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">cobra</span><span class="o">.</span><span class="n">Model</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add a new gene to a genome-scale metabolic cobra model.</span>

<span class="sd">    Args:</span>
<span class="sd">        - model (cobra.Model): </span>
<span class="sd">            The model.</span>
<span class="sd">        - reaction (str): </span>
<span class="sd">            The reaction id to add the gene to.</span>
<span class="sd">        - row (pd.Series): </span>
<span class="sd">            A single row of the output table of :py:func:`~specimen.core.refinement.extension.map_BiGG_reactions`.</span>
<span class="sd">        - first (bool, optional): </span>
<span class="sd">            Shows, if gene is the first gene to be added to the reaction. </span>
<span class="sd">            True if gene is first to be added.</span>
<span class="sd">            Defaults to False.</span>

<span class="sd">    Returns:</span>
<span class="sd">        cobra.Model: </span>
<span class="sd">            Updated model.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># add gene</span>
    <span class="k">if</span> <span class="n">first</span> <span class="ow">or</span> <span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">reaction</span><span class="p">)</span><span class="o">.</span><span class="n">gene_reaction_rule</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
        <span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">reaction</span><span class="p">)</span><span class="o">.</span><span class="n">gene_reaction_rule</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">reaction</span><span class="p">)</span><span class="o">.</span><span class="n">gene_reaction_rule</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">reaction</span><span class="p">)</span><span class="o">.</span><span class="n">gene_reaction_rule</span> <span class="o">+</span> <span class="s1">&#39; or &#39;</span> <span class="o">+</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># add name</span>
    <span class="n">model</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># add annotations</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">pd</span><span class="o">.</span><span class="n">isnull</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">4</span><span class="p">]):</span>
        <span class="n">model</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;ncbigene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
    <span class="n">model</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;ncbiprotein&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1"># note: annotations like sbo, kegg.genes and uniprot missing</span>

    <span class="k">return</span> <span class="n">model</span></div>



<span class="c1"># build metabolites</span>
<span class="c1"># -----------------</span>
<span class="c1"># @TODO</span>
<span class="c1"># UNDER CONSTRUCTION</span>
<div class="viewcode-block" id="build_metabolite_mnx">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.build_metabolite_mnx">[docs]</a>
<span class="k">def</span> <span class="nf">build_metabolite_mnx</span><span class="p">(</span><span class="n">metabolite</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span><span class="n">cobra</span><span class="o">.</span><span class="n">Model</span><span class="p">,</span> 
                         <span class="n">mnx_chem_prop</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">mnx_chem_xref</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> 
                         <span class="n">bigg_metabolites</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">namespace</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">cobra</span><span class="o">.</span><span class="n">Metabolite</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create or retrieve (from model) a metabolite based on its MetaNetX ID.</span>

<span class="sd">    Args:</span>
<span class="sd">        - metabolite (str): </span>
<span class="sd">            The MetaNetX ID of the metabolite.</span>
<span class="sd">        - model (cobra.Model):  </span>
<span class="sd">            The underlying genome-scale metabolic model.</span>
<span class="sd">        - mnx_chem_prop (pd.DataFrame): </span>
<span class="sd">            The chem_xref table from MetaNetX.</span>
<span class="sd">        - mnx_chem_xref (pd.DataFrame): </span>
<span class="sd">            The chem_prop table from MetaNetX.</span>
<span class="sd">        - bigg_metabolites (pd.DataFrame): </span>
<span class="sd">            The BiGG compound namespace table.</span>
<span class="sd">        - namespace (pd.DataFrame): </span>
<span class="sd">            The namespace of the model.</span>

<span class="sd">    Returns:</span>
<span class="sd">        cobra.Metabolite: </span>
<span class="sd">            The retrieved or newly build metabolite.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">metabolite_prop</span> <span class="o">=</span> <span class="n">mnx_chem_prop</span><span class="p">[</span><span class="n">mnx_chem_prop</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">==</span><span class="n">metabolite</span><span class="p">]</span>
    <span class="n">metabolite_anno</span> <span class="o">=</span> <span class="n">mnx_chem_xref</span><span class="p">[</span><span class="n">mnx_chem_xref</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">==</span><span class="n">metabolite</span><span class="p">]</span>
    <span class="n">model_mnx</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;metanetx.chemical&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">metabolites</span> <span class="k">if</span> <span class="s1">&#39;metanetx.chemical&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">annotation</span><span class="p">]</span>

    <span class="c1"># fast check if compound already in model</span>
    <span class="c1"># ------------------------------------------</span>
    <span class="c1"># @TODO ..........................................</span>
        <span class="c1">#   currently no checking for compartments</span>
        <span class="c1">#   first match will be taken (most often cytosol one)</span>
        <span class="c1">#   regardless of the compartment</span>
        <span class="c1">#.............................................</span>
    <span class="c1"># step 1: check if MetaNetX ID in model</span>
    <span class="k">if</span> <span class="n">metabolite</span> <span class="ow">in</span> <span class="n">model_mnx</span><span class="p">:</span>
        <span class="n">matches</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">metabolites</span> <span class="k">if</span> <span class="s1">&#39;metanetx.chemical&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">annotation</span> <span class="ow">and</span> <span class="n">x</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;metanetx.chemical&#39;</span><span class="p">]</span><span class="o">==</span><span class="n">metabolite</span><span class="p">]</span>

    <span class="c1"># step 2: if yes, retrieve metabolite from model</span>
    <span class="c1">#  case 1: multiple matches found</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">matches</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="c1"># ................</span>
            <span class="c1"># @TODO see above</span>
            <span class="c1"># ................</span>
            <span class="n">match</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">metabolites</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">matches</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="c1">#  case 2: only one match found</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">match</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">metabolites</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">matches</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

        <span class="c1"># step 3: add metabolite</span>
        <span class="k">return</span> <span class="n">match</span>

    <span class="c1"># if not, create new metabolite</span>
    <span class="c1"># -----------------------------</span>
    <span class="k">else</span><span class="p">:</span>

        <span class="c1"># step 1: create a random metabolite ID</span>
        <span class="c1"># ...........................</span>
        <span class="c1"># @TODO : compartment problem </span>
        <span class="c1"># - does it have to be in the name?</span>
        <span class="c1"># ...........................</span>
        <span class="n">new_metabolite</span> <span class="o">=</span> <span class="n">cobra</span><span class="o">.</span><span class="n">Metabolite</span><span class="p">(</span><span class="n">create_random_id</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="s1">&#39;meta&#39;</span><span class="p">,</span><span class="s1">&#39;SPECIMEN&#39;</span><span class="p">))</span> 


        <span class="c1"># step 2: add features</span>
        <span class="c1"># --------------------</span>
        <span class="c1"># @TODO ..........................................</span>
        <span class="c1">#   currently no checking for compartments</span>
        <span class="c1">#   defaults to c</span>
        <span class="c1">#   makes it difficult to add exchange reactions</span>
        <span class="c1">#.................................................</span>
        <span class="n">new_metabolite</span><span class="o">.</span><span class="n">formula</span> <span class="o">=</span> <span class="n">metabolite_prop</span><span class="p">[</span><span class="s1">&#39;formula&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">new_metabolite</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">metabolite_prop</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">new_metabolite</span><span class="o">.</span><span class="n">charge</span> <span class="o">=</span> <span class="n">metabolite_prop</span><span class="p">[</span><span class="s1">&#39;charge&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">new_metabolite</span><span class="o">.</span><span class="n">compartment</span> <span class="o">=</span> <span class="s1">&#39;c&#39;</span>

        <span class="c1"># step 3: add notes</span>
        <span class="c1"># -----------------</span>
        <span class="n">new_metabolite</span><span class="o">.</span><span class="n">notes</span><span class="p">[</span><span class="s1">&#39;added via&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;metanetx.chemical&#39;</span>

        <span class="c1"># step 4: add annotations</span>
        <span class="c1"># -----------------------</span>
        <span class="n">new_metabolite</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;metanetx.chemical&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metabolite_prop</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">new_metabolite</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;chebi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metabolite_prop</span><span class="p">[</span><span class="s1">&#39;reference&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">pd</span><span class="o">.</span><span class="n">isnull</span><span class="p">(</span><span class="n">metabolite_prop</span><span class="p">[</span><span class="s1">&#39;InChIKey&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
            <span class="n">new_metabolite</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;inchikey&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metabolite_prop</span><span class="p">[</span><span class="s1">&#39;InChIKey&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;=&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">db</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;kegg.compound&#39;</span><span class="p">,</span><span class="s1">&#39;metacyc.compound&#39;</span><span class="p">,</span><span class="s1">&#39;seed.compound&#39;</span><span class="p">,</span><span class="s1">&#39;bigg.metabolite&#39;</span><span class="p">]:</span>
            <span class="n">db_matches</span> <span class="o">=</span> <span class="n">metabolite_anno</span><span class="p">[</span><span class="n">metabolite_anno</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">db</span><span class="p">)]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">db_matches</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                 <span class="n">new_metabolite</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">db</span><span class="p">]</span> <span class="o">=</span> <span class="n">db_matches</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">db_matches</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">new_metabolite</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">db</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">m</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">db_matches</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span>

        <span class="c1"># if no BiGG was found in MetaNetX, try reverse search in BiGG</span>
        <span class="k">if</span> <span class="n">metabolite</span> <span class="ow">in</span> <span class="n">bigg_metabolites</span><span class="p">[</span><span class="s1">&#39;MetaNetX (MNX) Chemical&#39;</span><span class="p">]:</span>
            <span class="n">new_metabolite</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;bigg.metabolite&#39;</span><span class="p">]</span> <span class="o">=</span>  <span class="n">bigg_metabolites</span><span class="p">[</span><span class="n">bigg_metabolites</span><span class="p">[</span><span class="s1">&#39;MetaNetX (MNX) Chemical&#39;</span><span class="p">]</span><span class="o">==</span><span class="n">metabolite</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        
        <span class="c1"># add additional information from bigg if possible    </span>
        <span class="k">if</span> <span class="s1">&#39;bigg.metabolite&#39;</span> <span class="ow">in</span> <span class="n">new_metabolite</span><span class="o">.</span><span class="n">annotation</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">bigg_information</span> <span class="o">=</span> <span class="n">bigg_metabolites</span><span class="p">[</span><span class="n">bigg_metabolites</span><span class="p">[</span><span class="s1">&#39;bigg_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;|&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">new_metabolite</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;bigg.metabolite&#39;</span><span class="p">]))]</span>
            <span class="n">db_id_bigg</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;BioCyc&#39;</span><span class="p">:</span><span class="s1">&#39;biocyc&#39;</span><span class="p">,</span> <span class="s1">&#39;MetaNetX (MNX) Chemical&#39;</span><span class="p">:</span><span class="s1">&#39;metanetx.chemical&#39;</span><span class="p">,</span><span class="s1">&#39;SEED Compound&#39;</span><span class="p">:</span><span class="s1">&#39;seed.compound&#39;</span><span class="p">,</span><span class="s1">&#39;InChI Key&#39;</span><span class="p">:</span><span class="s1">&#39;inchikey&#39;</span><span class="p">}</span>
            <span class="k">for</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">db_id_bigg</span><span class="p">:</span>
                <span class="n">info</span> <span class="o">=</span> <span class="n">bigg_information</span><span class="p">[</span><span class="n">db</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">info</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">info</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">info</span><span class="p">)</span>
                    <span class="n">info</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">info</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)]</span> <span class="c1"># make sure all entries are a separate list object</span>
                    <span class="n">new_metabolite</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">db_id_bigg</span><span class="p">[</span><span class="n">db</span><span class="p">]]</span> <span class="o">=</span> <span class="n">info</span>

        <span class="c1"># step 5: change ID according to namespace</span>
        <span class="c1"># ----------------------------------------</span>
        <span class="n">match_id_to_namespace</span><span class="p">(</span><span class="n">new_metabolite</span><span class="p">,</span><span class="n">namespace</span><span class="p">)</span>
       
        <span class="c1"># step 6: re-check existence of ID in model</span>
        <span class="c1"># -----------------------------------------</span>
        <span class="c1"># @TODO : check complete annotations? </span>
        <span class="c1">#        - or let those be covered by the duplicate check later on?</span>
        <span class="k">if</span> <span class="n">new_metabolite</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="p">[</span><span class="n">_</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">metabolites</span><span class="p">]:</span>
            <span class="k">return</span> <span class="n">model</span><span class="o">.</span><span class="n">metabolites</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">new_metabolite</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
        
    <span class="k">return</span> <span class="n">new_metabolite</span></div>


<span class="c1"># @TODO</span>
<span class="c1"># UNDER CONSTRUCTION</span>
<div class="viewcode-block" id="build_metabolite_kegg">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.build_metabolite_kegg">[docs]</a>
<span class="k">def</span> <span class="nf">build_metabolite_kegg</span><span class="p">(</span><span class="n">kegg_id</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span><span class="n">cobra</span><span class="o">.</span><span class="n">Model</span><span class="p">,</span> 
                          <span class="n">model_kegg_ids</span><span class="p">:</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">bigg_metabolites</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> 
                          <span class="n">namespace</span><span class="p">:</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;BiGG&#39;</span><span class="p">]</span><span class="o">=</span><span class="s1">&#39;BiGG&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">cobra</span><span class="o">.</span><span class="n">Metabolite</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create or retrieve (from model) a metabolite based on its KEGG ID.</span>

<span class="sd">    Args:</span>
<span class="sd">        - kegg_id (str): </span>
<span class="sd">            The KEGG.compound ID of the metabolite in question.</span>
<span class="sd">        - model (cobra.Model): </span>
<span class="sd">            The model.</span>
<span class="sd">        - model_kegg_ids (list[str]): </span>
<span class="sd">            List of all annotated KEGG Compound IDs in the model.</span>
<span class="sd">        - bigg_metabolites (pd.DataFrame):  </span>
<span class="sd">            The BiGG compound namespace table.</span>
<span class="sd">        - namespace (Literal[&#39;BiGG&#39;], optional): </span>
<span class="sd">            Namespace of the model. Defaults to &#39;BiGG&#39;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        cobra.Metabolite:  </span>
<span class="sd">            The retrieved or newly build metabolite.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># retrieve KEGG entry for compound</span>
    <span class="c1"># --------------------------------</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">kegg_handle</span> <span class="o">=</span> <span class="n">REST</span><span class="o">.</span><span class="n">kegg_get</span><span class="p">(</span><span class="n">kegg_id</span><span class="p">)</span>
        <span class="n">kegg_record</span> <span class="o">=</span> <span class="p">[</span><span class="n">r</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">Compound</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">kegg_handle</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">except</span> <span class="n">urllib</span><span class="o">.</span><span class="n">error</span><span class="o">.</span><span class="n">HTTPError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;HTTPError: </span><span class="si">{</span><span class="n">kegg_id</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">cobra</span><span class="o">.</span><span class="n">Metabolite</span><span class="p">()</span>
    <span class="k">except</span> <span class="ne">ConnectionResetError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;ConnectionResetError: </span><span class="si">{</span><span class="n">kegg_id</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">cobra</span><span class="o">.</span><span class="n">Metabolite</span><span class="p">()</span>
    <span class="k">except</span> <span class="n">urllib</span><span class="o">.</span><span class="n">error</span><span class="o">.</span><span class="n">URLError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;URLError: </span><span class="si">{</span><span class="n">kegg_id</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">cobra</span><span class="o">.</span><span class="n">Metabolite</span><span class="p">()</span>

    <span class="c1"># ---------------------------------------</span>
    <span class="c1"># fast check if compound already in model</span>
    <span class="c1"># ---------------------------------------</span>
    <span class="c1"># @TODO ..........................................</span>
        <span class="c1">#   currently no checking for compartments</span>
        <span class="c1">#   first match will be taken (most often cytosol one)</span>
        <span class="c1">#   regardless of the compartment</span>
        <span class="c1">#.............................................</span>
    <span class="c1"># step 1: check via KEGG ID</span>
    <span class="k">if</span> <span class="n">kegg_id</span> <span class="ow">in</span> <span class="n">model_kegg_ids</span><span class="p">:</span>
        <span class="n">matches</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">metabolites</span> <span class="k">if</span> <span class="p">(</span><span class="s1">&#39;kegg.compound&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">annotation</span> <span class="ow">and</span> <span class="n">x</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;kegg.compound&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">kegg_id</span><span class="p">)]</span>

        <span class="c1"># step 2: model id --&gt; metabolite object</span>
        <span class="c1">#  case 1: multiple matches found</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">matches</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">match</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">metabolites</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">matches</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="c1">#  case 2: only one match found</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">match</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">metabolites</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">matches</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

        <span class="c1"># step 3: add metabolite</span>
        <span class="k">return</span> <span class="n">match</span>

    <span class="c1"># -----------------------------</span>
    <span class="c1"># if not, create new metabolite</span>
    <span class="c1"># -----------------------------</span>
    <span class="c1"># ...............</span>
    <span class="c1"># @TODO</span>
    <span class="c1">#     compartment</span>
    <span class="c1"># ...............</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># step 1: create a random metabolite ID</span>
        <span class="c1"># -------------------------------------</span>
        <span class="c1"># ...........................</span>
        <span class="c1"># @TODO : compartment problem </span>
        <span class="c1"># - does it have to be in the name?</span>
        <span class="c1"># ...........................</span>
        <span class="n">new_metabolite</span> <span class="o">=</span> <span class="n">cobra</span><span class="o">.</span><span class="n">Metabolite</span><span class="p">(</span><span class="n">create_random_id</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="s1">&#39;meta&#39;</span><span class="p">,</span><span class="s1">&#39;SPECIMEN&#39;</span><span class="p">))</span> 

        <span class="c1"># step 2: add features</span>
        <span class="c1"># --------------------</span>
        <span class="c1"># @TODO ..........................................</span>
        <span class="c1">#   currently no checking for compartments</span>
        <span class="c1">#.............................................</span>
        <span class="c1"># set name from KEGG and additionally use it as ID if there is none yet</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">kegg_record</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">kegg_record</span><span class="o">.</span><span class="n">name</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">new_metabolite</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">kegg_record</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">new_metabolite</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">kegg_record</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">new_metabolite</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">kegg_record</span><span class="o">.</span><span class="n">name</span>
        <span class="c1"># set compartment</span>
        <span class="n">new_metabolite</span><span class="o">.</span><span class="n">compartment</span> <span class="o">=</span> <span class="s1">&#39;c&#39;</span>
        <span class="c1"># set formula</span>
        <span class="n">new_metabolite</span><span class="o">.</span><span class="n">formula</span> <span class="o">=</span> <span class="n">kegg_record</span><span class="o">.</span><span class="n">formula</span>

        <span class="c1"># step 3: add notes</span>
        <span class="c1"># -----------------</span>
        <span class="n">new_metabolite</span><span class="o">.</span><span class="n">notes</span><span class="p">[</span><span class="s1">&#39;added via&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;KEGG.compound&#39;</span>

        <span class="c1"># step 4: add annotations</span>
        <span class="c1"># -----------------------</span>
        <span class="n">new_metabolite</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;kegg.compound&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kegg_id</span>
        <span class="n">db_idtf</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;CAS&#39;</span><span class="p">:</span><span class="s1">&#39;cas&#39;</span><span class="p">,</span><span class="s1">&#39;PubChem&#39;</span><span class="p">:</span><span class="s1">&#39;pubchem.compound&#39;</span><span class="p">,</span><span class="s1">&#39;ChEBI&#39;</span><span class="p">:</span><span class="s1">&#39;chebi&#39;</span><span class="p">}</span>
        <span class="k">for</span> <span class="n">db</span><span class="p">,</span><span class="n">ids</span> <span class="ow">in</span> <span class="n">kegg_record</span><span class="o">.</span><span class="n">dblinks</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">db_idtf</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ids</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">new_metabolite</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">db_idtf</span><span class="p">[</span><span class="n">db</span><span class="p">]]</span> <span class="o">=</span> <span class="n">ids</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">new_metabolite</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">db_idtf</span><span class="p">[</span><span class="n">db</span><span class="p">]]</span> <span class="o">=</span> <span class="n">ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="c1"># add additional information from BiGG</span>
        <span class="k">if</span> <span class="n">kegg_id</span> <span class="ow">in</span> <span class="n">bigg_metabolites</span><span class="p">[</span><span class="s1">&#39;KEGG Compound&#39;</span><span class="p">]:</span>

            <span class="n">bigg_information</span> <span class="o">=</span> <span class="n">bigg_metabolites</span><span class="p">[</span><span class="n">bigg_metabolites</span><span class="p">[</span><span class="s1">&#39;KEGG Compound&#39;</span><span class="p">]</span><span class="o">==</span><span class="n">kegg_id</span><span class="p">]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">bigg_information</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>

                <span class="n">new_metabolite</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;bigg.metabolite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bigg_information</span><span class="p">[</span><span class="s1">&#39;bigg_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>

                <span class="n">db_id_bigg</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;BioCyc&#39;</span><span class="p">:</span><span class="s1">&#39;biocyc&#39;</span><span class="p">,</span> <span class="s1">&#39;MetaNetX (MNX) Chemical&#39;</span><span class="p">:</span><span class="s1">&#39;metanetx.chemical&#39;</span><span class="p">,</span><span class="s1">&#39;SEED Compound&#39;</span><span class="p">:</span><span class="s1">&#39;seed.compound&#39;</span><span class="p">,</span><span class="s1">&#39;InChI Key&#39;</span><span class="p">:</span><span class="s1">&#39;inchikey&#39;</span><span class="p">}</span>
                <span class="k">for</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">db_id_bigg</span><span class="p">:</span>
                    <span class="n">info</span> <span class="o">=</span> <span class="n">bigg_information</span><span class="p">[</span><span class="n">db</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">info</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="n">info</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">info</span><span class="p">)</span>
                        <span class="n">info</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">info</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)]</span> <span class="c1"># make sure all entries are a separate list object</span>
                        <span class="n">new_metabolite</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">db_id_bigg</span><span class="p">[</span><span class="n">db</span><span class="p">]]</span> <span class="o">=</span> <span class="n">info</span>

        <span class="c1"># step 5: change ID according to namespace</span>
        <span class="c1"># ----------------------------------------</span>
        <span class="n">match_id_to_namespace</span><span class="p">(</span><span class="n">new_metabolite</span><span class="p">,</span><span class="n">namespace</span><span class="p">)</span>
       
        <span class="c1"># step 6: re-check existence of ID in model</span>
        <span class="c1"># -----------------------------------------</span>
        <span class="c1"># @TODO : check complete annotations? </span>
        <span class="c1">#        - or let those be covered by the duplicate check later on?</span>
        <span class="k">if</span> <span class="n">new_metabolite</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="p">[</span><span class="n">_</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">metabolites</span><span class="p">]:</span>
            <span class="k">return</span> <span class="n">model</span><span class="o">.</span><span class="n">metabolites</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">new_metabolite</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">new_metabolite</span></div>


<span class="c1"># @TODO</span>
<span class="c1"># UNDER CONSTRUCTION</span>
<div class="viewcode-block" id="get_metabolites_mnx">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.get_metabolites_mnx">[docs]</a>
<span class="k">def</span> <span class="nf">get_metabolites_mnx</span><span class="p">(</span><span class="n">model</span><span class="p">:</span><span class="n">cobra</span><span class="o">.</span><span class="n">Model</span><span class="p">,</span><span class="n">equation</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span>
                        <span class="n">mnx_chem_xref</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span><span class="n">mnx_chem_prop</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
                        <span class="n">bigg_metabolites</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">namespace</span><span class="p">:</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;BiGG&#39;</span><span class="p">]</span><span class="o">=</span><span class="s1">&#39;BiGG&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="n">cobra</span><span class="o">.</span><span class="n">Metabolite</span><span class="p">,</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Based on a given MetaNetX equation and a model, get or</span>
<span class="sd">    create metabolite entires in/for the model.</span>

<span class="sd">    Args:</span>
<span class="sd">        - model (cobra.Model): </span>
<span class="sd">            The GEM.</span>
<span class="sd">        - equation (str): </span>
<span class="sd">            The equation from MetaNetX.</span>
<span class="sd">        - mnx_chem_xref (pd.DataFrame):  </span>
<span class="sd">            The chem_xref table from MetaNetX.</span>
<span class="sd">        - mnx_chem_prop (pd.DataFrame):  </span>
<span class="sd">            The chem_prop table from MetaNetX.</span>
<span class="sd">        - bigg_metabolites (pd.DataFrame): </span>
<span class="sd">            The BiGG compound namespace table.</span>
<span class="sd">        - namespace (Literal[&#39;BiGG&#39;], optional): </span>
<span class="sd">            Namespace of the model. </span>
<span class="sd">            Defaults to &#39;BiGG&#39;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict[cobra.Metabolite,int]: </span>
<span class="sd">            Dictonary of metabolites and stoichiometric factors.</span>
<span class="sd">    &quot;&quot;&quot;</span>  

    <span class="c1"># @TODO ...................................</span>
    <span class="c1">#   currently no checking for compartments</span>
    <span class="c1">#..........................................</span>

    <span class="n">model_metabolites</span> <span class="o">=</span> <span class="p">[</span><span class="n">m</span><span class="o">.</span><span class="n">formula</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">metabolites</span><span class="p">]</span>
    <span class="n">metabolites</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">produced</span> <span class="o">=</span> <span class="o">-</span><span class="mf">1.0</span>
    <span class="n">factor</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">equation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">):</span>
        <span class="c1"># switch from reactants to products</span>
        <span class="k">if</span> <span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;=&#39;</span><span class="p">:</span>
            <span class="n">produced</span> <span class="o">=</span> <span class="mf">1.0</span>
        <span class="c1"># found stoichiometric factor</span>
        <span class="k">elif</span> <span class="n">s</span><span class="o">.</span><span class="n">isnumeric</span><span class="p">():</span>
            <span class="n">factor</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
        <span class="c1"># skip</span>
        <span class="k">elif</span> <span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="c1"># found metabolite</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># get information from MetaNetX</span>
            <span class="n">metabolite</span><span class="p">,</span> <span class="n">compartment</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;@&#39;</span><span class="p">)</span>
            <span class="c1"># build or identify metabolite</span>
            <span class="n">new_metabolite</span> <span class="o">=</span> <span class="n">build_metabolite_mnx</span><span class="p">(</span><span class="n">metabolite</span><span class="p">,</span> <span class="n">model</span><span class="p">,</span> <span class="n">mnx_chem_prop</span><span class="p">,</span> <span class="n">mnx_chem_xref</span><span class="p">,</span><span class="n">bigg_metabolites</span><span class="p">,</span> <span class="n">namespace</span><span class="p">)</span>
            <span class="c1"># add metabolite</span>
            <span class="k">if</span> <span class="n">new_metabolite</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="p">[</span><span class="n">_</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">metabolites</span><span class="p">]:</span>
                <span class="c1"># ......................................................</span>
                <span class="c1"># @TODO: </span>
                <span class="c1">#   check if metabolite if both reactant and product</span>
                <span class="c1">#   suggests exchange reaction </span>
                <span class="c1">#   -&gt; maybe a good place to change compartment for one?</span>
                <span class="c1">#   -&gt; what about name and directions???</span>
                <span class="c1"># ......................................................</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">test</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">metabolites</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">new_metabolite</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                    <span class="n">new_metabolite</span> <span class="o">=</span> <span class="n">new_metabolite</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
                    <span class="n">new_metabolite</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">new_metabolite</span><span class="o">.</span><span class="n">id</span> <span class="o">+</span> <span class="s1">&#39;_i&#39;</span>
                <span class="k">except</span><span class="p">:</span>
                    <span class="n">new_metabolite</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">new_metabolite</span><span class="o">.</span><span class="n">id</span> <span class="o">+</span> <span class="s1">&#39;_i&#39;</span>

            <span class="n">metabolites</span><span class="p">[</span><span class="n">new_metabolite</span><span class="p">]</span> <span class="o">=</span> <span class="n">factor</span> <span class="o">*</span> <span class="n">produced</span>

    <span class="k">return</span> <span class="n">metabolites</span></div>



<span class="c1"># @TODO</span>
<span class="c1"># UNDER CONSTRUCTION</span>
<div class="viewcode-block" id="get_metabolites_kegg">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.get_metabolites_kegg">[docs]</a>
<span class="k">def</span> <span class="nf">get_metabolites_kegg</span><span class="p">(</span><span class="n">model</span><span class="p">:</span><span class="n">cobra</span><span class="o">.</span><span class="n">Model</span><span class="p">,</span><span class="n">equation</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span>
                         <span class="n">chem_xref</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span><span class="n">chem_prop</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
                         <span class="n">bigg_metabolites</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> 
                         <span class="n">namespace</span><span class="p">:</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;BiGG&#39;</span><span class="p">]</span><span class="o">=</span><span class="s1">&#39;BiGG&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="n">cobra</span><span class="o">.</span><span class="n">Metabolite</span><span class="p">,</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Based on a given KEGG equation and a model, get or</span>
<span class="sd">    create metabolite entires in/for the model.</span>

<span class="sd">    Args:</span>
<span class="sd">        - model (cobra.Model): </span>
<span class="sd">            A GEM.</span>
<span class="sd">        - equation (str): </span>
<span class="sd">            The equation from KEGG.</span>
<span class="sd">        - chem_xref (pd.DataFrame): </span>
<span class="sd">            The chem_xref table from MetaNetX.</span>
<span class="sd">        - chem_prop (pd.DataFrame): </span>
<span class="sd">            The chem_prop table from MetaNetX.</span>
<span class="sd">        - bigg_metabolites (pd.DataFrame): </span>
<span class="sd">            The BiGG compound namespace table.</span>
<span class="sd">        - namespace (Literal[&#39;BiGG&#39;], optional): </span>
<span class="sd">            The namespace of the model. </span>
<span class="sd">            Defaults to &#39;BiGG&#39;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict[cobra.Metabolite,int]: </span>
<span class="sd">            Dictonary of metabolites and stoichiometric factors.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># @TODO ...................................</span>
    <span class="c1">#   currently no checking for compartments</span>
    <span class="c1">#..........................................</span>

    <span class="n">model_metabolites</span> <span class="o">=</span> <span class="p">[</span><span class="n">m</span><span class="o">.</span><span class="n">formula</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">metabolites</span><span class="p">]</span>
    <span class="n">model_kegg_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">m</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;kegg.compound&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">metabolites</span> <span class="k">if</span> <span class="s1">&#39;kegg.compound&#39;</span> <span class="ow">in</span> <span class="n">m</span><span class="o">.</span><span class="n">annotation</span><span class="p">]</span>
    <span class="n">metabolites</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">produced</span> <span class="o">=</span> <span class="o">-</span><span class="mf">1.0</span>
    <span class="n">factor</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">mnx_id</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>

    <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">equation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">):</span>
        <span class="c1"># switch from reactants to products</span>
        <span class="k">if</span> <span class="s1">&#39;=&#39;</span> <span class="ow">in</span> <span class="n">s</span><span class="p">:</span>
            <span class="n">produced</span> <span class="o">=</span> <span class="mf">1.0</span>
        <span class="c1"># found stoichiometric factor</span>
        <span class="k">elif</span> <span class="n">s</span><span class="o">.</span><span class="n">isnumeric</span><span class="p">():</span>
            <span class="n">factor</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
        <span class="c1"># skip</span>
        <span class="k">elif</span> <span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="c1"># found metabolite</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># check if s is a valid ID</span>
            <span class="k">if</span> <span class="s1">&#39;(&#39;</span> <span class="ow">in</span> <span class="n">s</span><span class="p">:</span>
                <span class="n">s</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="c1"># ..................................</span>
                <span class="c1"># @TODO</span>
                <span class="c1">#     known case: DNA(n) --&gt; DNA(n+1)</span>
                <span class="c1">#     currently note in brackets gets ignored</span>
                <span class="c1"># ..................................</span>
            <span class="k">elif</span> <span class="ow">not</span> <span class="n">s</span><span class="o">.</span><span class="n">isalnum</span><span class="p">():</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Problem: unknown character in ID inside get_metabolites_kegg() detected.</span><span class="se">\n</span><span class="s1">Please contact dev about your problem.&#39;</span><span class="p">)</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

            <span class="n">mnx_id</span> <span class="o">=</span> <span class="n">chem_xref</span><span class="p">[</span><span class="n">chem_xref</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="sa">F</span><span class="s1">&#39;kegg.compound:</span><span class="si">{</span><span class="n">s</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span>
            <span class="c1"># case 1:</span>
            <span class="c1">#     add metabolite via MetaNetX</span>
            <span class="c1">#     -&gt; make sure, only 1 ID match is found (match is unambiguous)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mnx_id</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">mnx_id</span> <span class="o">=</span> <span class="n">mnx_id</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
                <span class="n">metabolite</span> <span class="o">=</span> <span class="n">build_metabolite_mnx</span><span class="p">(</span><span class="n">mnx_id</span><span class="p">,</span> <span class="n">model</span><span class="p">,</span> <span class="n">chem_prop</span><span class="p">,</span> <span class="n">chem_xref</span><span class="p">,</span> <span class="n">bigg_metabolites</span><span class="p">,</span> <span class="n">namespace</span><span class="p">)</span>
            <span class="c1"># case 2:</span>
            <span class="c1">#     add metabolite via KEGG</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">metabolite</span> <span class="o">=</span> <span class="n">build_metabolite_kegg</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">model</span><span class="p">,</span> <span class="n">model_kegg_ids</span><span class="p">,</span> <span class="n">bigg_metabolites</span><span class="p">,</span> <span class="n">namespace</span><span class="p">)</span>

            <span class="c1"># add new metabolite</span>
            <span class="c1"># @TODO : place to check for exchanges?</span>
            <span class="k">if</span> <span class="n">metabolite</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="p">[</span><span class="n">_</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">metabolites</span><span class="p">]:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">test</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">metabolites</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">metabolite</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                    <span class="n">metabolite</span> <span class="o">=</span> <span class="n">metabolite</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
                    <span class="n">metabolite</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">metabolite</span><span class="o">.</span><span class="n">id</span> <span class="o">+</span> <span class="s1">&#39;_i&#39;</span>
                <span class="k">except</span><span class="p">:</span>
                    <span class="n">metabolite</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">metabolite</span><span class="o">.</span><span class="n">id</span> <span class="o">+</span> <span class="s1">&#39;_i&#39;</span>

            <span class="n">metabolites</span><span class="p">[</span><span class="n">metabolite</span><span class="p">]</span> <span class="o">=</span> <span class="n">factor</span> <span class="o">*</span> <span class="n">produced</span>

    <span class="k">return</span> <span class="n">metabolites</span></div>



<span class="c1"># build reaction</span>
<span class="c1"># --------------</span>
<span class="c1">#@TODO</span>
<span class="c1"># UNDER CONSTRUCTION</span>
<div class="viewcode-block" id="add_reaction">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.add_reaction">[docs]</a>
<span class="k">def</span> <span class="nf">add_reaction</span><span class="p">(</span><span class="n">model</span><span class="p">:</span><span class="n">cobra</span><span class="o">.</span><span class="n">Model</span><span class="p">,</span><span class="n">row</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span>
                 <span class="n">reac_xref</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span><span class="n">reac_prop</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
                 <span class="n">chem_xref</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span><span class="n">chem_prop</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
                 <span class="n">bigg_metabolites</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> 
                 <span class="n">namespace</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">&#39;BiGG&#39;</span><span class="p">,</span> 
                 <span class="n">exclude_dna</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exclude_rna</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">cobra</span><span class="o">.</span><span class="n">Model</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Helper function for extent_model. For a given row of a gene to be added, add the </span>
<span class="sd">    corresponding reactions and metabolites via MetaNetX/KEGG as needed and available.</span>

<span class="sd">    Args:</span>
<span class="sd">        - model (cobra.Model): </span>
<span class="sd">            The model.</span>
<span class="sd">        - row (pd.Series): </span>
<span class="sd">            A row from the table in function extend_model.</span>
<span class="sd">        - reac_xref (pd.DataFrame): </span>
<span class="sd">            The MetaNetX reac_xref table.</span>
<span class="sd">        - reac_prop (pd.DataFrame): </span>
<span class="sd">            The MetaNetX reac_prop table.</span>
<span class="sd">        - chem_xref (pd.DataFrame): </span>
<span class="sd">            The MetaNetX chem_xref table.</span>
<span class="sd">        - chem_prop (pd.DataFrame): </span>
<span class="sd">            The MetaNetX chem_prop table.</span>
<span class="sd">        - bigg_metabolites (pd.DataFrame): </span>
<span class="sd">            The BiGG compound namespace table.</span>
<span class="sd">        - namespace (str, optional): </span>
<span class="sd">            Namespace of the model. </span>
<span class="sd">            Defaults to &#39;BiGG&#39;.</span>
<span class="sd">        - exclude_dna (bool, optional): </span>
<span class="sd">            Flag to exclude the addition of </span>
<span class="sd">            directly DNA-related reactions. </span>
<span class="sd">            Defaults to True.</span>
<span class="sd">        - exclude_rna (bool, optional): </span>
<span class="sd">            Flag to exclude the addition of </span>
<span class="sd">            directly RNA-related reactions. </span>
<span class="sd">            Defaults to True.</span>

<span class="sd">    Returns:</span>
<span class="sd">        cobra.Model: </span>
<span class="sd">            The extended model.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># create reaction object</span>
    <span class="n">reac</span> <span class="o">=</span> <span class="n">cobra</span><span class="o">.</span><span class="n">Reaction</span><span class="p">(</span><span class="n">create_random_id</span><span class="p">(</span><span class="n">model</span><span class="p">,</span><span class="s1">&#39;reac&#39;</span><span class="p">,</span><span class="s1">&#39;SPECIMEN&#39;</span><span class="p">))</span>

    <span class="c1"># ----------------------------</span>
    <span class="c1"># curate reaction via MetaNetX</span>
    <span class="c1"># ----------------------------</span>
    <span class="c1"># try kegg.reaction --&gt; metanetx.reaction</span>
    <span class="k">if</span> <span class="sa">F</span><span class="s1">&#39;kegg.reaction:</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;KEGG.reaction&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">reac_xref</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">]):</span>
        
        <span class="c1"># get MetaNetX ID</span>
        <span class="n">met_reac_kegg</span> <span class="o">=</span> <span class="n">reac_xref</span><span class="p">[</span><span class="n">reac_xref</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">]</span><span class="o">==</span><span class="sa">F</span><span class="s1">&#39;kegg.reaction:</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;KEGG.reaction&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span>
        <span class="n">met_reac</span> <span class="o">=</span> <span class="n">reac_prop</span><span class="p">[</span><span class="n">reac_prop</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">==</span><span class="n">met_reac_kegg</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>

        <span class="c1"># make sure exactly one entry is parsed</span>
        <span class="c1"># @TODO : parallel parsing</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">met_reac</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;Warning: multiple matches for kegg.reaction </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;KEGG.reaction&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1"> found. Only first one will be used.&#39;</span><span class="p">)</span>
            <span class="n">met_reac</span> <span class="o">=</span> <span class="n">met_reac</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

        <span class="c1"># add name</span>
        <span class="c1"># --------</span>
        <span class="c1">#     from MetaNetX KEGG description</span>
        <span class="n">reac</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">met_reac_kegg</span><span class="p">[</span><span class="s1">&#39;description&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;|&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

        <span class="c1"># add notes</span>
        <span class="c1"># ---------</span>
        <span class="n">reac</span><span class="o">.</span><span class="n">notes</span><span class="p">[</span><span class="s1">&#39;creation&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;via MetaNetX&#39;</span>
        <span class="n">reac</span><span class="o">.</span><span class="n">notes</span><span class="p">[</span><span class="s1">&#39;KEGG.information&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;KEGG.notes&#39;</span><span class="p">]</span>

        <span class="c1"># add metabolites</span>
        <span class="c1"># ----------------</span>
        <span class="n">reac</span><span class="o">.</span><span class="n">add_metabolites</span><span class="p">(</span><span class="n">get_metabolites_mnx</span><span class="p">(</span><span class="n">model</span><span class="p">,</span><span class="n">met_reac</span><span class="p">[</span><span class="s1">&#39;mnx equation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">chem_xref</span><span class="p">,</span><span class="n">chem_prop</span><span class="p">,</span><span class="n">bigg_metabolites</span><span class="p">,</span> <span class="n">namespace</span><span class="p">))</span>
        <span class="c1">#@TODO .............</span>
        <span class="c1">#   direction of reaction</span>
        <span class="c1">#   ---&gt; current solution:</span>
        <span class="c1">#        use one direction only</span>
        <span class="c1"># ..................</span>

        <span class="c1"># add annotations</span>
        <span class="c1"># ---------------</span>
        <span class="n">reac</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;ec-code&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;EC number&#39;</span><span class="p">]</span>
        <span class="n">reac</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;kegg.reaction&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;KEGG.reaction&#39;</span><span class="p">]</span>
        <span class="n">reac</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;metanetx.reaction&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">met_reac_kegg</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">met_reac_anno</span> <span class="o">=</span> <span class="n">reac_xref</span><span class="p">[</span><span class="n">reac_xref</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">==</span><span class="n">met_reac_kegg</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
        <span class="k">for</span> <span class="n">db</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;metacyc.reaction&#39;</span><span class="p">,</span><span class="s1">&#39;seed.reaction&#39;</span><span class="p">,</span><span class="s1">&#39;rhea&#39;</span><span class="p">,</span><span class="s1">&#39;bigg.reaction&#39;</span><span class="p">]:</span>
            <span class="n">db_matches</span> <span class="o">=</span> <span class="n">met_reac_anno</span><span class="p">[</span><span class="n">met_reac_anno</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">db</span><span class="p">)]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">db_matches</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">reac</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">db</span><span class="p">]</span> <span class="o">=</span> <span class="n">db_matches</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">db_matches</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">reac</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">db</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">r</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">db_matches</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">continue</span>
    
    <span class="c1"># if not possible, use information from KEGG only</span>
    <span class="c1"># ------------------------</span>
    <span class="c1"># curate reaction via KEGG</span>
    <span class="c1"># ------------------------</span>
    <span class="k">else</span><span class="p">:</span>
        
        <span class="c1"># retrieve reaction information from KEGG</span>
        <span class="n">reac_kegg</span> <span class="o">=</span> <span class="n">kegg_reaction_parser</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;KEGG.reaction&#39;</span><span class="p">])</span>

        <span class="c1"># add name</span>
        <span class="c1"># --------</span>
        <span class="c1">#     from KEGG name</span>
        <span class="n">reac</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">reac_kegg</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span>

        <span class="c1"># add notes</span>
        <span class="c1"># ---------</span>
        <span class="n">reac</span><span class="o">.</span><span class="n">notes</span><span class="p">[</span><span class="s1">&#39;creation&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;via KEGG&#39;</span>
        <span class="n">reac</span><span class="o">.</span><span class="n">notes</span><span class="p">[</span><span class="s1">&#39;KEGG.information&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;KEGG.notes&#39;</span><span class="p">]</span>

        <span class="c1"># add metabolites</span>
        <span class="c1"># ----------------</span>
        <span class="n">reac</span><span class="o">.</span><span class="n">add_metabolites</span><span class="p">(</span><span class="n">get_metabolites_kegg</span><span class="p">(</span><span class="n">model</span><span class="p">,</span><span class="n">reac_kegg</span><span class="p">[</span><span class="s1">&#39;equation&#39;</span><span class="p">],</span><span class="n">chem_xref</span><span class="p">,</span><span class="n">chem_prop</span><span class="p">,</span><span class="n">bigg_metabolites</span><span class="p">,</span> <span class="n">namespace</span><span class="p">))</span>
            <span class="c1">#@TODO .............</span>
            <span class="c1">#   direction of reaction</span>
            <span class="c1">#   ---&gt; current solution:</span>
            <span class="c1">#        use one direction only</span>
            <span class="c1"># ..................</span>

        <span class="c1"># add annotations</span>
        <span class="c1"># ---------------</span>
        <span class="n">reac</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;ec-code&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;EC number&#39;</span><span class="p">]</span>
        <span class="n">reac</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="s1">&#39;kegg.reaction&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;KEGG.reaction&#39;</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">db</span><span class="p">,</span> <span class="n">identifiers</span> <span class="ow">in</span> <span class="n">reac_kegg</span><span class="p">[</span><span class="s1">&#39;db&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">identifiers</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">reac</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">db</span><span class="p">]</span> <span class="o">=</span> <span class="n">identifiers</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">reac</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">db</span><span class="p">]</span> <span class="o">=</span> <span class="n">identifiers</span>


    <span class="c1"># --------------------------------------</span>
    <span class="c1"># re-set ID to fit namespace if possible</span>
    <span class="c1"># --------------------------------------</span>
    <span class="n">match_id_to_namespace</span><span class="p">(</span><span class="n">reac</span><span class="p">,</span> <span class="n">namespace</span><span class="p">)</span>

    <span class="c1"># ---------------------</span>
    <span class="c1"># add reaction to model</span>
    <span class="c1"># ---------------------</span>
    
    <span class="c1"># if the ID change results in an ID already in the model, use that reaction</span>
    <span class="k">if</span> <span class="n">reac</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="p">[</span><span class="n">_</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="p">]:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">reac</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s1"> already in model, not added a second time.&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># check if reaction is complete</span>
        <span class="c1"># and fullfills the requirements / parameters</span>
        <span class="k">if</span> <span class="n">isreaction_complete</span><span class="p">(</span><span class="n">reac</span><span class="p">,</span> <span class="n">exclude_dna</span><span class="p">,</span> <span class="n">exclude_rna</span><span class="p">):</span>
            <span class="n">model</span><span class="o">.</span><span class="n">add_reactions</span><span class="p">([</span><span class="n">reac</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;reaction </span><span class="si">{</span><span class="n">reac</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s1"> for gene </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;locus_tag&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1"> could not be completely reconstructed, not added to model.&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">model</span>

    <span class="c1"># --------</span>
    <span class="c1"># add gene</span>
    <span class="c1"># --------</span>
    <span class="c1"># check if gene is already in model</span>
    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span>
        <span class="c1"># if - for whatever reason - gene already in gpr, skip</span>
        <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">reac</span><span class="o">.</span><span class="n">id</span><span class="p">)</span><span class="o">.</span><span class="n">gene_reaction_rule</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">model</span>
        <span class="c1"># create new gpr, if nonexistent</span>
        <span class="k">elif</span> <span class="ow">not</span> <span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">reac</span><span class="o">.</span><span class="n">id</span><span class="p">)</span><span class="o">.</span><span class="n">gene_reaction_rule</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">reac</span><span class="o">.</span><span class="n">id</span><span class="p">)</span><span class="o">.</span><span class="n">gene_reaction_rule</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">reac</span><span class="o">.</span><span class="n">id</span><span class="p">)</span><span class="o">.</span><span class="n">gene_reaction_rule</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span>
        <span class="c1"># add gene to existing gpr</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">reac</span><span class="o">.</span><span class="n">id</span><span class="p">)</span><span class="o">.</span><span class="n">gene_reaction_rule</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">reac</span><span class="o">.</span><span class="n">id</span><span class="p">)</span><span class="o">.</span><span class="n">gene_reaction_rule</span> <span class="o">+</span> <span class="s1">&#39; or &#39;</span> <span class="o">+</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># add (to) gene reaction rule and curate new gene object</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">reac</span><span class="o">.</span><span class="n">id</span><span class="p">)</span><span class="o">.</span><span class="n">gene_reaction_rule</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">reac</span><span class="o">.</span><span class="n">id</span><span class="p">)</span><span class="o">.</span><span class="n">gene_reaction_rule</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">model</span> <span class="o">=</span> <span class="n">add_gene</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">reac</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">row</span><span class="p">,</span> <span class="n">first</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">model</span> <span class="o">=</span> <span class="n">add_gene</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">reac</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">row</span><span class="p">,</span> <span class="n">first</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">model</span></div>



<span class="c1"># extend model</span>
<span class="c1"># ------------</span>
<span class="c1"># notes</span>
<span class="c1"># @CHECK : connections, e.g. input is now a param short </span>
<div class="viewcode-block" id="extent_model">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.extent_model">[docs]</a>
<span class="k">def</span> <span class="nf">extent_model</span><span class="p">(</span><span class="n">table</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span><span class="n">cobra</span><span class="o">.</span><span class="n">Model</span><span class="p">,</span>
                 <span class="n">chem_prop_file</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span><span class="n">chem_xref_file</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span>
                 <span class="n">reac_prop_file</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span><span class="n">reac_xref_file</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> 
                 <span class="n">namespace</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">&#39;BiGG&#39;</span><span class="p">,</span> 
                 <span class="n">exclude_dna</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exclude_rna</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">cobra</span><span class="o">.</span><span class="n">Model</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add reactions, metabolites and genes to a model based on the output of </span>
<span class="sd">    :py:func:`~specimen.core.refinement.extension.map_to_bigg`.</span>

<span class="sd">    Args:</span>
<span class="sd">        - table (pd.DataFrame): </span>
<span class="sd">            The table with the information to be added to the model.</span>
<span class="sd">            Output of :py:func:`~specimen.core.refinement.extension.map_to_bigg`.</span>
<span class="sd">        - model (cobra.Model): </span>
<span class="sd">            The model to extend.</span>
<span class="sd">        - chem_prop_file (str): </span>
<span class="sd">            Path to the MetaNetX chem_prop file.</span>
<span class="sd">        - chem_xref_file (str): </span>
<span class="sd">            Path to the MetaNetX chem_xref file.</span>
<span class="sd">        - reac_prop_file (str): </span>
<span class="sd">            Path to the MetaNetX reac_prop file.</span>
<span class="sd">        - reac_xref_file (str): </span>
<span class="sd">            Path to the MetaNetX reac_xref file.</span>
<span class="sd">        - namespace (str, optional): </span>
<span class="sd">            Namespace of the model. </span>
<span class="sd">            Defaults to &#39;BiGG&#39;.</span>
<span class="sd">        - exclude_dna (bool, optional): </span>
<span class="sd">            Tag to include or exclude DNA reactions. </span>
<span class="sd">            Defaults to True.</span>
<span class="sd">        - exclude_rna (bool, optional): </span>
<span class="sd">            Tag to include or exclude RNA reactions. </span>
<span class="sd">            Defaults to True.</span>

<span class="sd">    Returns:</span>
<span class="sd">        cobra.Model: </span>
<span class="sd">            The extended model.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># load MetaNetX database / namespace</span>
    <span class="n">chem_prop</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">chem_prop_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s1">&#39;#&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">,</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;reference&#39;</span><span class="p">,</span><span class="s1">&#39;formula&#39;</span><span class="p">,</span><span class="s1">&#39;charge&#39;</span><span class="p">,</span><span class="s1">&#39;mass&#39;</span><span class="p">,</span><span class="s1">&#39;InChI&#39;</span><span class="p">,</span><span class="s1">&#39;InChIKey&#39;</span><span class="p">,</span><span class="s1">&#39;SMILES&#39;</span><span class="p">])</span>
    <span class="n">chem_xref</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">chem_xref_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s1">&#39;#&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">,</span><span class="s1">&#39;ID&#39;</span><span class="p">,</span><span class="s1">&#39;description&#39;</span><span class="p">])</span>

    <span class="n">reac_prop</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">reac_prop_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s1">&#39;#&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">,</span><span class="s1">&#39;mnx equation&#39;</span><span class="p">,</span><span class="s1">&#39;reference&#39;</span><span class="p">,</span><span class="s1">&#39;classifs&#39;</span><span class="p">,</span><span class="s1">&#39;is_balanced&#39;</span><span class="p">,</span><span class="s1">&#39;is_transport&#39;</span><span class="p">])</span>
    <span class="n">reac_xref</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">reac_xref_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s1">&#39;#&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">,</span><span class="s1">&#39;ID&#39;</span><span class="p">,</span><span class="s1">&#39;description&#39;</span><span class="p">])</span>

    <span class="c1"># load bigg metabolite namespace</span>
    <span class="n">bigg_metabolites</span> <span class="o">=</span> <span class="n">load_a_table_from_database</span><span class="p">(</span><span class="s1">&#39;bigg_metabolites&#39;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
    <span class="n">bigg_metabolites</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span><span class="s1">&#39;bigg_id&#39;</span><span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">bigg_metabolites</span> <span class="o">=</span> <span class="n">bigg_metabolites</span><span class="p">[[</span><span class="s1">&#39;bigg_id&#39;</span><span class="p">,</span><span class="s1">&#39;universal_bigg_id&#39;</span><span class="p">,</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;CHEBI&#39;</span><span class="p">,</span><span class="s1">&#39;BioCyc&#39;</span><span class="p">,</span><span class="s1">&#39;KEGG Compound&#39;</span><span class="p">,</span><span class="s1">&#39;MetaNetX (MNX) Chemical&#39;</span><span class="p">,</span><span class="s1">&#39;SEED Compound&#39;</span><span class="p">,</span><span class="s1">&#39;InChI Key&#39;</span><span class="p">]]</span>

    <span class="c1"># add genes one by one to model</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Adding genes and if needed reactions and metabolites to model:&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">row_idx</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">table</span><span class="o">.</span><span class="n">index</span><span class="p">):</span>

        <span class="c1"># generate Name -&gt; KEGG.reaction dictionary</span>
        <span class="n">react_dict</span> <span class="o">=</span> <span class="n">get_reaction_annotation_dict</span><span class="p">(</span><span class="n">model</span><span class="p">,</span><span class="s1">&#39;KEGG&#39;</span><span class="p">)</span>
        <span class="c1"># generate Name -&gt; BiGG.reaction dictionary</span>
        <span class="n">react_dict_2</span> <span class="o">=</span> <span class="n">get_reaction_annotation_dict</span><span class="p">(</span><span class="n">model</span><span class="p">,</span><span class="s1">&#39;BiGG&#39;</span><span class="p">)</span>

        <span class="c1"># get row in pandas format</span>
        <span class="n">row</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">row_idx</span><span class="p">]</span>


        <span class="c1"># case 1: BiGG name already in model = reaction in model</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">pd</span><span class="o">.</span><span class="n">isnull</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;bigg_id&#39;</span><span class="p">])</span> <span class="ow">and</span> <span class="nb">any</span><span class="p">((</span><span class="kc">True</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;bigg_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">react_dict_2</span><span class="o">.</span><span class="n">values</span><span class="p">())):</span>
            <span class="c1"># get matching reaction id(s)</span>
            <span class="n">reac_found</span> <span class="o">=</span> <span class="p">[</span><span class="n">_</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;bigg_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">react_dict_2</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>
            <span class="c1"># add genes to all reactions</span>
            <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">reac_found</span><span class="p">:</span>
                <span class="n">model</span> <span class="o">=</span> <span class="n">add_gene</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">r</span><span class="p">,</span> <span class="n">row</span><span class="p">)</span>

        <span class="c1"># case 1: KEGG reaction ID in model = reaction probably in model as well</span>
        <span class="k">elif</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;KEGG.reaction&#39;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">react_dict</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="c1"># get corresponding reaction</span>
            <span class="n">react_found</span> <span class="o">=</span> <span class="p">[</span><span class="n">_</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">react_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;KEGG.reaction&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">react_dict</span><span class="p">[</span><span class="n">_</span><span class="p">]]</span>
            <span class="c1"># add gene to all reactions found</span>
            <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">react_found</span><span class="p">:</span>
                <span class="n">model</span> <span class="o">=</span> <span class="n">add_gene</span><span class="p">(</span><span class="n">model</span><span class="p">,</span><span class="n">r</span><span class="p">,</span><span class="n">row</span><span class="p">)</span>

        <span class="c1"># case 3: reaction not in model</span>
        <span class="c1">#         -&gt; add reaction(s), gene and metabolites if needed</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># case 3.1: one reaction</span>
            <span class="n">react</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;KEGG.reaction&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">react</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">model</span> <span class="o">=</span> <span class="n">add_reaction</span><span class="p">(</span><span class="n">model</span><span class="p">,</span><span class="n">row</span><span class="p">,</span><span class="n">reac_xref</span><span class="p">,</span><span class="n">reac_prop</span><span class="p">,</span><span class="n">chem_xref</span><span class="p">,</span><span class="n">chem_prop</span><span class="p">,</span><span class="n">bigg_metabolites</span><span class="p">,</span> <span class="n">namespace</span><span class="p">,</span> <span class="n">exclude_dna</span><span class="p">,</span> <span class="n">exclude_rna</span><span class="p">)</span>

            <span class="c1"># case 3.2: multiple reactions</span>
            <span class="c1">#           add each reaction separatly, with the same gene for th gene reaction rule</span>
            <span class="c1"># note: zero reactions not possible due to previous filtering</span>
            <span class="k">else</span><span class="p">:</span>

                <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">react</span><span class="p">:</span>
                    <span class="n">temp_row</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">deep</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                    <span class="n">temp_row</span><span class="p">[</span><span class="s1">&#39;KEGG.reaction&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">r</span>
                    <span class="n">model</span> <span class="o">=</span> <span class="n">add_reaction</span><span class="p">(</span><span class="n">model</span><span class="p">,</span><span class="n">temp_row</span><span class="p">,</span><span class="n">reac_xref</span><span class="p">,</span><span class="n">reac_prop</span><span class="p">,</span><span class="n">chem_xref</span><span class="p">,</span><span class="n">chem_prop</span><span class="p">,</span><span class="n">bigg_metabolites</span><span class="p">,</span> <span class="n">namespace</span><span class="p">,</span> <span class="n">exclude_dna</span><span class="p">,</span> <span class="n">exclude_rna</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">model</span></div>


<span class="c1"># main</span>
<span class="c1"># ----</span>

<span class="c1"># @TEST : fitted to refinegems</span>
<span class="c1"># @CHECK : connections, e.g. input is now two params short </span>
<span class="c1"># run the main as function</span>
<div class="viewcode-block" id="run">
<a class="viewcode-back" href="../../../../../modules/core.html#specimen.hqtb.core.refinement.extension.run">[docs]</a>
<span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="n">draft</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">gene_list</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">fasta</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> 
        <span class="n">db</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dir</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> 
        <span class="n">mnx_chem_prop</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">mnx_chem_xref</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> 
        <span class="n">mnx_reac_prop</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">mnx_reac_xref</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> 
        <span class="n">ncbi_map</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">ncbi_dat</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> 
        <span class="nb">id</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">&#39;locus_tag&#39;</span><span class="p">,</span> 
        <span class="n">sensitivity</span><span class="p">:</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;sensitive&#39;</span><span class="p">,</span> <span class="s1">&#39;more-sensitive&#39;</span><span class="p">,</span> <span class="s1">&#39;very-sensitive&#39;</span><span class="p">,</span><span class="s1">&#39;ultra-sensitive&#39;</span><span class="p">]</span><span class="o">=</span><span class="s1">&#39;more-sensitive&#39;</span><span class="p">,</span> 
        <span class="n">coverage</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">95.0</span><span class="p">,</span> <span class="n">pid</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">90.0</span><span class="p">,</span> 
        <span class="n">threads</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> 
        <span class="n">exclude_dna</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exclude_rna</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> 
        <span class="n">memote</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extend a draft model.</span>

<span class="sd">    Using a multitude of information from MetaNetX, KEGG, NCBI and more,</span>
<span class="sd">    extend a draft model by yet unadded genes (&quot;missing genes&quot;) by mapping them against</span>
<span class="sd">    a database using DIAMOND.</span>

<span class="sd">    Args:</span>
<span class="sd">        - draft (str): </span>
<span class="sd">            Path to the draft model.</span>
<span class="sd">        - gene_list (str): </span>
<span class="sd">            Path to a csv file containing information on all the genes found in the annotated genome.</span>
<span class="sd">        - fasta (str): </span>
<span class="sd">            Path to the (protein) FASTA file containing the CDS sequences.</span>
<span class="sd">        - db (str): </span>
<span class="sd">            Path to the database used for running DIAMOND.</span>
<span class="sd">        - dir (str):  </span>
<span class="sd">            Path to the directory for the output (directories).</span>
<span class="sd">        - mnx_chem_prop (str): </span>
<span class="sd">            Path to the MetaNetX chem_prop namespace file.</span>
<span class="sd">        - mnx_chem_xref (str): </span>
<span class="sd">            Path to the MetaNetX chem_xref namespace file.</span>
<span class="sd">        - mnx_reac_prop (str): </span>
<span class="sd">            Path to the MetaNetX reac_prop namespace file.</span>
<span class="sd">        - mnx_reac_xref (str): </span>
<span class="sd">            Path to the MetaNetX reac_xref namespace file.</span>
<span class="sd">        - ncbi_map (str): </span>
<span class="sd">            Path to the NCBI information mapping file. Optional, but recommended.</span>
<span class="sd">        - ncbi_dat (str):</span>
<span class="sd">            Path to the NCBI database information file. Optional, but recommended.</span>
<span class="sd">        - id (str, optional): </span>
<span class="sd">            Name of the column of the csv file that contains the entries that were </span>
<span class="sd">            used as gene identifiers in the draft model. </span>
<span class="sd">            Defaults to &#39;locus_tag&#39;.</span>
<span class="sd">        - sensitivity (Literal[&#39;sensitive&#39;, &#39;more-sensitive&#39;, &#39;very-sensitive&#39;,&#39;ultra-sensitive&#39;], optional): </span>
<span class="sd">            Sensitivity mode for DIAMOND blastp run.</span>
<span class="sd">            Can be sensitive, more-sensitive, very-sensitive or ultra-sensitive.</span>
<span class="sd">            Defaults to &#39;more-sensitive&#39;.</span>
<span class="sd">        - coverage (float, optional): </span>
<span class="sd">            Threshold value for the query coverage for DIAMOND. </span>
<span class="sd">            Defaults to 95.0.</span>
<span class="sd">        - pid (float, optional): </span>
<span class="sd">            PID (percentage identity value) to filter the blast hist by. </span>
<span class="sd">            Only hits equal or above the given value are kept.</span>
<span class="sd">            Defaults to 90.0.</span>
<span class="sd">        - threads (int, optional): </span>
<span class="sd">            Number of threads to be used for DIAMOND. </span>
<span class="sd">            Defaults to 2.</span>
<span class="sd">        - exclude_dna (bool, optional): </span>
<span class="sd">            Exclude reactions with DNA in their name when added. </span>
<span class="sd">            Defaults to True.</span>
<span class="sd">        - exclude_rna (bool, optional): </span>
<span class="sd">            Exclude reactions with RNA in their name when added. </span>
<span class="sd">            Defaults to True.</span>
<span class="sd">        - memote (bool, optional): </span>
<span class="sd">            Use memote on the extended model. </span>
<span class="sd">            Defaults to False.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: Unknown sensitivity mode.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    

    <span class="k">if</span> <span class="ow">not</span> <span class="n">sensitivity</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;sensitive&#39;</span><span class="p">,</span><span class="s1">&#39;more-sensitive&#39;</span><span class="p">,</span><span class="s1">&#39;very-sensitive&#39;</span><span class="p">,</span><span class="s1">&#39;ultra-sensitive&#39;</span><span class="p">]:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;Unknown sensitivity mode </span><span class="si">{</span><span class="n">sensitivity</span><span class="si">}</span><span class="s1">. Choose from: sensitive, more-sensitive, very-sensitive, ultra-sensitive&#39;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">refinement step 1: extension</span><span class="se">\n</span><span class="s1">################################################################################</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>


    <span class="c1"># -----------------------</span>
    <span class="c1"># create output directory</span>
    <span class="c1"># -----------------------</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s2">&quot;step1-extension&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;Creating new directory </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s2">&quot;step1-extension&quot;</span><span class="p">))</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">FileExistsError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Given directory already has required structure.&#39;</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s2">&quot;manual_curation&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;Creating new directory </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s2">&quot;manual_curation&quot;</span><span class="p">))</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">FileExistsError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Given directory already has required structure.&#39;</span><span class="p">)</span>

    <span class="c1"># ----------------------</span>
    <span class="c1"># identify missing genes</span>
    <span class="c1"># ----------------------</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1"># ----------------------</span><span class="se">\n</span><span class="s1"># identify missing genes</span><span class="se">\n</span><span class="s1"># ----------------------&#39;</span><span class="p">)</span>

    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

    <span class="c1"># read in the draft model</span>
    <span class="n">draft</span> <span class="o">=</span> <span class="n">cobra</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read_sbml_model</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">draft</span><span class="p">))</span>
    <span class="c1"># identify missing genes and create new FASTA</span>
    <span class="n">missing_genes</span> <span class="o">=</span> <span class="n">identify_missing_genes</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">draft</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">fasta</span><span class="p">,</span> <span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s1">&#39;step1-extension&#39;</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">missing genes: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">missing_genes</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="n">end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">time: </span><span class="si">{</span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="si">}</span><span class="s1">s&#39;</span><span class="p">)</span>

    <span class="c1"># ------------------------------------------------------</span>
    <span class="c1"># use DIAMOND to identify homologs for the missing genes</span>
    <span class="c1"># ------------------------------------------------------</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1"># -----------</span><span class="se">\n</span><span class="s1"># run DIAMOND</span><span class="se">\n</span><span class="s1"># -----------&#39;</span><span class="p">)</span>

    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

    <span class="n">outname_diamond</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s1">&#39;step1-extension&#39;</span><span class="p">,</span><span class="s1">&#39;DIAMOND_results.tsv&#39;</span><span class="p">)</span>
    <span class="n">bl</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\\</span><span class="s2"> &quot;</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Running the following command:&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;diamond blastp -d </span><span class="si">{</span><span class="n">db</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span><span class="n">bl</span><span class="p">)</span><span class="si">}</span><span class="s1"> -q </span><span class="si">{</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span><span class="n">bl</span><span class="p">),</span><span class="s2">&quot;step1-extension&quot;</span><span class="p">,</span><span class="s2">&quot;missing_genes.fasta&quot;</span><span class="p">)</span><span class="si">}</span><span class="s1"> --</span><span class="si">{</span><span class="n">sensitivity</span><span class="si">}</span><span class="s1"> --query-cover </span><span class="si">{</span><span class="n">coverage</span><span class="si">}</span><span class="s1"> -p </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">threads</span><span class="p">)</span><span class="si">}</span><span class="s1"> -o </span><span class="si">{</span><span class="n">outname_diamond</span><span class="si">}</span><span class="s1"> --outfmt 6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore&#39;</span><span class="p">)</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="sa">F</span><span class="s1">&#39;diamond blastp -d </span><span class="si">{</span><span class="n">db</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span><span class="n">bl</span><span class="p">)</span><span class="si">}</span><span class="s1"> -q </span><span class="si">{</span><span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span><span class="n">bl</span><span class="p">),</span><span class="s2">&quot;step1-extension&quot;</span><span class="p">,</span><span class="s2">&quot;missing_genes.fasta&quot;</span><span class="p">)</span><span class="si">}</span><span class="s1"> --</span><span class="si">{</span><span class="n">sensitivity</span><span class="si">}</span><span class="s1"> --query-cover </span><span class="si">{</span><span class="n">coverage</span><span class="si">}</span><span class="s1"> -p </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">threads</span><span class="p">)</span><span class="si">}</span><span class="s1"> -o </span><span class="si">{</span><span class="n">outname_diamond</span><span class="si">}</span><span class="s1"> --outfmt 6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore&#39;</span><span class="p">],</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">time: </span><span class="si">{</span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="si">}</span><span class="s1">s&#39;</span><span class="p">)</span>

    <span class="c1"># -----------------------------------------</span>
    <span class="c1"># find best hits for each gene, if possible</span>
    <span class="c1"># -----------------------------------------</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1"># --------------------</span><span class="se">\n</span><span class="s1"># filter out best hits</span><span class="se">\n</span><span class="s1"># --------------------&#39;</span><span class="p">)</span>

    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

    <span class="c1"># best hit for each query / missing gene, if one exists</span>
    <span class="n">best_hits</span> <span class="o">=</span> <span class="n">find_best_diamond_hits</span><span class="p">(</span><span class="n">outname_diamond</span><span class="p">,</span> <span class="n">pid</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">genes after DIAMOND with coverage &gt; </span><span class="si">{</span><span class="n">coverage</span><span class="si">}</span><span class="s1">, PID </span><span class="si">{</span><span class="n">pid</span><span class="si">}</span><span class="s1">: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">best_hits</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="n">end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">time: </span><span class="si">{</span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="si">}</span><span class="s1">s&#39;</span><span class="p">)</span>

    <span class="c1"># --------------------------------------------</span>
    <span class="c1"># get further information on the missing genes</span>
    <span class="c1"># --------------------------------------------</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1"># --------------------------</span><span class="se">\n</span><span class="s1"># get additional information</span><span class="se">\n</span><span class="s1"># --------------------------&#39;</span><span class="p">)</span>

    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

    <span class="c1"># from NCBI</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Retrieve information from NCBI via accession version number&#39;</span><span class="p">)</span>
    <span class="n">genes_to_add</span> <span class="o">=</span> <span class="n">get_ncbi_info</span><span class="p">(</span><span class="n">best_hits</span><span class="p">,</span> <span class="n">ncbi_map</span><span class="p">)</span>

    <span class="c1"># map to KEGG (gene or ec --&gt; enzyme and reaction)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Map to KEGG&#39;</span><span class="p">)</span>
    <span class="n">genes_to_add</span> <span class="o">=</span> <span class="n">map_to_KEGG</span><span class="p">(</span><span class="n">genes_to_add</span><span class="p">,</span> <span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s2">&quot;step1-extension&quot;</span><span class="p">),</span> <span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s2">&quot;manual_curation&quot;</span><span class="p">),</span> <span class="n">ncbi_dat</span><span class="p">)</span>

    <span class="c1"># map to BiGG</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">map information to BiGG namespace via EC number AND KEGG.reaction ID&#39;</span><span class="p">)</span>
    <span class="n">genes_to_add</span> <span class="o">=</span> <span class="n">map_BiGG_reactions</span><span class="p">(</span><span class="n">genes_to_add</span><span class="p">)</span>

    <span class="n">end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">time: </span><span class="si">{</span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="si">}</span><span class="s1">s&#39;</span><span class="p">)</span>

    <span class="c1"># ------------------</span>
    <span class="c1"># add genes to model</span>
    <span class="c1"># ------------------</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1"># ------------------</span><span class="se">\n</span><span class="s1"># add genes to model</span><span class="se">\n</span><span class="s1"># ------------------&#39;</span><span class="p">)</span>

    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="c1"># save starting values</span>
    <span class="n">m_before</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">draft</span><span class="o">.</span><span class="n">metabolites</span><span class="p">)</span>
    <span class="n">r_before</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">draft</span><span class="o">.</span><span class="n">reactions</span><span class="p">)</span>
    <span class="n">g_before</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">draft</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>

    <span class="c1"># extent the model</span>
    <span class="n">draft</span> <span class="o">=</span> <span class="n">extent_model</span><span class="p">(</span><span class="n">genes_to_add</span><span class="p">,</span><span class="n">draft</span><span class="p">,</span><span class="n">mnx_chem_prop</span><span class="p">,</span><span class="n">mnx_chem_xref</span><span class="p">,</span><span class="n">mnx_reac_prop</span><span class="p">,</span><span class="n">mnx_reac_xref</span><span class="p">,</span> <span class="n">exclude_dna</span><span class="p">,</span> <span class="n">exclude_rna</span><span class="p">)</span>
    <span class="c1"># save it</span>
    <span class="n">name</span> <span class="o">=</span> <span class="sa">F</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">draft</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s1">_extended&#39;</span>
    <span class="n">cobra</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">write_sbml_model</span><span class="p">(</span><span class="n">draft</span><span class="p">,</span> <span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s1">&#39;step1-extension&#39;</span><span class="p">,</span><span class="n">name</span><span class="o">+</span><span class="s1">&#39;.xml&#39;</span><span class="p">))</span>

    <span class="c1"># find out the differences to before</span>
    <span class="n">r_after</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">draft</span><span class="o">.</span><span class="n">reactions</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">number of added reactions: </span><span class="si">{</span><span class="n">r_after</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">r_before</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">m_after</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">draft</span><span class="o">.</span><span class="n">metabolites</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">number of added metabolites: </span><span class="si">{</span><span class="n">m_after</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">m_before</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">g_after</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">draft</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">number of added genes: </span><span class="si">{</span><span class="n">g_after</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">g_before</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="n">end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">time: </span><span class="si">{</span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="si">}</span><span class="s1">s&#39;</span><span class="p">)</span>

    <span class="c1"># ---------------------------------</span>
    <span class="c1"># assess model quality using memote</span>
    <span class="c1"># ---------------------------------</span>

    <span class="k">if</span> <span class="n">memote</span><span class="p">:</span>
        <span class="n">memote_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="s1">&#39;step1-extension&#39;</span><span class="p">,</span><span class="n">name</span><span class="o">+</span><span class="s1">&#39;.html&#39;</span><span class="p">)</span>
        <span class="n">run_memote</span><span class="p">(</span><span class="n">draft</span><span class="p">,</span> <span class="s1">&#39;html&#39;</span><span class="p">,</span> <span class="n">return_res</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">save_res</span><span class="o">=</span><span class="n">memote_path</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Carolin Brune.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>