<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>specimen.core package &mdash; SPECIMEN 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=559046d4" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
      <link rel="stylesheet" type="text/css" href="../_static/custom_theme.css?v=ef5873f9" />

  
    <link rel="shortcut icon" href="../_static/LogoSPECIMEN.png"/>
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=8d563738"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
        <script src="../_static/copybutton.js?v=7859de39"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="specimen.util package" href="util.html" />
    <link rel="prev" title="specimen.classes package" href="classes.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html">
            
              <img src="../_static/LogoSPECIMEN.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../about-pipeline.html">About SPECIMEN</a></li>
<li class="toctree-l1"><a class="reference internal" href="../run-pipeline.html">Run the Pipeline</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../specimen.html">Contents of SPECIMEN</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../specimen.html#specimen-subpackages">specimen subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="classes.html">specimen.classes</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">specimen.core</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#specimen-core-refinement-subpackage">specimen.core.refinement subpackage</a></li>
<li class="toctree-l4"><a class="reference internal" href="#specimen-core-submodules">specimen.core submodules</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="util.html">specimen.util</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../specimen.html#specimen-submodules">specimen submodules</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../help.html">Help &amp; FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dev-notes.html">Notes for Developers</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">SPECIMEN</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../specimen.html">Contents of SPECIMEN</a></li>
      <li class="breadcrumb-item active">specimen.core package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/modules/core.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="specimen-core-package">
<h1>specimen.core package<a class="headerlink" href="#specimen-core-package" title="Link to this heading"></a></h1>
<section id="specimen-core-refinement-subpackage">
<h2>specimen.core.refinement subpackage<a class="headerlink" href="#specimen-core-refinement-subpackage" title="Link to this heading"></a></h2>
<section id="module-specimen.core.refinement.annotation">
<span id="specimen-core-refinement-annotation-submodule"></span><h3>specimen.core.refinement.annotation submodule<a class="headerlink" href="#module-specimen.core.refinement.annotation" title="Link to this heading"></a></h3>
<p>Further annotate a model.</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.annotation.kegg_reaction_to_kegg_pathway">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.annotation.</span></span><span class="sig-name descname"><span class="pre">kegg_reaction_to_kegg_pathway</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">viaEC</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">viaRC</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/specimen/core/refinement/annotation.html#kegg_reaction_to_kegg_pathway"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.annotation.kegg_reaction_to_kegg_pathway" title="Link to this definition"></a></dt>
<dd><p>Retrieve the KEGG pathways for existing KEGG reaction IDs, if
they have yet to be added. Depending on the given options, only the
reactions is searched or additional searches are started using the
EC number and reactions class if the first search was unsuccesful.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>model (cobra.Model): </dt><dd><p>The model - loaded with COBRApy - to be annotated.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>viaEC (bool, optional): </dt><dd><p>Option to search for KEGG pathway ID 
using the EC number if previous searches were unsuccesful. 
Defaults to False.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>viaRC (bool, optional): </dt><dd><p>Option to search for KEGG pathway ID 
using the reaction class if previous searches were unsuccesful. 
Defaults to False.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.annotation.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.annotation.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kegg_viaEC</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kegg_viaRC</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memote</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/specimen/core/refinement/annotation.html#run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.annotation.run" title="Link to this definition"></a></dt>
<dd><p>Further annotate a given model.</p>
<p>Currently can add annotations for:</p>
<ul class="simple">
<li><p>SBO using SBOannotator</p></li>
<li><p>KEGG.reaction -&gt; KEGG.pathway</p></li>
</ul>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>model (str): </dt><dd><p>Path to the model.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>dir (str): </dt><dd><p>Path to the output directory.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>kegg_viaEC (bool, optional): </dt><dd><p>Option to search for KEGG pathway ID using the 
EC number if previous searches were unsuccesful. 
Defaults to False.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>kegg_viaRC (bool, optional): </dt><dd><p>Option to search for KEGG pathway ID using the 
reaction class if previous searches were unsuccesful. 
Defaults to False.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>memote (bool, optional): </dt><dd><p>Optionally run memote after the annotation. 
Defaults to False.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.core.refinement.cleanup">
<span id="specimen-core-refinement-cleanup-submodule"></span><h3>specimen.core.refinement.cleanup submodule<a class="headerlink" href="#module-specimen.core.refinement.cleanup" title="Link to this heading"></a></h3>
<p>Perform part 3 of the refinement: clean-up.</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.cleanup.check_direction">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.cleanup.</span></span><span class="sig-name descname"><span class="pre">check_direction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Model</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/cleanup.html#check_direction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.cleanup.check_direction" title="Link to this definition"></a></dt>
<dd><p>Check the direction of newly created reactions (01-extention) by searching for matching MetaCyc,
KEGG and MetaNetX IDs as well as EC number in a downloaded BioCyc (MetaCyc)
database table (need to contain at least the following columns:
Reactions (MetaCyc ID),EC-Number,KEGG reaction,METANETX,Reaction-Direction.</p>
<p>&#64;NOTE: checks only creations that do not contain the notes[‘creation’] == ‘via template’</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>model (cobra.Model): </dt><dd><p>The GEM containing the reactions to be check for direction.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>data_file (str): </dt><dd><p>Path to the MetabCyc (BioCyc) smart table.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>cobra.Model: </dt><dd><p>The updated model.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.cleanup.cobra_gapfill_wrapper">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.cleanup.</span></span><span class="sig-name descname"><span class="pre">cobra_gapfill_wrapper</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">universal</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Medium</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'BiGG'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'BiGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">growth_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunk_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Model</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/cleanup.html#cobra_gapfill_wrapper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.cleanup.cobra_gapfill_wrapper" title="Link to this definition"></a></dt>
<dd><p>Wrapper for <a class="reference internal" href="#specimen.core.refinement.cleanup.single_cobra_gapfill" title="specimen.core.refinement.cleanup.single_cobra_gapfill"><code class="xref py py-func docutils literal notranslate"><span class="pre">single_cobra_gapfill()</span></code></a>.</p>
<p>Either use the full set of reactions in universal model by setting iteration to
0 or None or use them in randomized chunks for faster runtime (useful
on laptops). Note: when using the second option, be aware that this does not
test all reaction combinations exhaustively (heuristic approach!!!).</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>model (cobra.Model): </dt><dd><p>he model to perform gapfilling on.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>universal (cobra.Model): </dt><dd><p>A model with reactions to be potentially used for the gapfilling.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>medium (Medium): </dt><dd><p>A medium the model should grow on.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (Literal[‘BiGG’], optional): </dt><dd><p>Namespace to use for the model. 
Options include ‘BiGG’.
Defaults to ‘BiGG’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>growth_threshold (float, optional): </dt><dd><p>Growth threshold for the gapfilling. 
Defaults to 0.05.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>iterations (int, optional): </dt><dd><p>Number of iterations for the heuristic version of the gapfilling.
If 0 or None is given, uses full set of reactions.
Defaults to 3.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>chunk_size (int, optional): </dt><dd><p>Number of reactions to be used for gapfilling at the same time. 
If None or 0 is given, use full set, not heuristic.
Defaults to 10000.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>cobra.Model: </dt><dd><p>The gapfilled model, if a solution was found.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.cleanup.multiple_cobra_gapfill">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.cleanup.</span></span><span class="sig-name descname"><span class="pre">multiple_cobra_gapfill</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">universal</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">media_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Medium</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'BiGG'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'BiGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">growth_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunk_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Model</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/cleanup.html#multiple_cobra_gapfill"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.cleanup.multiple_cobra_gapfill" title="Link to this definition"></a></dt>
<dd><p>Perform <a class="reference internal" href="#specimen.core.refinement.cleanup.single_cobra_gapfill" title="specimen.core.refinement.cleanup.single_cobra_gapfill"><code class="xref py py-func docutils literal notranslate"><span class="pre">single_cobra_gapfill()</span></code></a> on a list of media.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>model (cobra.Model): </dt><dd><p>The model to be gapfilled.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>universal (cobra.Model): </dt><dd><p>The model to use reactions for gapfilling from.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>media_list (list[Medium]): </dt><dd><p>List ofmedia the model is supposed to grow on.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>growth_threshold (float, optional): </dt><dd><p>Growth threshold for the gapfilling. 
Defaults to 0.05.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>iterations (int, optional): </dt><dd><p>Number of iterations for the heuristic version of the gapfilling.
If 0 or None is given, uses full set of reactions.
Defaults to 3.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>chunk_size (int, optional): </dt><dd><p>Number of reactions to be used for gapfilling at the same time. 
If None or 0 is given, use full set, not heuristic.
Defaults to 10000.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>cobra.Model: </dt><dd><p>The gapfilled model, if a solution was found.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.cleanup.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.cleanup.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biocyc_db</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_dupl_reac</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_dupl_meta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'default'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_unused_meta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_dupl_reac</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_dupl_meta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">universal</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">media_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'BiGG'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'BiGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">growth_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunk_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memote</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/specimen/core/refinement/cleanup.html#run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.cleanup.run" title="Link to this definition"></a></dt>
<dd><p>Perform the second refinement step, cleanup, on a model.</p>
<p>The second refinement step resolves the following issues:</p>
<ol class="arabic simple">
<li><p>(optional) checking direction of reactions with BioCyc</p></li>
<li><p>resolve BioCyc/MetaCyc annotation inconsistencies</p></li>
<li><p>find and/or resolve duplicates (reactions and metabolites)</p></li>
<li><p>gapfilling using cobra + universal model with reactions + a set of media (&#64;TODO: under developement)</p></li>
</ol>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>model (str): </dt><dd><p>The Path to an sbml model.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>dir (str): </dt><dd><p>Path to the directory of the output.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>biocyc_db (str, optional): </dt><dd><p>Path to the BioCyc/MetaCyc reaction database file. 
Defaults to None, which leads to skipping the direction check.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>check_dupl_reac (bool, optional): </dt><dd><p>Option to check for duplicate reactions. 
Defaults to False.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>check_dupl_meta (bool, optional): </dt><dd><p>Option to check for duplicate metabolites. 
Defaults to ‘default’, which checks based on MetaNetX first.
Further options include ‘skip’ and ‘exhaustive’ (check for all possibilities).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>remove_unused_meta (bool, optional): </dt><dd><p>Option to remove unused metabolites from the model. 
Defaults to False.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>remove_dupl_reac (bool, optional): </dt><dd><p>Option to remove the duplicate reactions. 
True is only applied, if check_dupl_reac is also True.
Defaults to False.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>remove_dupl_meta (bool, optional): </dt><dd><p>Option to remove the duplicate metabolites. 
True is only applied, if check_dupl_meta is also True.
Defaults to False.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>universal (str, optional): </dt><dd><p>Path to a universal model for gapfilling. 
Defaults to None, which skips the gapfilling.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>media_path (str, optional): </dt><dd><p>Path to a medium config file for gapfilling. 
Defaults to None.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (Literal[‘BiGG’], optional): </dt><dd><p>Namespace to use for the model. 
Options include ‘BiGG’.
Defaults to ‘BiGG’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>growth_threshold (float, optional): </dt><dd><p>Growth threshold for the gapfilling. 
Defaults to 0.05.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>iterations (int, optional): </dt><dd><p>Number of iterations for the heuristic version of the gapfilling.
If 0 or None is given, uses full set of reactions.
Defaults to 3.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>chunk_size (int, optional): </dt><dd><p>Number of reactions to be used for gapfilling at the same time. 
If None or 0 is given, use full set, not heuristic.
Defaults to 10000.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>memote (bool, optional): </dt><dd><p>Option to run memote on the cleaned model. 
Defaults to False.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Raises:</dt><dd><p>ValueError: Unknown option for check_dupl_meta</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.cleanup.single_cobra_gapfill">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.cleanup.</span></span><span class="sig-name descname"><span class="pre">single_cobra_gapfill</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">universal</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Medium</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'BiGG'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'BiGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">growth_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.05</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bool</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/cleanup.html#single_cobra_gapfill"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.cleanup.single_cobra_gapfill" title="Link to this definition"></a></dt>
<dd><p>Attempt gapfilling (with COBRApy) for a given model to allow growth on a given
medium.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>model (cobra.Model): </dt><dd><p>The model to perform gapfilling on.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>universal (cobra.Model):  </dt><dd><p>A model with reactions to be potentially used for the gapfilling.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>medium (Medium): </dt><dd><p>A medium the model should grow on.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (Literal[‘BiGG’], optional): Namespace to use for the model. </dt><dd><p>Defaults to ‘BiGG’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>growth_threshold (float, optional):  Minimal rate for the model to be considered growing.</dt><dd><p>Defaults to 0.05.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>Union[list[str],True]: </dt><dd><p>List of reactions to be added to the model to allow growth
or True, if the model already grows.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.core.refinement.extension">
<span id="specimen-core-refinement-extension-submodule"></span><h3>specimen.core.refinement.extension submodule<a class="headerlink" href="#module-specimen.core.refinement.extension" title="Link to this heading"></a></h3>
<p>Part one of the third step of the pipeline: refinement - extension.</p>
<p>Extents the model by mapping missing genes to reactions via NCBI, KEGG,
MetaNetX and BiGG.</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.add_gene">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">add_gene</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reaction</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Series</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">first</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Model</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#add_gene"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.add_gene" title="Link to this definition"></a></dt>
<dd><p>Add a new gene to a genome-scale metabolic cobra model.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>model (cobra.Model): </dt><dd><p>The model.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>reaction (str): </dt><dd><p>The reaction id to add the gene to.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>row (pd.Series): </dt><dd><p>A single row of the output table of <a class="reference internal" href="#specimen.core.refinement.extension.map_BiGG_reactions" title="specimen.core.refinement.extension.map_BiGG_reactions"><code class="xref py py-func docutils literal notranslate"><span class="pre">map_BiGG_reactions()</span></code></a>.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>first (bool, optional): </dt><dd><p>Shows, if gene is the first gene to be added to the reaction. 
True if gene is first to be added.
Defaults to False.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>cobra.Model: </dt><dd><p>Updated model.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.add_reaction">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">add_reaction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Series</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reac_xref</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reac_prop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chem_xref</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chem_prop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigg_metabolites</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'BiGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_dna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_rna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Model</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#add_reaction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.add_reaction" title="Link to this definition"></a></dt>
<dd><p>Helper function for extent_model. For a given row of a gene to be added, add the 
corresponding reactions and metabolites via MetaNetX/KEGG as needed and available.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>model (cobra.Model): </dt><dd><p>The model.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>row (pd.Series): </dt><dd><p>A row from the table in function extend_model.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>reac_xref (pd.DataFrame): </dt><dd><p>The MetaNetX reac_xref table.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>reac_prop (pd.DataFrame): </dt><dd><p>The MetaNetX reac_prop table.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>chem_xref (pd.DataFrame): </dt><dd><p>The MetaNetX chem_xref table.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>chem_prop (pd.DataFrame): </dt><dd><p>The MetaNetX chem_prop table.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>bigg_metabolites (pd.DataFrame): </dt><dd><p>The BiGG compound namespace table.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (str, optional): </dt><dd><p>Namespace of the model. 
Defaults to ‘BiGG’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>exclude_dna (bool, optional): </dt><dd><p>Flag to exclude the addition of 
directly DNA-related reactions. 
Defaults to True.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>exclude_rna (bool, optional): </dt><dd><p>Flag to exclude the addition of 
directly RNA-related reactions. 
Defaults to True.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>cobra.Model: </dt><dd><p>The extended model.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.build_metabolite_kegg">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">build_metabolite_kegg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kegg_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_kegg_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigg_metabolites</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'BiGG'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'BiGG'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Metabolite</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#build_metabolite_kegg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.build_metabolite_kegg" title="Link to this definition"></a></dt>
<dd><p>Create or retrieve (from model) a metabolite based on its KEGG ID.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>kegg_id (str): </dt><dd><p>The KEGG.compound ID of the metabolite in question.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>model (cobra.Model): </dt><dd><p>The model.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>model_kegg_ids (list[str]): </dt><dd><p>List of all annotated KEGG Compound IDs in the model.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>bigg_metabolites (pd.DataFrame):  </dt><dd><p>The BiGG compound namespace table.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (Literal[‘BiGG’], optional): </dt><dd><p>Namespace of the model. Defaults to ‘BiGG’.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>cobra.Metabolite:  </dt><dd><p>The retrieved or newly build metabolite.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.build_metabolite_mnx">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">build_metabolite_mnx</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metabolite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnx_chem_prop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnx_chem_xref</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigg_metabolites</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Metabolite</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#build_metabolite_mnx"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.build_metabolite_mnx" title="Link to this definition"></a></dt>
<dd><p>Create or retrieve (from model) a metabolite based on its MetaNetX ID.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>metabolite (str): </dt><dd><p>The MetaNetX ID of the metabolite.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>model (cobra.Model):  </dt><dd><p>The underlying genome-scale metabolic model.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>mnx_chem_prop (pd.DataFrame): </dt><dd><p>The chem_xref table from MetaNetX.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>mnx_chem_xref (pd.DataFrame): </dt><dd><p>The chem_prop table from MetaNetX.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>bigg_metabolites (pd.DataFrame): </dt><dd><p>The BiGG compound namespace table.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (pd.DataFrame): </dt><dd><p>The namespace of the model.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>cobra.Metabolite: </dt><dd><p>The retrieved or newly build metabolite.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.extent_model">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">extent_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chem_prop_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chem_xref_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reac_prop_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reac_xref_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'BiGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_dna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_rna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Model</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#extent_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.extent_model" title="Link to this definition"></a></dt>
<dd><p>Add reactions, metabolites and genes to a model based on the output of 
<code class="xref py py-func docutils literal notranslate"><span class="pre">map_to_bigg()</span></code>.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>table (pd.DataFrame): </dt><dd><p>The table with the information to be added to the model.
Output of <code class="xref py py-func docutils literal notranslate"><span class="pre">map_to_bigg()</span></code>.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>model (cobra.Model): </dt><dd><p>The model to extend.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>chem_prop_file (str): </dt><dd><p>Path to the MetaNetX chem_prop file.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>chem_xref_file (str): </dt><dd><p>Path to the MetaNetX chem_xref file.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>reac_prop_file (str): </dt><dd><p>Path to the MetaNetX reac_prop file.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>reac_xref_file (str): </dt><dd><p>Path to the MetaNetX reac_xref file.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (str, optional): </dt><dd><p>Namespace of the model. 
Defaults to ‘BiGG’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>exclude_dna (bool, optional): </dt><dd><p>Tag to include or exclude DNA reactions. 
Defaults to True.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>exclude_rna (bool, optional): </dt><dd><p>Tag to include or exclude RNA reactions. 
Defaults to True.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>cobra.Model: </dt><dd><p>The extended model.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.find_best_diamond_hits">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">find_best_diamond_hits</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#find_best_diamond_hits"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.find_best_diamond_hits" title="Link to this definition"></a></dt>
<dd><p>Identify the best hits from the DIAMOND run.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>file (str): </dt><dd><p>File name (path) of the “missing_genes” file.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>pid (float): </dt><dd><p>Threshold value for the PID. 
Should be between 0.0 and 100.0 (given in percentage)</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>pd.DataFrame: </dt><dd><p>Table of the locus tags of the query and the NCBI accession version numbers of the RefSeqs.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.get_metabolites_kegg">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">get_metabolites_kegg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">equation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chem_xref</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chem_prop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigg_metabolites</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'BiGG'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'BiGG'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">Metabolite</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#get_metabolites_kegg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.get_metabolites_kegg" title="Link to this definition"></a></dt>
<dd><p>Based on a given KEGG equation and a model, get or
create metabolite entires in/for the model.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>model (cobra.Model): </dt><dd><p>A GEM.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>equation (str): </dt><dd><p>The equation from KEGG.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>chem_xref (pd.DataFrame): </dt><dd><p>The chem_xref table from MetaNetX.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>chem_prop (pd.DataFrame): </dt><dd><p>The chem_prop table from MetaNetX.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>bigg_metabolites (pd.DataFrame): </dt><dd><p>The BiGG compound namespace table.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (Literal[‘BiGG’], optional): </dt><dd><p>The namespace of the model. 
Defaults to ‘BiGG’.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>dict[cobra.Metabolite,int]: </dt><dd><p>Dictonary of metabolites and stoichiometric factors.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.get_metabolites_mnx">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">get_metabolites_mnx</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">equation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnx_chem_xref</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnx_chem_prop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigg_metabolites</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'BiGG'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'BiGG'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">Metabolite</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#get_metabolites_mnx"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.get_metabolites_mnx" title="Link to this definition"></a></dt>
<dd><p>Based on a given MetaNetX equation and a model, get or
create metabolite entires in/for the model.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>model (cobra.Model): </dt><dd><p>The GEM.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>equation (str): </dt><dd><p>The equation from MetaNetX.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>mnx_chem_xref (pd.DataFrame):  </dt><dd><p>The chem_xref table from MetaNetX.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>mnx_chem_prop (pd.DataFrame):  </dt><dd><p>The chem_prop table from MetaNetX.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>bigg_metabolites (pd.DataFrame): </dt><dd><p>The BiGG compound namespace table.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (Literal[‘BiGG’], optional): </dt><dd><p>Namespace of the model. 
Defaults to ‘BiGG’.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>dict[cobra.Metabolite,int]: </dt><dd><p>Dictonary of metabolites and stoichiometric factors.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.get_ncbi_info">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">get_ncbi_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncbi_map_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#get_ncbi_info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.get_ncbi_info" title="Link to this definition"></a></dt>
<dd><p>Retrieve information from NCBI via mapping to precomputed files.</p>
<blockquote>
<div><p>If no mapping exists, try via Entrez - implemented, but does not work well</p>
<p>-&gt; problems with connecting successfully to NCBI.</p>
<p>&#64;TODO if time, rewrite using the Bioppython package, see if it works better</p>
</div></blockquote>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>table (pd.DataFrame):  </dt><dd><p>A table containing the output of <a class="reference internal" href="#specimen.core.refinement.extension.find_best_diamond_hits" title="specimen.core.refinement.extension.find_best_diamond_hits"><code class="xref py py-func docutils literal notranslate"><span class="pre">specimen.core.refinement.extension.find_best_diamond_hits()</span></code></a>.
Must have the column ‘ncbi_accession_version’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>ncbi_map_file (str, optional): </dt><dd><p>A csv file containing the ncbi mapping. 
Defaults to None.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>pd.DataFrame:  </dt><dd><p>A table with the added information.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.identify_missing_genes">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">identify_missing_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fasta_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#identify_missing_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.identify_missing_genes" title="Link to this definition"></a></dt>
<dd><p>Identify missing genes in a model based on the draft model and a list of genes.
Save the missing genes’ sequences in a new FASTA file.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>gene_path (str): </dt><dd><p>Path to the csv file containing the list of genes.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>model (cobra.Model): </dt><dd><p>The draft model.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>id (str): </dt><dd><p>Name of the column in gene_path, that equal the ids in the model.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>fasta_path (str): </dt><dd><p>Path to the FASTA file containing all genes for the genome.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>dir (str): </dt><dd><p>Path to the directory to save the new FASTA in.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>pd.DataFrame: </dt><dd><p>A table with the missing genes’ information.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.isreaction_complete">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">isreaction_complete</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Reaction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_dna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_rna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#isreaction_complete"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.isreaction_complete" title="Link to this definition"></a></dt>
<dd><p>Check, if a reaction is complete and ready to be added to the model.
Additionally, it is possible to check for DNA and RNA reations
and set them to be excluded or included.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>reac (cobra.Reaction): </dt><dd><p>The reaction to be checked.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>exclude_dna (bool, optional): </dt><dd><p>Tag to include or exclude DNA reactions. 
Defaults to True.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>exclude_rna (bool, optional): </dt><dd><p>Tag to include or exclude RNA reactions. 
Defaults to True.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>bool: </dt><dd><p>True if the check is successful, else false.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.map_BiGG_reactions">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">map_BiGG_reactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#map_BiGG_reactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.map_BiGG_reactions" title="Link to this definition"></a></dt>
<dd><p>Map the output of <a class="reference internal" href="#specimen.core.refinement.extension.map_to_KEGG" title="specimen.core.refinement.extension.map_to_KEGG"><code class="xref py py-func docutils literal notranslate"><span class="pre">map_to_KEGG()</span></code></a>
to a BiGG namespace file (rewritten-type, see auxilliaries).</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>table_file (str): </dt><dd><p>The path to the saved table from running map_to_KEGG().</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>pd.DataFrame: </dt><dd><p>The table with an additional column for the mapping to BiGG reactions.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.map_BiGG_reactions_row">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">map_BiGG_reactions_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Series</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Series</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#map_BiGG_reactions_row"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.map_BiGG_reactions_row" title="Link to this definition"></a></dt>
<dd><p>Map a single entry from the table in <a class="reference internal" href="#specimen.core.refinement.extension.map_BiGG_reactions" title="specimen.core.refinement.extension.map_BiGG_reactions"><code class="xref py py-func docutils literal notranslate"><span class="pre">map_BiGG_reactions()</span></code></a> to the BiGG reaction namespace.</p>
<p>&#64;TODO</p>
<p>NOTE:</p>
<blockquote>
<div><p>only works for cases, where KEGG.reaction in row contains EXACTLY one entry
in the rare case that multiple reactions belong to one enzyme, they are omitted
in this search</p>
</div></blockquote>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>row (pd.Series): </dt><dd><p>A single row of the table.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (pd.DataFrame): </dt><dd><p>The BiGG reaction namespace table.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>pd.Series: </dt><dd><p>The edited row.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.map_to_KEGG">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">map_to_KEGG</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">working_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">manual_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_summary</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Path</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#map_to_KEGG"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.map_to_KEGG" title="Link to this definition"></a></dt>
<dd><p>Map entries of the table from <a class="reference internal" href="#specimen.core.refinement.extension.get_ncbi_info" title="specimen.core.refinement.extension.get_ncbi_info"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_ncbi_info()</span></code></a> to KEGG.
Get KEGG.reaction ID from locus tag, if available, otherwise try EC number.
Entries of the finalised table are saved depending whether they were assigned
a KEGG.reaction ID (further use) or not (possible starting point for manual curation).</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>gene_table (pd.DataFrame): </dt><dd><p>The input table containing the output of get_ncbi_info().</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>working_dir (str): </dt><dd><p>Path to the working directory - place to save genes with a KEGG.reaction ID.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>manual_dir (str): </dt><dd><p>Path to a directory to save the genes without KEGG assignment for possible manual curation.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>genome_summary (str, optional): </dt><dd><p>Path to a information file for the reference genomes to check if they are in KEGG. 
Defaults to None.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>Path: </dt><dd><p>The path to the file containing the genes with KEGG assignment.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.map_to_KEGG_row">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">map_to_KEGG_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Series</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loci_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Series</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#map_to_KEGG_row"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.map_to_KEGG_row" title="Link to this definition"></a></dt>
<dd><p>Map a single entry of the table in <a class="reference internal" href="#specimen.core.refinement.extension.map_to_KEGG" title="specimen.core.refinement.extension.map_to_KEGG"><code class="xref py py-func docutils literal notranslate"><span class="pre">map_to_KEGG()</span></code></a> to KEGG.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>row (pd.Series): </dt><dd><p>A single row of the table.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>loci_list (list[str], optional):</dt><dd><p>List of starting letters of locus tags belong to organisms, that are listed in KEGG
and the corresponding organism three-letter code. 
Defaults to None (= search all entries).</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>pd.Series: </dt><dd><p>The mapped row.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.map_to_NCBI_efetch_row">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">map_to_NCBI_efetch_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Series</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Series</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#map_to_NCBI_efetch_row"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.map_to_NCBI_efetch_row" title="Link to this definition"></a></dt>
<dd><p>Map a single entry from the table in <a class="reference internal" href="#specimen.core.refinement.extension.get_ncbi_info" title="specimen.core.refinement.extension.get_ncbi_info"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_ncbi_info()</span></code></a> 
to NCBI using EntrezDirect.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>row (pd.Series): </dt><dd><p>A single row of the table.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>pd.Series: </dt><dd><p>The edited row.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.extension.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.extension.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">draft</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fasta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnx_chem_prop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnx_chem_xref</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnx_reac_prop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnx_reac_xref</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncbi_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncbi_dat</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'locus_tag'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensitivity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'sensitive'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'more-sensitive'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'very-sensitive'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'ultra-sensitive'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'more-sensitive'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coverage</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">95.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">90.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_dna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_rna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memote</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/specimen/core/refinement/extension.html#run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.extension.run" title="Link to this definition"></a></dt>
<dd><p>Extend a draft model.</p>
<p>Using a multitude of information from MetaNetX, KEGG, NCBI and more,
extend a draft model by yet unadded genes (“missing genes”) by mapping them against
a database using DIAMOND.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>draft (str): </dt><dd><p>Path to the draft model.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>gene_list (str): </dt><dd><p>Path to a csv file containing information on all the genes found in the annotated genome.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>fasta (str): </dt><dd><p>Path to the (protein) FASTA file containing the CDS sequences.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>db (str): </dt><dd><p>Path to the database used for running DIAMOND.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>dir (str):  </dt><dd><p>Path to the directory for the output (directories).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>mnx_chem_prop (str): </dt><dd><p>Path to the MetaNetX chem_prop namespace file.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>mnx_chem_xref (str): </dt><dd><p>Path to the MetaNetX chem_xref namespace file.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>mnx_reac_prop (str): </dt><dd><p>Path to the MetaNetX reac_prop namespace file.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>mnx_reac_xref (str): </dt><dd><p>Path to the MetaNetX reac_xref namespace file.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>ncbi_map (str): </dt><dd><p>Path to the NCBI information mapping file. Optional, but recommended.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>ncbi_dat (str):</dt><dd><p>Path to the NCBI database information file. Optional, but recommended.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>id (str, optional): </dt><dd><p>Name of the column of the csv file that contains the entries that were 
used as gene identifiers in the draft model. 
Defaults to ‘locus_tag’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>sensitivity (Literal[‘sensitive’, ‘more-sensitive’, ‘very-sensitive’,’ultra-sensitive’], optional): </dt><dd><p>Sensitivity mode for DIAMOND blastp run.
Can be sensitive, more-sensitive, very-sensitive or ultra-sensitive.
Defaults to ‘more-sensitive’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>coverage (float, optional): </dt><dd><p>Threshold value for the query coverage for DIAMOND. 
Defaults to 95.0.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>pid (float, optional): </dt><dd><p>PID (percentage identity value) to filter the blast hist by. 
Only hits equal or above the given value are kept.
Defaults to 90.0.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>threads (int, optional): </dt><dd><p>Number of threads to be used for DIAMOND. 
Defaults to 2.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>exclude_dna (bool, optional): </dt><dd><p>Exclude reactions with DNA in their name when added. 
Defaults to True.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>exclude_rna (bool, optional): </dt><dd><p>Exclude reactions with RNA in their name when added. 
Defaults to True.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>memote (bool, optional): </dt><dd><p>Use memote on the extended model. 
Defaults to False.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Raises:</dt><dd><p>ValueError: Unknown sensitivity mode.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.core.refinement.smoothing">
<span id="specimen-core-refinement-smoothing-submodule"></span><h3>specimen.core.refinement.smoothing submodule<a class="headerlink" href="#module-specimen.core.refinement.smoothing" title="Link to this heading"></a></h3>
<p>Perform part 4 of the refinement: smoothing.</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.smoothing.adjust_BOF">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.smoothing.</span></span><span class="sig-name descname"><span class="pre">adjust_BOF</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dna_weight_fraction</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weight_frac</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/smoothing.html#adjust_BOF"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.smoothing.adjust_BOF" title="Link to this definition"></a></dt>
<dd><p>Adjust the model’s BOF using BOFdat. Currently implemented are step 1
DNA coefficients and step 2.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>genome (str): </dt><dd><p>Path to the genome (e.g. .fna) FASTA file.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>model_file (str): </dt><dd><p>Path to the sbml (.xml) file of the model.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>model (cobra.Model): </dt><dd><p>The genome-scale metabolic model (from the string above), loaded with COBRApy.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>dna_weight_fraction (float): </dt><dd><p>DNA weight fraction for BOF step 1.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>weight_frac (float): </dt><dd><p>Weight fraction for the second step of BOFdat (enzymes and ions)</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>str: </dt><dd><p>The updated BOF reaction as a reaction string.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.smoothing.perform_mcc">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.smoothing.</span></span><span class="sig-name descname"><span class="pre">perform_mcc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">apply</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Model</span></span></span><a class="reference internal" href="../_modules/specimen/core/refinement/smoothing.html#perform_mcc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.smoothing.perform_mcc" title="Link to this definition"></a></dt>
<dd><p>Run the MassChargeCuration toll on the model and optionally directly apply 
the solution.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>model (cobra.Model): </dt><dd><p>The model to use the tool on.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>dir (str): </dt><dd><p>Path of the directory to save MCC output in.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>apply (bool, optional): </dt><dd><p>If True, model is directly updated with the results. 
Defaults to True.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>cobra.Model: </dt><dd><p>The model (updated or not)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.refinement.smoothing.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.refinement.smoothing.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mcc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'skip'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">egc_solver</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'greedy'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'BiGG'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'BiGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dna_weight_frac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.023</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ion_weight_frac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memote</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/specimen/core/refinement/smoothing.html#run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.refinement.smoothing.run" title="Link to this definition"></a></dt>
<dd><p>Perform the fourth step of the refinement, smoothing, on a model.</p>
<p>The fourth step of the refinment, smoothing, includes:</p>
<ul class="simple">
<li><p>Mass-Charge curation</p></li>
<li><p>checking energy generating cycles</p></li>
<li><p>adjusting Biomass objective function parameters based on genome</p></li>
</ul>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>genome (str): </dt><dd><p>Path to the genome FASTA (e.g. .fna) file of your genome.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>model (str): </dt><dd><p>Path to the model. Should be sbml-format.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>dir (str): </dt><dd><p>Path of the output directory.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>mcc (str, optional): </dt><dd><p>Option for the Mass-Charge curation.
Can be ‘skip’, ‘apply’ (directly use MCC on model) or ‘extra’ (only generate information). 
Defaults to ‘skip’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>egc_solver (None | Literal[‘greedy’], optional): </dt><dd><p>If given, uses the option to solve EGCs.
Current options include greedy (Greedy Solver).
Defaults to ‘greedy’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (Literal[‘BiGG’], optional): </dt><dd><p>Namespace to use for the model.
Defaults to ‘BiGG’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>dna_weight_frac (float, optional): </dt><dd><p>DNA weight fraction to use for BOFdat.
Default is 0.023 for Klebsiella pneumoniae based on Liao et al.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>ion_weight_frac (float, optional): </dt><dd><p>Ion weight fraction to use for BOFdat.
Defaults to 0.05.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>memote (bool, optional): </dt><dd><p>Option to run memote after the refinement.
Defaults to False.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
</section>
<section id="specimen-core-submodules">
<h2>specimen.core submodules<a class="headerlink" href="#specimen-core-submodules" title="Link to this heading"></a></h2>
<section id="module-specimen.core.analysis">
<span id="specimen-core-analysis"></span><h3>specimen.core.analysis<a class="headerlink" href="#module-specimen.core.analysis" title="Link to this heading"></a></h3>
<p>Analyse a model (part 5 of the workflow).</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.analysis.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.analysis.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">media_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'BiGG'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'BiGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc_model_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc_based_on</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'id'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'id'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_aa_auxotrophies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pathway</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/specimen/core/analysis.html#run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.analysis.run" title="Link to this definition"></a></dt>
<dd><p>SPECIMEN Step 5: Analyse the generated model.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>model_path (str): </dt><dd><p>Path to the model.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>dir (str): </dt><dd><p>Path to the output directory.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>media_path (str, optional): </dt><dd><p>Path to a media config file. 
Using this enables growth simulation.
Defaults to None.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (Literal[‘BiGG’], optional): </dt><dd><p>Namespace to work on. 
Defaults to ‘BiGG’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>pc_model_path (str, optional): </dt><dd><p>Path to a core-pan model. Defaults to None.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>pc_based_on (Literal[‘id’], optional): </dt><dd><p>How to compare the model to the core-pan model. 
Defaults to ‘id’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>test_aa_auxotrophies (bool, optional): </dt><dd><p>Option to enable the amino acid
auxotrophy simulation. Defaults to True.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>pathway (bool, optional): </dt><dd><p>Optional to enable KEGG pathway analysis. 
Defaults to True.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.core.bidirectional_blast">
<span id="specimen-core-bidirectional-blast"></span><h3>specimen.core.bidirectional_blast<a class="headerlink" href="#module-specimen.core.bidirectional_blast" title="Link to this heading"></a></h3>
<p>Perform a bidirectional blastp using DIAMOND on an input and a template (annotated genomes).</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.bdbp_diamond">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">bdbp_diamond</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">template_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">template_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensitivity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sensitive'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/specimen/core/bidirectional_blast.html#bdbp_diamond"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.bidirectional_blast.bdbp_diamond" title="Link to this definition"></a></dt>
<dd><p>Perform bidirectional blastp using DIAMOND.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>dir (str): </dt><dd><p>Path to the directory parent to in/out.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>template_name (str): </dt><dd><p>Name of the template genome.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>input_name (str): </dt><dd><p>Name of the input genome.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>template_path (str): </dt><dd><p>Path to the CDS FASTA-file of the template.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>input_path (str): </dt><dd><p>Path to the CDS FASTA-file of the input.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>sensitivity (str, optional): </dt><dd><p>Sensitivity mode for DIAMOND. 
Defaults to ‘sensitive’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>threads (int, optional): </dt><dd><p>Number of threads to use when running DIAMOND. 
Defaults to 2.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.create_diamond_db">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">create_diamond_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/specimen/core/bidirectional_blast.html#create_diamond_db"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.bidirectional_blast.create_diamond_db" title="Link to this definition"></a></dt>
<dd><p>Create a DIAMOND database for a given protein FASTA file.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>dir (str): </dt><dd><p>Path to the data directory.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>name (str): </dt><dd><p>Name of the genome/database.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>path (str): </dt><dd><p>Path to the FASTA-file.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>threads (int): </dt><dd><p>Number of threads to use.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.extract_bestbdbp_hits">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">extract_bestbdbp_hits</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tvq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qvt</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cov</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.25</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/specimen/core/bidirectional_blast.html#extract_bestbdbp_hits"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.bidirectional_blast.extract_bestbdbp_hits" title="Link to this definition"></a></dt>
<dd><p>Extract the best directional blastp hits from two tsv files,
which were generate by :py:func`~specimen.core.bidirectional_blast.bdbp_diamond` 
generated or similar steps.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>tvq (str): </dt><dd><p>Path to the template vs. query file.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>qvt (str): </dt><dd><p>Path to the query vs. template file.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>name (str): </dt><dd><p>Name (path) of the output file.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>cov (float, optional): </dt><dd><p>Cut-off value for the coverage.
All hits with coverage &lt; cov will be excluded. 
Defaults to 0.25.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.extract_cds">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">extract_cds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collect_info</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">identifier</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="../_modules/specimen/core/bidirectional_blast.html#extract_cds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.bidirectional_blast.extract_cds" title="Link to this definition"></a></dt>
<dd><p>Extract the CDS from a genbank file (annotated genome). 
Produces a FASTA-file.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>file (str): </dt><dd><p>File to extract CDS from.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>name (str): </dt><dd><p>Name of the genome.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>dir (str): </dt><dd><p>Directory for the ouput.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>collect_info (list): </dt><dd><p>Feature identifiers to collect information from.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>identifier (str): </dt><dd><p>Feature identifier to use of the header of the FASTA.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>str: </dt><dd><p>Name of the FASTA-file</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">template</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">template_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp_header</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'protein_id'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_header</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'locus_tag'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_info</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['locus_tag',</span> <span class="pre">'product',</span> <span class="pre">'protein_id']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensitivity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'sensitive'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'more-sensitive'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'very-sensitive'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'ultra-sensitive'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'more-sensitive'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/specimen/core/bidirectional_blast.html#run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.bidirectional_blast.run" title="Link to this definition"></a></dt>
<dd><p>Run the bidirectional blast on a template and input genome (annotated).</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>template (str): </dt><dd><p>Path to the annotated genome file used as a template.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>input (str): </dt><dd><p>Path to the annotated genome file used as a input.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>dir (str): </dt><dd><p>Path to the output directory.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>template_name (str, optional): </dt><dd><p>Name of the annotated genome file used as a template.. 
Defaults to None.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>input_name (str, optional): </dt><dd><p>Name of the annotated genome file used as input.. 
Defaults to None.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>temp_header (str, optional): </dt><dd><p>Feature qualifier of the gbff (NCBI) / faa (PROKKA) of the template to use as header for the FASTA files.
If None is given, sets it based on file extension (currently only implememted for gbff and faa).
Defaults to ‘protein_id’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>in_header (str, optional): </dt><dd><p>Feature qualifier of the gbff (NCBI) / faa (PROKKA) of the input to use as header for the FASTA files. 
If None is given, sets it based on file extension (currently only implememted for gbff and faa).
Defaults to ‘locus_tag’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>threads (int, optional):  </dt><dd><p>Number of threads to be used for DIAMOND. 
Defaults to 2.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>extra_info (list[str], optional):  </dt><dd><p>List of feature qualifiers to be extracted from the annotated genome files as additional information. 
Defaults to [‘locus_tag’, ‘product’, ‘protein_id’].</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>sensitivity (Literal[‘sensitive’, ‘more-sensitive’, ‘very-sensitive’, ‘ultra-sensitive’], optional): </dt><dd><p>Sensitivity mode for DIAMOND blastp run.. 
Defaults to ‘more-sensitive’.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Raises:</dt><dd><p>ValueError: Unknown file extension. Please set value for temp_header manually or check file.
ValueError: Unknown file extension. Please set value for in_header manually or check file.
ValueError: Unknown sensitive mode</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.bidirectional_blast.run_diamond_blastp">
<span class="sig-prename descclassname"><span class="pre">specimen.core.bidirectional_blast.</span></span><span class="sig-name descname"><span class="pre">run_diamond_blastp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fasta_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensitivity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/specimen/core/bidirectional_blast.html#run_diamond_blastp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.bidirectional_blast.run_diamond_blastp" title="Link to this definition"></a></dt>
<dd><p>Run DIAMOND blastp for a given database name and FASTA - relies on the structure
created by this file (bidirectional_blast.py).</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>dir (str): </dt><dd><p>Parent directory of the place to save the files to.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>db (str): </dt><dd><p>Name of the genome/database used as the database.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>query (str): </dt><dd><p>Name of the genome used as the query.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>fasta_path (str): </dt><dd><p>Path to the FASTA-file containing the CDS.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>sensitivity (str): </dt><dd><p>Sensitivity mode to use for DIAMOND blastp.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>threads (int): </dt><dd><p>Number of threads that will be used for running DIAMOND</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.core.generate_draft_model">
<span id="specimen-core-generate-draft-model"></span><h3>specimen.core.generate_draft_model<a class="headerlink" href="#module-specimen.core.generate_draft_model" title="Link to this heading"></a></h3>
<p>Generate a draft model from a template.</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.check_unchanged">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">check_unchanged</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">draft</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bbh</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Model</span></span></span><a class="reference internal" href="../_modules/specimen/core/generate_draft_model.html#check_unchanged"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.generate_draft_model.check_unchanged" title="Link to this definition"></a></dt>
<dd><p>Check the genes names (more correctly, the IDs) for still existing original col_names.
Depending on the case, decide if to keep or remove them.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>draft (cobra.Model): </dt><dd><p>The draft model currently in the making.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>bbh (pd.DataFrame): </dt><dd><p>The table from the bidirectional_blast.py script containing the bidirectional blastp best hits information.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>cobra.Model: </dt><dd><p>The model after the check and possible removal of genes.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.edit_template_identifiers">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">edit_template_identifiers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'no'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'dot-to-underscore'</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="../_modules/specimen/core/generate_draft_model.html#edit_template_identifiers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.generate_draft_model.edit_template_identifiers" title="Link to this definition"></a></dt>
<dd><p>Edit the subject IDs to fit the gene IDs of the template model.
Requires further extention, if needed edits are not included.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>data (pd.DataFrame): </dt><dd><p>The data frame containing the bidirectional blastp best hits information.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>edit (Literal[‘no’,’dot-to-underscore’]): </dt><dd><p>Type of edit to perform.
Currently possible options: no, dot-to-underscore.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>pd.DataFrame: </dt><dd><p>The (un)edited DataFrame.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.gen_draft_model">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">gen_draft_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bbh</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'no'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'dot-to-underscore'</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'default'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'BiGG'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'BiGG'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Model</span></span></span><a class="reference internal" href="../_modules/specimen/core/generate_draft_model.html#gen_draft_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.generate_draft_model.gen_draft_model" title="Link to this definition"></a></dt>
<dd><p>Generate a draft model from a template model and the results of a bidirectional blastp (blast best hits) table
and save it as a new model.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>model (cobra.Model): </dt><dd><p>The template model.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>bbh (pd.DataFrame): </dt><dd><p>The bidirectional blastp best hits table.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>name (str): </dt><dd><p>Name of the newly generated model.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>dir (str): </dt><dd><p>Path to the directory to save the new model in.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>edit (Literal[‘no’,’dot-to-underscore’):  </dt><dd><p>Type of edit to perform.
Currently possible options: no, dot-to-underscore.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>medium (str, optional):  </dt><dd><p>Name of the to be loaded from the refineGEMs database or ‘default’ = the one
from the template model. If given the keyword ‘exchanges’, will use all exchange reactions in the model as a medium.
Defaults to ‘default’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (Literal[‘BiGG’], optional): </dt><dd><p>Namespace of the model. 
Defaults to ‘BiGG’.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>cobra.Model: </dt><dd><p>The generated draft model.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.pid_filter">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">pid_filter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="../_modules/specimen/core/generate_draft_model.html#pid_filter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.generate_draft_model.pid_filter" title="Link to this definition"></a></dt>
<dd><p>Filter the data based on PID threshold. Entries above the given value are kept.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>data (pd.DataFrame): </dt><dd><p>The data from bidirectional_blast.py containing at least a ‘PID’ column.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>pid (float): </dt><dd><p>PID threshold value, given in percentage e.g. 80.0.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>pd.DataFrame: </dt><dd><p>The filtered data.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.remove_absent_genes">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">remove_absent_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Model</span></span></span><a class="reference internal" href="../_modules/specimen/core/generate_draft_model.html#remove_absent_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.generate_draft_model.remove_absent_genes" title="Link to this definition"></a></dt>
<dd><p>Remove a list of genes from a given model.
Note: genes that are not found in the model are skipped.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>model (cobra.Model): </dt><dd><p>A template model to delete genes from. 
A copy will be created before deleting.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>genes (list[str]): </dt><dd><p>Gene identifiers of genes that should be deleted.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>cobra.Model: </dt><dd><p>A new model with the given genes deleted, if found in the original model.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.rename_found_homologs">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">rename_found_homologs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">draft</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bbh</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Model</span></span></span><a class="reference internal" href="../_modules/specimen/core/generate_draft_model.html#rename_found_homologs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.generate_draft_model.rename_found_homologs" title="Link to this definition"></a></dt>
<dd><p>Rename the genes in the model correnspondingly to the homologous ones found in the query.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>draft (cobra.Model): </dt><dd><p>The draft model with the to-be-renamed genes.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>bbh (pd.DataFrame): </dt><dd><p>The table from the bidirectional_blast.py script containing the bidirectional blastp best hits information</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>cobra.Model: </dt><dd><p>The draft model with renamed genes.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.generate_draft_model.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.generate_draft_model.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">template</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bpbbh</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edit_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'no'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'dot-to-underscore'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'no'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">80.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'default'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'BiGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memote</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/specimen/core/generate_draft_model.html#run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.generate_draft_model.run" title="Link to this definition"></a></dt>
<dd><p>Generate a draft model from a blastp best hits tsv file and a template model.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>template (str): </dt><dd><p>Path to the file containing the template model.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>bpbbh (str): </dt><dd><p>Path to the blastp bidirectional best hits.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>dir (str): </dt><dd><p>Path to output directory.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>edit_names (Literal[‘no’,’dot-to-underscore’, optional):  </dt><dd><p>Type of edit to perform.
Currently possible options: no, dot-to-underscore. 
Defaults to ‘no’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>pid (float, optional): </dt><dd><p>Threshold value for determining, if a gene is counted as present or absent. 
Given in percentage, e.g. 80.0 = 80%.
Defaults to 80.0.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>name (Union[str,None], optional): </dt><dd><p>Name of the output model. 
If not given, takes name from filename. 
Defaults to None.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>medium (str, optional):  </dt><dd><p>Name of the to be loaded from the refineGEMs database or ‘default’ = the one
from the template model. If given the keyword ‘exchanges’, will use all exchange reactions in the model as a medium.
Defaults to ‘default’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>namespace (str, optional): </dt><dd><p>Namespace of the model. 
Defaults to ‘BiGG’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>memote (bool, optional): </dt><dd><p>Option to run memote after creating the draft model. 
Defaults to False.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Raises:</dt><dd><p>ValueError: ‘Edit_names value {edit_names} not in list of allowed values: no, dot-to-underscore’</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-specimen.core.validation">
<span id="specimen-core-validation"></span><h3>specimen.core.validation<a class="headerlink" href="#module-specimen.core.validation" title="Link to this heading"></a></h3>
<p>Validate a model (part 4 of the workflow).</p>
<p>Implemented tests in include:
- cobra/sbml check using cobrapy</p>
<dl class="py function">
<dt class="sig sig-object py" id="specimen.core.validation.run">
<span class="sig-prename descclassname"><span class="pre">specimen.core.validation.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tests</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'cobra'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_all</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/specimen/core/validation.html#run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#specimen.core.validation.run" title="Link to this definition"></a></dt>
<dd><p>SPECIMEN Step 4: Validate the model.</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>dir (str): </dt><dd><p>Path to the output directory.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>model_path (str): </dt><dd><p>Path to the model to be validated</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>tests (None | Literal[‘cobra’], optional): </dt><dd><p>Tests to perform. 
Defaults to None.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>run_all (bool, optional): </dt><dd><p>Run al available tests. If True, overwrites
the previous parameter.
Defaults to True.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="classes.html" class="btn btn-neutral float-left" title="specimen.classes package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="util.html" class="btn btn-neutral float-right" title="specimen.util package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Carolin Brune.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>